{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pandas as pd\n",
    "# import sqlite3\n",
    "# from PIL import Image\n",
    "# import matplotlib.pyplot as plt\n",
    "# from plotnine import *\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "# from collections import Counter\n",
    "# import seaborn as sns\n",
    "# import matplotlib.colors as mcolors\n",
    "# from matplotlib.ticker import FuncFormatter\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "# import patchworklib as pw\n",
    "# sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "# import lcwgsus\n",
    "# from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# Genuine logics changed from the main script were marked by string **X** in the original R files.\n",
    "# Simplicity changes were marked by string **Y**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'DRB1'\n",
    "nucleotides = ['A', 'T', 'C', 'G']\n",
    "ipd_gen_file = '/Users/sus_zhang/Desktop/Suuuuuuuus/misc_data/alignments/' + gene + '_gen.txt'\n",
    "if not os.path.exists(ipd_gen_file):\n",
    "    ipd_gen_file = '/well/band/users/rbx225/recyclable_files/hla/alignments/' + gene + '_gen.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-A</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29942554</td>\n",
       "      <td>29945741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31353367</td>\n",
       "      <td>31357155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31268257</td>\n",
       "      <td>31272071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-DMA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32948765</td>\n",
       "      <td>32951900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HLA-DMB</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32934806</td>\n",
       "      <td>32940044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HLA-DOA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33006304</td>\n",
       "      <td>33009150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-DOB</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32812540</td>\n",
       "      <td>32816899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-DPA1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33064319</td>\n",
       "      <td>33073562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-DPA2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33091485</td>\n",
       "      <td>33097139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33076065</td>\n",
       "      <td>33087147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HLA-DPB2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33113855</td>\n",
       "      <td>33129686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HLA-DQA1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32637480</td>\n",
       "      <td>32643199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HLA-DQA2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32741532</td>\n",
       "      <td>32747214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HLA-DQB1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32660035</td>\n",
       "      <td>32666603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32440129</td>\n",
       "      <td>32445274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578780</td>\n",
       "      <td>32589729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32518625</td>\n",
       "      <td>32530185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HLA-E</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30489540</td>\n",
       "      <td>30492916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HLA-F</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29723501</td>\n",
       "      <td>29726666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HLA-G</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29827859</td>\n",
       "      <td>29830682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HLA-HFE</td>\n",
       "      <td>chr6</td>\n",
       "      <td>26087319</td>\n",
       "      <td>26098343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HLA-H</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29887803</td>\n",
       "      <td>29890883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HLA-J</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30006723</td>\n",
       "      <td>30009476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HLA-K</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29926466</td>\n",
       "      <td>29929702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HLA-L</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30259648</td>\n",
       "      <td>30263000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HLA-MICA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31403653</td>\n",
       "      <td>31415816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HLA-MICB</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31498274</td>\n",
       "      <td>31510557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HLA-N</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30351570</td>\n",
       "      <td>30351761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HLA-P</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29800524</td>\n",
       "      <td>29802776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HLA-S</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31381834</td>\n",
       "      <td>31382377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HLA-TAP1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32845139</td>\n",
       "      <td>32853398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HLA-TAP2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32828449</td>\n",
       "      <td>32837693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HLA-T</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29896662</td>\n",
       "      <td>29898450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HLA-U</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29934121</td>\n",
       "      <td>29934596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HLA-V</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29792334</td>\n",
       "      <td>29793434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HLA-W</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29956609</td>\n",
       "      <td>29959055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name   Chr     Start       End\n",
       "0      HLA-A  chr6  29942554  29945741\n",
       "1      HLA-B  chr6  31353367  31357155\n",
       "2      HLA-C  chr6  31268257  31272071\n",
       "3    HLA-DMA  chr6  32948765  32951900\n",
       "4    HLA-DMB  chr6  32934806  32940044\n",
       "5    HLA-DOA  chr6  33006304  33009150\n",
       "6    HLA-DOB  chr6  32812540  32816899\n",
       "7   HLA-DPA1  chr6  33064319  33073562\n",
       "8   HLA-DPA2  chr6  33091485  33097139\n",
       "9   HLA-DPB1  chr6  33076065  33087147\n",
       "10  HLA-DPB2  chr6  33113855  33129686\n",
       "11  HLA-DQA1  chr6  32637480  32643199\n",
       "12  HLA-DQA2  chr6  32741532  32747214\n",
       "13  HLA-DQB1  chr6  32660035  32666603\n",
       "14   HLA-DRA  chr6  32440129  32445274\n",
       "15  HLA-DRB1  chr6  32578780  32589729\n",
       "16  HLA-DRB5  chr6  32518625  32530185\n",
       "17     HLA-E  chr6  30489540  30492916\n",
       "18     HLA-F  chr6  29723501  29726666\n",
       "19     HLA-G  chr6  29827859  29830682\n",
       "20   HLA-HFE  chr6  26087319  26098343\n",
       "21     HLA-H  chr6  29887803  29890883\n",
       "22     HLA-J  chr6  30006723  30009476\n",
       "23     HLA-K  chr6  29926466  29929702\n",
       "24     HLA-L  chr6  30259648  30263000\n",
       "25  HLA-MICA  chr6  31403653  31415816\n",
       "26  HLA-MICB  chr6  31498274  31510557\n",
       "27     HLA-N  chr6  30351570  30351761\n",
       "28     HLA-P  chr6  29800524  29802776\n",
       "29     HLA-S  chr6  31381834  31382377\n",
       "30  HLA-TAP1  chr6  32845139  32853398\n",
       "31  HLA-TAP2  chr6  32828449  32837693\n",
       "32     HLA-T  chr6  29896662  29898450\n",
       "33     HLA-U  chr6  29934121  29934596\n",
       "34     HLA-V  chr6  29792334  29793434\n",
       "35     HLA-W  chr6  29956609  29959055"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hlagenes.txt', sep = ' ')\n",
    "hla_gene_information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(ipd_gen_file, \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "gDNA_idx = []\n",
    "names = []\n",
    "i = 0\n",
    "while len(gDNA_idx) < 2:\n",
    "    l = lines[i]\n",
    "    if 'gDNA' in l:\n",
    "        gDNA_idx.append(i)\n",
    "    elif l.lstrip(' ').split(' ')[0].startswith(gene + '*'):\n",
    "        name = l.lstrip(' ').split(' ')[0]\n",
    "        names.append(name)\n",
    "    i += 1\n",
    "    \n",
    "first_base = int(lines[gDNA_idx[0]].split(' ')[-1].split('\\n')[0])\n",
    "n_alleles = gDNA_idx[1] - gDNA_idx[0] - 3\n",
    "\n",
    "alleles_dict = {k:'' for k in names}\n",
    "for i, s in enumerate(lines):\n",
    "    r = s.lstrip(' ')\n",
    "    if r.startswith(gene):\n",
    "        r = r.rstrip(' \\n')\n",
    "        name = r.split(' ')[0]\n",
    "        sequence = r.split(' ')[2:]\n",
    "        sequence = ''.join(sequence)\n",
    "        alleles_dict[name] = alleles_dict[name] + sequence\n",
    "    \n",
    "df = pd.DataFrame({key: list(value) for key, value in alleles_dict.items()}).T\n",
    "df = df.drop(columns=df.columns[df.eq('|').all()])\n",
    "df.columns = range(df.shape[1])\n",
    "\n",
    "length = len(df.columns)\n",
    "positions = [first_base]*length\n",
    "\n",
    "df.loc[len(df)] = positions\n",
    "\n",
    "r_idx = len(df) - 1\n",
    "for i in df.columns[1:]:\n",
    "    if df.iloc[0, i] == '.':\n",
    "        df.iloc[r_idx, i] = df.iloc[r_idx, i-1]\n",
    "    else:\n",
    "        if df.iloc[r_idx, i-1] != -1:\n",
    "            df.iloc[r_idx, i] = df.iloc[r_idx, i-1] + 1\n",
    "        else:\n",
    "            df.iloc[r_idx, i] = df.iloc[r_idx, i-1] + 2 # No zero in position. ATG is encoded by 1.\n",
    "df.columns = df.iloc[r_idx]\n",
    "df = df.iloc[:r_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_refdir = '/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/QUILT_ref_files/'\n",
    "# all_haplotypes = pyreadr.read_r(original_refdir + 'quilt.hrc.hla.all.haplotypes.RData')\n",
    "snpformatalleles = pyreadr.read_r(original_refdir + 'hlaDRB1snpformatalleles.RData')\n",
    "fullallelesfilledin = pyreadr.read_r(original_refdir + 'HLADRB1fullallelesfilledin.RData')\n",
    "full = pyreadr.read_r(original_refdir + 'hlaDRB1full.RData')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oldsnpinfo -> prepared_vcf\n",
    "# ss -> variant_alleles\n",
    "# samples -> alleles\n",
    "# qq -> positions\n",
    "knownvarsfiltered = snpformatalleles['knownvarsfiltered'].copy()\n",
    "resmat = snpformatalleles['resmat'].copy()\n",
    "knownvarsfiltered = knownvarsfiltered[knownvarsfiltered[1].isin(nucleotides) & knownvarsfiltered[2].isin(nucleotides)].reset_index(drop = True)\n",
    "\n",
    "ids = \"chr6:\" + knownvarsfiltered[0].astype(str)\n",
    "prepared_vcf = pd.concat([ids, knownvarsfiltered], axis=1)\n",
    "alleles = resmat.index\n",
    "prepared_vcf.columns = [\"id\", \"position\", \"a0\", \"a1\"]\n",
    "prepared_vcf['position']  = prepared_vcf['position'].astype(int)\n",
    "\n",
    "ourpos = fullallelesfilledin['ourpos']['ourpos'].tolist()\n",
    "fullalleles = fullallelesfilledin['fullalleles'].copy()\n",
    "\n",
    "retained_index = []\n",
    "variant_alleles = pd.DataFrame(index=fullalleles.index)\n",
    "\n",
    "for i in range(len(prepared_vcf)):\n",
    "    _, pos, ref, alt = prepared_vcf.iloc[i,:]\n",
    "    index = np.where(ourpos == pos)[0]\n",
    "    if index.size != 0:\n",
    "        index = index[0]\n",
    "        retained_index.append(index)\n",
    "        variant_alleles[variant_alleles.shape[1]] = fullalleles[index].map({ref: '0', alt: '1'}).fillna(fullalleles[index])\n",
    "    else:\n",
    "        prepared_vcf.iloc[i,2] = 'tbr' # tbr: To be removed\n",
    "\n",
    "prepared_vcf = prepared_vcf[prepared_vcf['a0'] != 'tbr'].reset_index(drop = True)\n",
    "variant_alleles = variant_alleles.T\n",
    "\n",
    "positions = prepared_vcf['position'].astype(float).values  # Convert position to double\n",
    "\n",
    "zz = np.zeros(variant_alleles.shape[0]) \n",
    "for i in range(len(zz)):\n",
    "    zz[i] = np.sum(ourpos == positions[i])\n",
    "zz2 = np.zeros_like(zz)\n",
    "zz2 = np.sum(~((variant_alleles == '0') | (variant_alleles == '1')), axis=1)\n",
    "zz3 = np.zeros_like(zz)\n",
    "for i in range(len(zz3)):\n",
    "    zz3[i] = np.sum(positions == positions[i])\n",
    "zz4 = np.sum(~((variant_alleles == '0') | (variant_alleles == '1') | (variant_alleles == \".\")), axis=1)\n",
    "filter_conditions = (zz3 == 1) & (zz == 1) & (zz2 < variant_alleles.shape[1] * 0.1) & (zz2 < 0.5 * np.sum(variant_alleles == '1', axis=1)) & (zz4 == 0)\n",
    "\n",
    "# keep sites uniquely mapping, not overlapping another SNP, with at most 10% gaps and 2-fold more non-ancestral than gaps\n",
    "snpinfo = prepared_vcf.loc[filter_conditions, :]\n",
    "variant_alleles = variant_alleles.loc[filter_conditions, :]\n",
    "variant_alleles[variant_alleles == \".\"] = 0 # indels set to 0\n",
    "haps = variant_alleles.copy().reset_index(drop = True) # Now everything is in string format\n",
    "\n",
    "# Continue from line 244"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
