{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import pickle\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "# import sqlite3\n",
    "# from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "# from plotnine import *\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "# from collections import Counter\n",
    "# import seaborn as sns\n",
    "# import matplotlib.colors as mcolors\n",
    "# from matplotlib.ticker import FuncFormatter\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "import psa\n",
    "import pywfa\n",
    "from IPython.display import display_html\n",
    "# import patchworklib as pw\n",
    "sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "from hla_phase import *\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# Genuine logics changed from the main script were marked by string **X** in the original R files.\n",
    "# Simplicity changes were marked by string **Y**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "bam = \"/well/band/users/rbx225/GAMCC/data/bams/IDT0482.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>18517</th>\n",
       "      <th>18516</th>\n",
       "      <th>18515</th>\n",
       "      <th>18514</th>\n",
       "      <th>18513</th>\n",
       "      <th>18512</th>\n",
       "      <th>18511</th>\n",
       "      <th>18510</th>\n",
       "      <th>18509</th>\n",
       "      <th>18508</th>\n",
       "      <th>...</th>\n",
       "      <th>9</th>\n",
       "      <th>8</th>\n",
       "      <th>7</th>\n",
       "      <th>6</th>\n",
       "      <th>5</th>\n",
       "      <th>4</th>\n",
       "      <th>3</th>\n",
       "      <th>2</th>\n",
       "      <th>1</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRB1*01:01:01:01</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*01:01:01:02</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*01:01:01:03</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*01:01:01:04</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*01:01:01:05</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*16:02:01:06</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*16:02:10</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*16:05:01</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*16:09:01</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*16:74</th>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>...</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>624 rows × 18518 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 18517 18516 18515 18514 18513 18512 18511 18510 18509 18508  \\\n",
       "DRB1*01:01:01:01     *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:02     *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:03     *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:04     *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:05     *     *     *     *     *     *     *     *     *     *   \n",
       "...                ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "DRB1*16:02:01:06     *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:02:10        *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:05:01        *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:09:01        *     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:74           *     *     *     *     *     *     *     *     *     *   \n",
       "\n",
       "                  ... 9     8     7     6     5     4     3     2     1      \\\n",
       "DRB1*01:01:01:01  ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:02  ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:03  ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:04  ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*01:01:01:05  ...     *     *     *     *     *     *     *     *     *   \n",
       "...               ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "DRB1*16:02:01:06  ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:02:10     ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:05:01     ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:09:01     ...     *     *     *     *     *     *     *     *     *   \n",
       "DRB1*16:74        ...     *     *     *     *     *     *     *     *     *   \n",
       "\n",
       "                 0      \n",
       "DRB1*01:01:01:01     *  \n",
       "DRB1*01:01:01:02     *  \n",
       "DRB1*01:01:01:03     *  \n",
       "DRB1*01:01:01:04     *  \n",
       "DRB1*01:01:01:05     *  \n",
       "...                ...  \n",
       "DRB1*16:02:01:06     *  \n",
       "DRB1*16:02:10        *  \n",
       "DRB1*16:05:01        *  \n",
       "DRB1*16:09:01        *  \n",
       "DRB1*16:74           *  \n",
       "\n",
       "[624 rows x 18518 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# def process_db_genfile(gene, \n",
    "#                             ipd_gen_file_dir, \n",
    "#                             hla_gene_information, \n",
    "#                             hlatypes,\n",
    "#                             phased_vcf, \n",
    "#                             reference_allele_ary, \n",
    "#                             strict_snp_filter = True,\n",
    "#                             read_from_QUILT = False, \n",
    "#                             subset_vcf_samples = None,\n",
    "#                             sample_linker = None):\n",
    "    \n",
    "strand = hla_gene_information[hla_gene_information['Name'] == ('HLA-' + gene)]['Strand'].iloc[0]\n",
    "#     reference_allele = reference_allele_ary[np.char.find(reference_allele_ary, gene) != -1][0]\n",
    "nucleotides = ['A', 'T', 'C', 'G']\n",
    "\n",
    "ipd_gen_file = ipd_gen_file_dir + gene + '_gen.txt'\n",
    "with open(ipd_gen_file, \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "gDNA_idx = []\n",
    "names = []\n",
    "i = 0\n",
    "while len(gDNA_idx) < 2:\n",
    "    l = lines[i]\n",
    "    if 'gDNA' in l:\n",
    "        gDNA_idx.append(i)\n",
    "    elif l.lstrip(' ').split(' ')[0].startswith(gene + '*'):\n",
    "        name = l.lstrip(' ').split(' ')[0]\n",
    "        names.append(name)\n",
    "    i += 1\n",
    "\n",
    "first_base = int(lines[gDNA_idx[0]].split(' ')[-1].split('\\n')[0])\n",
    "n_alleles = gDNA_idx[1] - gDNA_idx[0] - 3\n",
    "\n",
    "alleles_dict = {k:'' for k in names}\n",
    "for i, s in enumerate(lines):\n",
    "    r = s.lstrip(' ')\n",
    "    if r.startswith(gene):\n",
    "        r = r.rstrip(' \\n')\n",
    "        name = r.split(' ')[0]\n",
    "        sequence = r.split(' ')[1:]\n",
    "        sequence = ''.join(sequence)\n",
    "        alleles_dict[name] = alleles_dict[name] + sequence\n",
    "\n",
    "db = pd.DataFrame({key: list(value) for key, value in alleles_dict.items()}).T\n",
    "db = db.drop(columns=db.columns[db.eq('|').all()])\n",
    "db.columns = range(db.shape[1])\n",
    "\n",
    "db = db.apply(lambda c: c.replace('-', c[0]) ,axis = 0)\n",
    "\n",
    "if strand != 1:\n",
    "    db.replace({'A': 't', 'C': 'g', 'G': 'c', 'T': 'a'}, inplace=True)\n",
    "    db.replace({'a': 'A', 'c': 'C', 'g': 'G', 't': 'T'}, inplace=True)\n",
    "    db = db.iloc[:, ::-1]\n",
    "\n",
    "db\n",
    "\n",
    "# #     variants_df = []\n",
    "# #     for i in range(len(db)):\n",
    "# #         variants_df.append(extract_snps_from_row(db.loc[reference_allele,:], db.iloc[i,], allele_pos))\n",
    "# #     vcf = pd.concat(variants_df).drop_duplicates()\n",
    "# #     vcf = vcf[(vcf['ref'].isin(nucleotides)) & (vcf['alt'].isin(nucleotides))].sort_values(by = 'pos').reset_index(drop = True)\n",
    "\n",
    "# #     allele_pos = np.array(allele_pos)\n",
    "# dbcopy = pd.DataFrame()\n",
    "# #     aligned_df = pd.DataFrame()\n",
    "# variants_to_remove = []\n",
    "\n",
    "# for i in range(vcf.shape[0]):\n",
    "#     pos, a1, a2 = vcf.iloc[i, :3]\n",
    "#     index = np.where((allele_pos == pos))[0]\n",
    "\n",
    "#     if len(index) == 1:\n",
    "#         index = index[0]\n",
    "#         dbcopy[dbcopy.shape[1]] = db[index]        \n",
    "# #             aligned_df[aligned_df.shape[1]] = db[index].apply(recode_variant, args = (a2))\n",
    "#     else:\n",
    "#         variants_to_remove.append(i)\n",
    "\n",
    "# #     vcf = vcf[~vcf.index.isin(variants_to_remove)].reset_index(drop = True) \n",
    "# # This is a remedy when extract_snps_from_row fails to detect insertions. \n",
    "# # For DRB1 we get 2442 variants instead of 2452, fairly good I'd say\n",
    "# dbfull = db\n",
    "# db = dbcopy\n",
    "\n",
    "# #     sample = pd.DataFrame({\n",
    "# #         \"ID_1\": aligned_df.index,\n",
    "# #         \"ID_2\": aligned_df.index,\n",
    "# #         \"missing\": 0\n",
    "# #     })\n",
    "# #     vcf['id'] = 'chr6:' + vcf['pos'].astype(str)\n",
    "# #     haps = aligned_df.T\n",
    "# #     cond = (haps == -1).mean(axis=1) < 0.1\n",
    "# #     haps = haps[cond].reset_index(drop = True)\n",
    "# #     vcf = vcf[cond].reset_index(drop = True)\n",
    "\n",
    "# #     start = vcf.loc[0, 'pos']\n",
    "# #     end =vcf.loc[len(vcf) - 1, 'pos']\n",
    "\n",
    "# #     cond = (allele_pos >= start) & (allele_pos <= end)\n",
    "# #     allele_pos = allele_pos[cond]\n",
    "# #     dbfull = dbfull.iloc[:, np.where(cond == True)[0]]\n",
    "# #     dbfull.columns = range(dbfull.shape[1])\n",
    "\n",
    "# #     # Imputing database alleles\n",
    "# #     for i, a in enumerate(haps.columns):\n",
    "# #         haps_undetermined = haps[a] == -1\n",
    "\n",
    "# #         allele_col = haps[~haps_undetermined][a]\n",
    "# #         subseted_db =  haps[~haps_undetermined] \n",
    "# #         pairwise_distance = subseted_db.sub(subseted_db[a], axis=0)\n",
    "# #         counts = pairwise_distance.abs().sum().sort_values().iloc[1:]\n",
    "# #         closest_allele_idx = 0\n",
    "# #         substituted_nucleotides = np.array([np.nan]*np.sum(haps_undetermined))\n",
    "# #         substituted_indices = haps_undetermined[haps_undetermined == True].index\n",
    "\n",
    "# #         while np.isnan(substituted_nucleotides).any() > 0:\n",
    "# #             closest_allele = counts.index[closest_allele_idx]\n",
    "\n",
    "# #             target_haplotype = haps.loc[substituted_indices, closest_allele]\n",
    "# #             substitutable = target_haplotype[target_haplotype.values != -1]\n",
    "# #             substituted_nucleotides[np.where(np.in1d(substituted_indices, substitutable.index))[0]] = substitutable.values\n",
    "# #             closest_allele_idx += 1\n",
    "\n",
    "# #         haps.loc[substituted_indices, a] = substituted_nucleotides\n",
    "# #         # haps has dim n_variants * n_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>flag</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>map_quality</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>chr_alt</th>\n",
       "      <th>pos_alt</th>\n",
       "      <th>insert_size</th>\n",
       "      <th>sequence</th>\n",
       "      <th>base_quality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:1164:15157:28463</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32580978</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32581049</td>\n",
       "      <td>222</td>\n",
       "      <td>GCCATTTCGGGAGAAAAAAAGGATTTCAAATCACACTGAACAGTTA...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:1164:15157:28463</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32581049</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32580978</td>\n",
       "      <td>-222</td>\n",
       "      <td>AAATATTACAGCCTTGATGTAAGGCAAGAGTTCAACATCTGATCCA...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:1501:14696:8249</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32589048</td>\n",
       "      <td>60</td>\n",
       "      <td>139M12S</td>\n",
       "      <td>=</td>\n",
       "      <td>32588932</td>\n",
       "      <td>-255</td>\n",
       "      <td>AAAGAAAGAGCACTAAGTATAACTTCTGTCAGACAACCTACATACA...</td>\n",
       "      <td>FFF:FFFFFFFFFFFFFFFFF:FFFFFFF:FFFFFFFF,FF:FFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:1501:14696:8249</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32588932</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32589048</td>\n",
       "      <td>255</td>\n",
       "      <td>GAGGATGCCTTAAGTTCTTTAGGCACCAAAGAATACCTCATAAATG...</td>\n",
       "      <td>FFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:1619:32280:8046</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32583021</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32582884</td>\n",
       "      <td>-288</td>\n",
       "      <td>CATACAAACCACAAACTGGAGCAAATGCTGAATCAAATATCAATAA...</td>\n",
       "      <td>FFFFFFFFF:FFFFFFFF,FFFF,FFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2570:5222:23093</td>\n",
       "      <td>147</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32590480</td>\n",
       "      <td>40</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32590322</td>\n",
       "      <td>-309</td>\n",
       "      <td>TCTTAGTAAAAAGAGTTTTTTCAAGTGTATCTCCTGATGCTTCCCC...</td>\n",
       "      <td>FFFF,FFFFFFF::FF:FFFFFFFFFF:FFFF:FFFFF:F:FFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2635:21630:12649</td>\n",
       "      <td>99</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32590175</td>\n",
       "      <td>51</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32590246</td>\n",
       "      <td>218</td>\n",
       "      <td>AGATGGAGGGAACATGGTTGAAAATCAGGACACCTTTAATCTGGTC...</td>\n",
       "      <td>FFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2635:21630:12649</td>\n",
       "      <td>147</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32590246</td>\n",
       "      <td>40</td>\n",
       "      <td>79M3I36M1I32M</td>\n",
       "      <td>=</td>\n",
       "      <td>32590175</td>\n",
       "      <td>-218</td>\n",
       "      <td>TAGATTTGGGAAAATTACTTCATGTCTCACAGTTGATATGAAGGCA...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2640:3405:19069</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578686</td>\n",
       "      <td>60</td>\n",
       "      <td>37M10I104M</td>\n",
       "      <td>=</td>\n",
       "      <td>32578872</td>\n",
       "      <td>337</td>\n",
       "      <td>GTGAAAATCATTATTCTGTTAGGAAATAATCTTCCCTTTTTCTTTT...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFF:FFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2640:3405:19069</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578872</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32578686</td>\n",
       "      <td>-337</td>\n",
       "      <td>AAAGTTGAAGATGAGGCGCTGCCATCAATGCTGGGACTTCAGGCCA...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         ID flag   chr       pos map_quality  \\\n",
       "0   A00711:748:HHFNVDSX7:4:1164:15157:28463  163  chr6  32580978          60   \n",
       "1   A00711:748:HHFNVDSX7:4:1164:15157:28463   83  chr6  32581049          60   \n",
       "2    A00711:748:HHFNVDSX7:4:1501:14696:8249   83  chr6  32589048          60   \n",
       "3    A00711:748:HHFNVDSX7:4:1501:14696:8249  163  chr6  32588932          60   \n",
       "4    A00711:748:HHFNVDSX7:4:1619:32280:8046   83  chr6  32583021          60   \n",
       "..                                      ...  ...   ...       ...         ...   \n",
       "79   A00711:753:HHCG3DSX7:4:2570:5222:23093  147  chr6  32590480          40   \n",
       "80  A00711:753:HHCG3DSX7:4:2635:21630:12649   99  chr6  32590175          51   \n",
       "81  A00711:753:HHCG3DSX7:4:2635:21630:12649  147  chr6  32590246          40   \n",
       "82   A00711:753:HHCG3DSX7:4:2640:3405:19069  163  chr6  32578686          60   \n",
       "83   A00711:753:HHCG3DSX7:4:2640:3405:19069   83  chr6  32578872          60   \n",
       "\n",
       "            CIGAR chr_alt   pos_alt insert_size  \\\n",
       "0            151M       =  32581049         222   \n",
       "1            151M       =  32580978        -222   \n",
       "2         139M12S       =  32588932        -255   \n",
       "3            151M       =  32589048         255   \n",
       "4            151M       =  32582884        -288   \n",
       "..            ...     ...       ...         ...   \n",
       "79           151M       =  32590322        -309   \n",
       "80           151M       =  32590246         218   \n",
       "81  79M3I36M1I32M       =  32590175        -218   \n",
       "82     37M10I104M       =  32578872         337   \n",
       "83           151M       =  32578686        -337   \n",
       "\n",
       "                                             sequence  \\\n",
       "0   GCCATTTCGGGAGAAAAAAAGGATTTCAAATCACACTGAACAGTTA...   \n",
       "1   AAATATTACAGCCTTGATGTAAGGCAAGAGTTCAACATCTGATCCA...   \n",
       "2   AAAGAAAGAGCACTAAGTATAACTTCTGTCAGACAACCTACATACA...   \n",
       "3   GAGGATGCCTTAAGTTCTTTAGGCACCAAAGAATACCTCATAAATG...   \n",
       "4   CATACAAACCACAAACTGGAGCAAATGCTGAATCAAATATCAATAA...   \n",
       "..                                                ...   \n",
       "79  TCTTAGTAAAAAGAGTTTTTTCAAGTGTATCTCCTGATGCTTCCCC...   \n",
       "80  AGATGGAGGGAACATGGTTGAAAATCAGGACACCTTTAATCTGGTC...   \n",
       "81  TAGATTTGGGAAAATTACTTCATGTCTCACAGTTGATATGAAGGCA...   \n",
       "82  GTGAAAATCATTATTCTGTTAGGAAATAATCTTCCCTTTTTCTTTT...   \n",
       "83  AAAGTTGAAGATGAGGCGCTGCCATCAATGCTGGGACTTCAGGCCA...   \n",
       "\n",
       "                                         base_quality  \n",
       "0   FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "1   FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFF...  \n",
       "2   FFF:FFFFFFFFFFFFFFFFF:FFFFFFF:FFFFFFFF,FF:FFFF...  \n",
       "3   FFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "4   FFFFFFFFF:FFFFFFFF,FFFF,FFFFFFFFFFFFFFFFFFFFFF...  \n",
       "..                                                ...  \n",
       "79  FFFF,FFFFFFF::FF:FFFFFFFFFF:FFFF:FFFFF:F:FFFFF...  \n",
       "80  FFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "81  FFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "82  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFF:FFFFFFF...  \n",
       "83  FFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFF...  \n",
       "\n",
       "[84 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regstart = hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['Start'].values[0]\n",
    "regend = hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['End'].values[0]\n",
    "strand = hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['Strand'].values[0]\n",
    "\n",
    "reads_apart_max = 1000\n",
    "\n",
    "command = f\"samtools view {bam} chr6:{regstart-reads_apart_max}-{regend+reads_apart_max}\"\n",
    "reads = subprocess.run(command, shell = True, capture_output = True, text = True).stdout[:-1].split('\\n')\n",
    "\n",
    "reads = [i.split('\\t') for i in reads if '##' not in i]\n",
    "reads = pd.DataFrame(reads).iloc[:,:11]\n",
    "reads.columns = [\n",
    "    \"ID\", \"flag\", \"chr\", \"pos\", \"map_quality\", \"CIGAR\", \"chr_alt\", \"pos_alt\", \"insert_size\", \"sequence\", \"base_quality\"\n",
    "]\n",
    "\n",
    "reads['pos'] = reads['pos'].astype(int)\n",
    "reads['pos_alt'] = reads['pos_alt'].astype(int)\n",
    "\n",
    "reads = reads[\n",
    "    (reads['chr'] == 'chr6') &\n",
    "    (reads['chr_alt'].isin(['chr6', '='])) &\n",
    "    (reads['pos_alt'] + reads['sequence'].str.len() >= np.ones(len(reads))*(regstart - reads_apart_max)) & \n",
    "    (reads['pos_alt'] <= np.ones(len(reads))*(regend + reads_apart_max))\n",
    "]\n",
    "\n",
    "id_counts = reads['ID'].value_counts()\n",
    "valid_ids_ary = id_counts[id_counts == 2].index.tolist()\n",
    "reads = reads[reads['ID'].isin(valid_ids_ary)].sort_values(by = 'ID').reset_index(drop = True)\n",
    "\n",
    "reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-8"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pywfa import WavefrontAligner\n",
    "\n",
    "pattern = \"*CTTTACTCGCGCGTTGGAGAAATAAAATAGT\"\n",
    "text =      \"*CTTTACTCGCGCGTTTGGAGAAATAAAATAGT\"\n",
    "a = WavefrontAligner(pattern)\n",
    "score = a.wavefront_align(text)\n",
    "score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4I7M5D6M\n",
      "-30\n"
     ]
    }
   ],
   "source": [
    "import pywfa\n",
    "\n",
    "pattern = \"AAAAACCTTTTTAAAAAA\"\n",
    "text =       \"GGCCAAAAACCAAAAAA\"\n",
    "a = WavefrontAligner(pattern)\n",
    "\n",
    "res = a(text, clip_cigar=False)\n",
    "print(pywfa.cigartuples_to_str(res.cigartuples))\n",
    "\n",
    "print(a.wavefront_align(text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "ref_samples = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/oneKG.samples', sep = ' ')\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "ref_samples_removed = ref_samples[~ref_samples['SAMPLE'].isin(hlatypes['Sample ID'].tolist())]\n",
    "samples_to_remove = ref_samples_removed['SAMPLE'].tolist()\n",
    "hlatypes = hlatypes[~hlatypes['Sample ID'].isin(samples_to_remove)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/oneKG.chr6.vcf.gz\"\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "\n",
    "# subset_samples = None\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information,\n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            read_from_QUILT = False, \n",
    "                            subset_vcf_samples = None,\n",
    "                            sample_linker = None)\n",
    "\n",
    "individual = 'NA12878'\n",
    "ix = hlatypes.index[hlatypes['Sample ID'] == 'NA12878'][0]\n",
    "display_indices = np.arange(10)\n",
    "\n",
    "res = visualise_phase(gene, ix, hlatypes, return_dict)\n",
    "compare_phase(display_indices, res)\n",
    "\n",
    "# reference_exclude_samplelist_file = ''\n",
    "# if reference_exclude_samplelist_file != '':\n",
    "#     samples_to_remove = samples_to_remove + lcwgsus.read_tsv_as_lst(reference_exclude_samplelist_file)\n",
    "# if samples_to_remove is not none, save to \"hlauntyped.exclude.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For GAMCC\n",
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "\n",
    "subset_vcf_samples = lcwgsus.read_tsv_as_lst('/well/band/users/rbx225/GAMCC/data/sample_tsvs/fv_gm_names.tsv')\n",
    "subset_vcf_samples = ','.join(subset_vcf_samples)\n",
    "\n",
    "sample_linker = pd.read_csv('/well/band/users/rbx225/GAMCC/' + SAMPLE_LINKER_FILE)\n",
    "sample_linker = {k:v for k, v in zip(sample_linker['Sample_Name'], sample_linker['Chip_Name'])}\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information, \n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            read_from_QUILT = True, \n",
    "                            subset_vcf_samples = subset_vcf_samples,\n",
    "                            sample_linker = sample_linker)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
