{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import pickle\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "# import sqlite3\n",
    "# from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "# from plotnine import *\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "# from collections import Counter\n",
    "# import seaborn as sns\n",
    "# import matplotlib.colors as mcolors\n",
    "# from matplotlib.ticker import FuncFormatter\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "import psa\n",
    "import pywfa\n",
    "from IPython.display import display_html, IFrame\n",
    "# import patchworklib as pw\n",
    "sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "from hla_phase import *\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# Genuine logics changed from the main script were marked by string **X** in the original R files.\n",
    "# Simplicity changes were marked by string **Y**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "positional indexers are out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 31\u001b[0m\n\u001b[1;32m     28\u001b[0m html \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m1KG-DRB1.html\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     30\u001b[0m res \u001b[38;5;241m=\u001b[39m visualise_phase(gene, ix, hlatypes, return_dict, both_het \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 31\u001b[0m \u001b[43mcompare_phase\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdisplay_indices\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mres\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msave_html\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msave_name\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mhtml\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;66;03m# reference_exclude_samplelist_file = ''\u001b[39;00m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;66;03m# if reference_exclude_samplelist_file != '':\u001b[39;00m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m#     samples_to_remove = samples_to_remove + lcwgsus.read_tsv_as_lst(reference_exclude_samplelist_file)\u001b[39;00m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;66;03m# if samples_to_remove is not none, save to \"hlauntyped.exclude.txt\"\u001b[39;00m\n",
      "File \u001b[0;32m/gpfs3/well/band/users/rbx225/software/QUILT_sus/QUILT/Python/hla_phase.py:108\u001b[0m, in \u001b[0;36mcompare_phase\u001b[0;34m(indices, res, save_html, save_name)\u001b[0m\n\u001b[1;32m    107\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompare_phase\u001b[39m(indices, res, save_html \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m, save_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 108\u001b[0m     vcf_html \u001b[38;5;241m=\u001b[39m \u001b[43mres\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhet_variants\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mindices\u001b[49m\u001b[43m,\u001b[49m\u001b[43m:\u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mto_html()\n\u001b[1;32m    109\u001b[0m     df1_html \u001b[38;5;241m=\u001b[39m res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpossible_p1\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[indices,:]\u001b[38;5;241m.\u001b[39mto_html()\n\u001b[1;32m    110\u001b[0m     df2_html \u001b[38;5;241m=\u001b[39m res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpossible_p2\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[indices,:]\u001b[38;5;241m.\u001b[39mto_html()\n",
      "File \u001b[0;32m/well/band/users/rbx225/conda/skylake/envs/sus/lib/python3.8/site-packages/pandas/core/indexing.py:1097\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1095\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_scalar_access(key):\n\u001b[1;32m   1096\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_value(\u001b[38;5;241m*\u001b[39mkey, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n\u001b[0;32m-> 1097\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_tuple\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1098\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1099\u001b[0m     \u001b[38;5;66;03m# we by definition only have the 0th axis\u001b[39;00m\n\u001b[1;32m   1100\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m/well/band/users/rbx225/conda/skylake/envs/sus/lib/python3.8/site-packages/pandas/core/indexing.py:1594\u001b[0m, in \u001b[0;36m_iLocIndexer._getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1593\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_getitem_tuple\u001b[39m(\u001b[38;5;28mself\u001b[39m, tup: \u001b[38;5;28mtuple\u001b[39m):\n\u001b[0;32m-> 1594\u001b[0m     tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_tuple_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtup\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1595\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m suppress(IndexingError):\n\u001b[1;32m   1596\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_lowerdim(tup)\n",
      "File \u001b[0;32m/well/band/users/rbx225/conda/skylake/envs/sus/lib/python3.8/site-packages/pandas/core/indexing.py:904\u001b[0m, in \u001b[0;36m_LocationIndexer._validate_tuple_indexer\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    902\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, k \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(key):\n\u001b[1;32m    903\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 904\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_key\u001b[49m\u001b[43m(\u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mi\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    905\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    906\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    907\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLocation based indexing can only have \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    908\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_valid_types\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] types\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    909\u001b[0m         ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n",
      "File \u001b[0;32m/well/band/users/rbx225/conda/skylake/envs/sus/lib/python3.8/site-packages/pandas/core/indexing.py:1516\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_key\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1514\u001b[0m     \u001b[38;5;66;03m# check that the key does not exceed the maximum size of the index\u001b[39;00m\n\u001b[1;32m   1515\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(arr) \u001b[38;5;129;01mand\u001b[39;00m (arr\u001b[38;5;241m.\u001b[39mmax() \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m len_axis \u001b[38;5;129;01mor\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mmin() \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m-\u001b[39mlen_axis):\n\u001b[0;32m-> 1516\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpositional indexers are out-of-bounds\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1518\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCan only index by location with a [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_valid_types\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mIndexError\u001b[0m: positional indexers are out-of-bounds"
     ]
    }
   ],
   "source": [
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "ref_samples = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/oneKG.samples', sep = ' ')\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "ref_samples_removed = ref_samples[~ref_samples['SAMPLE'].isin(hlatypes['Sample ID'].tolist())]\n",
    "samples_to_remove = ref_samples_removed['SAMPLE'].tolist()\n",
    "hlatypes = hlatypes[~hlatypes['Sample ID'].isin(samples_to_remove)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/oneKG.chr6.vcf.gz\"\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information,\n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            read_from_QUILT = False, \n",
    "                            subset_vcf_samples = None,\n",
    "                            sample_linker = None)\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "individual = 'NA12878'\n",
    "ix = hlatypes.index[hlatypes['Sample ID'] == 'NA12878'][0]\n",
    "display_indices = np.arange(10)\n",
    "html = '1KG-DRB1.html'\n",
    "\n",
    "res = visualise_phase(gene, ix, hlatypes, return_dict, both_het = True)\n",
    "compare_phase(display_indices, res, save_html = True, save_name = html)\n",
    "\n",
    "plot_first_step_phase(return_dict)\n",
    "plot_last_step_phase(return_dict)\n",
    "# reference_exclude_samplelist_file = ''\n",
    "# if reference_exclude_samplelist_file != '':\n",
    "#     samples_to_remove = samples_to_remove + lcwgsus.read_tsv_as_lst(reference_exclude_samplelist_file)\n",
    "# if samples_to_remove is not none, save to \"hlauntyped.exclude.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unphased samples: 3\n",
      "Percentage agreed with QUILT phasing for DRB1:  97.38791423001949 %\n"
     ]
    }
   ],
   "source": [
    "quilt_phase = pd.read_csv('/well/band/users/rbx225/recyclable_files/hla/1KG_phased_' + gene + '.tsv', sep = '\\t')\n",
    "\n",
    "phase_df = return_dict['phase_df']\n",
    "phased_indices = phase_df.index[(phase_df['final_step_phase1'] | phase_df['final_step_phase2'])]\n",
    "print('Number of unphased samples:', len(phase_df) - len(phased_indices))\n",
    "\n",
    "df1 = phase_df.loc[phased_indices,:]\n",
    "df2 = quilt_phase.loc[phased_indices,:]\n",
    "\n",
    "matched_haps = df1['final_step_phase1'] == df2['phase1_final']\n",
    "percet_matched_haps = (matched_haps.sum() / len(df1)) * 100\n",
    "print(f'Percentage agreed with QUILT phasing for {gene}: ', percet_matched_haps, '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For GAMCC\n",
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "\n",
    "subset_vcf_samples = lcwgsus.read_tsv_as_lst('/well/band/users/rbx225/GAMCC/data/sample_tsvs/fv_gm_names.tsv')\n",
    "subset_vcf_samples = ','.join(subset_vcf_samples)\n",
    "\n",
    "sample_linker = pd.read_csv('/well/band/users/rbx225/GAMCC/' + SAMPLE_LINKER_FILE)\n",
    "sample_linker = {k:v for k, v in zip(sample_linker['Sample_Name'], sample_linker['Chip_Name'])}\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information, \n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            read_from_QUILT = True, \n",
    "                            subset_vcf_samples = subset_vcf_samples,\n",
    "                            sample_linker = sample_linker)\n",
    "\n",
    "hlatypes = return_dict['hlatypes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "unphased_samples_to_remove = hlatypes.loc[indices, 'Sample ID'].tolist()\n",
    "samples_to_remove = samples_to_remove + unphased_samples_to_remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'phase_trios' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m trios[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallele2\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mN/A\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     10\u001b[0m undetermined_trios \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m---> 12\u001b[0m trios \u001b[38;5;241m=\u001b[39m trios\u001b[38;5;241m.\u001b[39mapply(func \u001b[38;5;241m=\u001b[39m \u001b[43mphase_trios\u001b[49m, args \u001b[38;5;241m=\u001b[39m (gene, hlatypes), axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m     14\u001b[0m kgp_samples_idx \u001b[38;5;241m=\u001b[39m hlatypes\u001b[38;5;241m.\u001b[39mindex[hlatypes[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSample ID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(trios[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIndividual ID\u001b[39m\u001b[38;5;124m'\u001b[39m])]\u001b[38;5;241m.\u001b[39mto_numpy()\n\u001b[1;32m     15\u001b[0m kgp_samples_phase \u001b[38;5;241m=\u001b[39m phased1[kgp_samples_idx]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'phase_trios' is not defined"
     ]
    }
   ],
   "source": [
    "# QUILT phasing is in the order of paternal-maternal\n",
    "retained_samples = hlatypes['Sample ID'].tolist()\n",
    "trios = pd.read_csv('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/integrated_call_samples_v3.20200731.ALL.ped', sep = '\\t')\n",
    "trios = trios[trios['Individual ID'].isin(retained_samples)]\n",
    "trios = trios[(trios['Paternal ID'].isin(retained_samples)) | (trios['Maternal ID'].isin(retained_samples))]\n",
    "trios = trios[['Individual ID', 'Paternal ID', 'Maternal ID']]\n",
    "\n",
    "trios['allele1'] = 'N/A'\n",
    "trios['allele2'] = 'N/A'\n",
    "undetermined_trios = []\n",
    "\n",
    "trios = trios.apply(func = phase_trios, args = (gene, hlatypes), axis = 1)\n",
    "\n",
    "kgp_samples_idx = hlatypes.index[hlatypes['Sample ID'].isin(trios['Individual ID'])].to_numpy()\n",
    "kgp_samples_phase = phased1[kgp_samples_idx]\n",
    "\n",
    "phased_res = pd.DataFrame({'Sample ID': trios['Individual ID'].tolist(), 'allele1': reftypes1[kgp_samples_idx], 'allele2': reftypes2[kgp_samples_idx]})\n",
    "\n",
    "for i in range(len(phased_res)):\n",
    "    if kgp_samples_phase[i] == False:\n",
    "        tmp = phased_res.iloc[i, 1]\n",
    "        phased_res.iloc[i, 1] = phased_res.iloc[i, 2]\n",
    "        phased_res.iloc[i, 2] = tmp\n",
    "\n",
    "res = pd.merge(phased_res, trios[['Individual ID', 'allele1', 'allele2']], left_on = 'Sample ID', right_on = 'Individual ID', suffixes = ('_pred', '_true'))\n",
    "res = res.drop(columns = 'Individual ID')\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['G-14:54-T', 'T-09:01-G', 'G-08:03-T', 'G-14:54-G', 'G-14:04-T', 'T-09:01-G']\n",
      "['G-11:04-T', 'A-03:01-G', 'G-11:04-T', 'A-12:01-G', 'A-03:01-G', 'A-14:54-T']\n",
      "['C-12:02-C', 'T-12:02-A', 'C-12:02-C', 'T-10:01-C', 'C-12:02-C', 'T-08:03-A']\n",
      "['G-03:01-T', 'T-01:01-A', 'G-03:01-T', 'G-15:01-A', 'T-01:01-A', 'T-01:01-A']\n",
      "['G-12:01-T', 'A-16:02-A', 'A-13:01-T', 'G-12:01-T', 'A-16:02-A', 'G-11:01-T']\n",
      "['C-07:01-G', 'G-03:01-A', 'C-07:01-G', 'G-15:01-A', 'G-16:02-A', 'G-03:01-A']\n",
      "['T-01:02-T', 'G-14:06-G', 'G-01:02-T', 'T-04:11-T', 'G-04:07-T', 'G-14:06-G']\n"
     ]
    }
   ],
   "source": [
    "full_start = return_dict['b38_start']\n",
    "full_end = return_dict['b38_end']\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "# full_start = 32568456\n",
    "# full_end = 32589742\n",
    "\n",
    "retained_samples = hlatypes['Sample ID'].tolist()\n",
    "trios = pd.read_csv('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/integrated_call_samples_v3.20200731.ALL.ped', sep = '\\t')\n",
    "trios = trios[trios['Individual ID'].isin(retained_samples)]\n",
    "trios = trios[(trios['Paternal ID'].isin(retained_samples)) & (trios['Maternal ID'].isin(retained_samples))]\n",
    "trios = trios[['Individual ID', 'Paternal ID', 'Maternal ID']]\n",
    "\n",
    "trio_samples_all = trios.values.ravel().tolist()\n",
    "\n",
    "df = read_vcf(\n",
    "        full_start - 5000, \n",
    "        full_end + 5000, \n",
    "        phased_vcf, \n",
    "        hlatypes, \n",
    "        False, \n",
    "        ','.join(trio_samples_all), \n",
    "        None)\n",
    "\n",
    "df = df[(df['ref'].str.len() == 1) & (df['alt'].str.len() == 1)]\n",
    "\n",
    "df_up = df[df['pos'] < full_start]\n",
    "df_down = df[df['pos'] > full_end]\n",
    "\n",
    "phase_df = return_dict['phase_df']\n",
    "phase_df = phase_df[phase_df['Sample'].isin(trio_samples_all)]\n",
    "\n",
    "trio_haps = {}\n",
    "\n",
    "for child in trios['Individual ID']:\n",
    "    mat_id, pat_id = trios[trios['Individual ID'] == child][['Maternal ID', 'Paternal ID']].values[0]\n",
    "    \n",
    "    child_alleles = phase_df[phase_df['Sample'] == child][['allele1', 'allele2']].values[0]\n",
    "    mat_alleles = phase_df[phase_df['Sample'] == mat_id][['allele1', 'allele2']].values[0]\n",
    "    pat_alleles = phase_df[phase_df['Sample'] == pat_id][['allele1', 'allele2']].values[0]\n",
    "    \n",
    "    df_up_child = df_up[['pos', 'ref', 'alt'] + [child, mat_id, pat_id]]\n",
    "    df_up_child = df_up_child[~df_up_child[child].str.contains(r'\\.')]\n",
    "    df_up_child['pat'] = df_up_child[child].str.split('|').str.get(0)\n",
    "    df_up_child['mat'] = df_up_child[child].str.split('|').str.get(1)\n",
    "    het_sites = (df_up_child['mat'] != df_up_child['pat'])\n",
    "    \n",
    "    upstream_idx = het_sites[het_sites].index[-1]\n",
    "    ref, alt, mat, pat = df_up_child.loc[upstream_idx, ['ref', 'alt', 'mat', 'pat']]\n",
    "    \n",
    "    mat_genotypes = df_up_child.loc[upstream_idx, mat_id].split('|')\n",
    "    pat_genotypes = df_up_child.loc[upstream_idx, pat_id].split('|')\n",
    "    \n",
    "    if mat_genotypes[0] == '0':\n",
    "        mat_hap1 = ref + '-' + mat_alleles[0] + '-'\n",
    "    elif mat_genotypes[0] == '1':\n",
    "        mat_hap1 = alt + '-' + mat_alleles[0] + '-'\n",
    "    else:\n",
    "        mat_hap1 = 'N' + '-' + mat_alleles[0] + '-'\n",
    "        \n",
    "    if mat_genotypes[1] == '0':\n",
    "        mat_hap2 = ref + '-' + mat_alleles[1] + '-'\n",
    "    elif mat_genotypes[1] == '1':\n",
    "        mat_hap2 = alt + '-' + mat_alleles[1] + '-'\n",
    "    else:\n",
    "        mat_hap2 = 'N' + '-' + mat_alleles[1] + '-'\n",
    "    \n",
    "    if pat_genotypes[0] == '0':\n",
    "        pat_hap1 = ref + '-' + pat_alleles[0] + '-'\n",
    "    elif pat_genotypes[0] == '1':\n",
    "        pat_hap1 = alt + '-' + pat_alleles[0] + '-'\n",
    "    else:\n",
    "        pat_hap1 = 'N' + '-' + pat_alleles[0] + '-'\n",
    "        \n",
    "    if pat_genotypes[1] == '0':\n",
    "        pat_hap2 = ref + '-' + pat_alleles[1] + '-'\n",
    "    elif pat_genotypes[1] == '1':\n",
    "        pat_hap2 = alt + '-' + pat_alleles[1] + '-'\n",
    "    else:\n",
    "        pat_hap2 = 'N' + '-' + pat_alleles[1] + '-'\n",
    "    \n",
    "    if pat == '0':\n",
    "        child_hap1 = ref + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = alt + '-' + child_alleles[1] + '-'\n",
    "    elif pat == '1':\n",
    "        child_hap1 = alt + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = ref + '-' + child_alleles[1] + '-'\n",
    "    else:\n",
    "        child_hap1 = 'N' + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = 'N' + '-' + child_alleles[1] + '-'\n",
    "    \n",
    "    df_down_child = df_down[['pos', 'ref', 'alt'] + [child, mat_id, pat_id]]\n",
    "    df_down_child = df_down_child[~df_down_child[child].str.contains(r'\\.')]\n",
    "    df_down_child['pat'] = df_down_child[child].str.split('|').str.get(0)\n",
    "    df_down_child['mat'] = df_down_child[child].str.split('|').str.get(1)\n",
    "    het_sites = (df_down_child['mat'] != df_down_child['pat'])\n",
    "    downstream_idx = het_sites[het_sites].index[0]\n",
    "    ref, alt, mat, pat = df_down_child.loc[downstream_idx, ['ref', 'alt', 'mat', 'pat']]\n",
    "    \n",
    "    mat_genotypes = df_down_child.loc[downstream_idx, mat_id].split('|')\n",
    "    pat_genotypes = df_down_child.loc[downstream_idx, pat_id].split('|')\n",
    "    \n",
    "    if mat_genotypes[0] == '0':\n",
    "        mat_hap1 = mat_hap1 + ref\n",
    "    elif mat_genotypes[0] == '1':\n",
    "        mat_hap1 = mat_hap1 + alt\n",
    "    else:\n",
    "        mat_hap1 = mat_hap1 + 'N'\n",
    "        \n",
    "    if mat_genotypes[1] == '0':\n",
    "        mat_hap2 = mat_hap2 + ref\n",
    "    elif mat_genotypes[1] == '1':\n",
    "        mat_hap2 = mat_hap2 + alt\n",
    "    else:\n",
    "        mat_hap2 = mat_hap2 + 'N'\n",
    "    \n",
    "    if pat_genotypes[0] == '0':\n",
    "        pat_hap1 = pat_hap1 + ref\n",
    "    elif pat_genotypes[0] == '1':\n",
    "        pat_hap1 = pat_hap1 + alt\n",
    "    else:\n",
    "        pat_hap1 = pat_hap1 + 'N'\n",
    "        \n",
    "    if pat_genotypes[1] == '0':\n",
    "        pat_hap2 = pat_hap2 + ref\n",
    "    elif pat_genotypes[1] == '1':\n",
    "        pat_hap2 = pat_hap2 + alt\n",
    "    else:\n",
    "        pat_hap2 = pat_hap2 + 'N'\n",
    "    \n",
    "    if pat == '0':\n",
    "        child_hap1 = child_hap1 + ref\n",
    "        child_hap2 = child_hap2 + alt\n",
    "    elif pat == '1':\n",
    "        child_hap1 = child_hap1 + alt\n",
    "        child_hap2 = child_hap2 + ref\n",
    "    else:\n",
    "        break\n",
    "    \n",
    "    trio_haps[child] = [child_hap1, child_hap2, pat_hap1, pat_hap2, mat_hap1, mat_hap2]\n",
    "    \n",
    "for v in trio_haps.values():\n",
    "    print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RETAIN THIS!\n",
    "# gamcc_hla = pd.read_csv('/well/band/users/rbx225/GAMCC/data/hla_direct_sequencing/HLA_direct_sequencing_all.csv')\n",
    "# gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Included Alleles']]\n",
    "# for s in gamcc_hla['SampleID'].unique():\n",
    "#     tmps = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "#     for l in HLA_GENES:\n",
    "#         tmpl = tmps[tmps['Locus'] == l]\n",
    "#         repeat = 2 - tmpl.shape[0]\n",
    "#         if repeat == 2:\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, np.nan]\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, np.nan]\n",
    "#         if repeat == 1:\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, tmpl.iloc[0,2]]\n",
    "                \n",
    "# gamcc_hla = gamcc_hla[gamcc_hla['Locus'] == gene]\n",
    "# gamcc_hla = gamcc_hla[['SampleID', 'Included Alleles']]\n",
    "\n",
    "# gamcc_hla = pd.merge(gamcc_hla[::2], gamcc_hla[1::2], on = 'SampleID')\n",
    "# gamcc_hla.columns = ['Sample ID', 'HLA-DRB1 1', 'HLA-DRB1 2']\n",
    "# gamcc_hla\n",
    "\n",
    "# x = gamcc_hla.iloc[0, 1]\n",
    "# indices = []\n",
    "# db_alleles = db.index.to_numpy()\n",
    "\n",
    "# for a in x.split('/'):\n",
    "#     old_a = a\n",
    "#     while (gene + '*' + a) not in db.index and ':' in a:\n",
    "#         a = a.rsplit(':', 1)[0]\n",
    "#     if ':' in a:\n",
    "#         indices.append(np.where(db_alleles == gene + '*' + a)[0][0])\n",
    "\n",
    "# x = db.iloc[np.array(indices), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>NA19685</th>\n",
       "      <th>NA19660</th>\n",
       "      <th>NA19661</th>\n",
       "      <th>mat</th>\n",
       "      <th>pat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32566587</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32566604</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32566652</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32566682</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32566687</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>32566689</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>32566695</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>32566705</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>32566707</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>32566720</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>32566739</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>32567918</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>32568000</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>32568008</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>32568020</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>32568031</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>32568040</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>32568051</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>32568054</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>32568055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>32568057</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>32568070</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>32568080</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>32568088</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>32568097</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>32568104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>32568112</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>32568117</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32568120</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>32568137</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>32568138</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>32568150</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>32568157</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>32568158</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>32568165</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>32568174</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>32568174</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>32568240</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>32568343</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>32568435</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>32568443</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>32568448</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0        pos ref alt NA19685 NA19660 NA19661 mat pat\n",
       "0   32566587   A   T     0|0     0|0     0|0   0   0\n",
       "1   32566604   C   G     0|0     0|0     0|0   0   0\n",
       "2   32566652   T   C     0|0     0|0     0|0   0   0\n",
       "3   32566682   G   A     1|0     0|0     0|1   1   0\n",
       "4   32566687   C   T     1|0     0|0     0|1   1   0\n",
       "5   32566689   T   C     0|0     0|0     0|0   0   0\n",
       "6   32566695   A   G     0|0     0|0     0|0   0   0\n",
       "7   32566705   C   T     1|0     0|0     0|1   1   0\n",
       "8   32566707   G   A     1|0     0|0     0|1   1   0\n",
       "9   32566720   C   T     0|0     0|0     0|0   0   0\n",
       "10  32566739   T   C     0|0     0|0     0|0   0   0\n",
       "11  32567918   C   T     0|0     0|0     0|0   0   0\n",
       "14  32568000   C   T     0|0     0|0     0|0   0   0\n",
       "15  32568008   A   G     0|0     0|0     0|0   0   0\n",
       "17  32568020   G   T     0|0     0|0     0|0   0   0\n",
       "19  32568031   G   A     0|0     0|0     0|0   0   0\n",
       "20  32568040   C   T     0|0     0|0     0|0   0   0\n",
       "21  32568051   C   A     0|0     0|0     0|0   0   0\n",
       "22  32568054   T   C     0|0     0|0     0|0   0   0\n",
       "23  32568055   G   A     0|0     0|0     0|0   0   0\n",
       "24  32568057   G   A     0|0     0|0     0|0   0   0\n",
       "25  32568070   G   A     0|0     0|0     0|0   0   0\n",
       "26  32568080   G   C     0|0     0|0     0|0   0   0\n",
       "27  32568088   C   T     0|0     0|0     0|0   0   0\n",
       "28  32568097   A   G     0|0     0|0     0|0   0   0\n",
       "29  32568104   A   G     1|0     0|0     0|1   1   0\n",
       "30  32568112   G   T     0|0     0|0     0|0   0   0\n",
       "31  32568117   G   T     1|0     0|0     0|1   1   0\n",
       "32  32568120   T   C     0|0     0|0     1|0   0   0\n",
       "33  32568137   T   G     0|0     0|0     0|0   0   0\n",
       "34  32568138   G   C     0|0     0|0     0|0   0   0\n",
       "35  32568150   G   T     0|0     0|0     0|0   0   0\n",
       "36  32568157   C   G     0|0     0|0     0|0   0   0\n",
       "37  32568158   A   G     0|0     0|0     0|0   0   0\n",
       "38  32568165   G   A     0|0     0|0     0|0   0   0\n",
       "39  32568174   G   C     0|0     0|0     0|0   0   0\n",
       "40  32568174   G   T     0|0     0|0     0|0   0   0\n",
       "41  32568240   C   A     0|0     0|0     0|0   0   0\n",
       "44  32568343   T   C     0|0     0|0     0|0   0   0\n",
       "45  32568435   A   C     0|0     0|0     0|0   0   0\n",
       "46  32568443   T   A     0|0     0|0     0|0   0   0\n",
       "47  32568448   C   G     0|0     0|0     0|0   0   0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_up_child"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
