{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import pickle\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "# import sqlite3\n",
    "# from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "# from plotnine import *\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "# from collections import Counter\n",
    "# import seaborn as sns\n",
    "# import matplotlib.colors as mcolors\n",
    "# from matplotlib.ticker import FuncFormatter\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "import psa\n",
    "import pywfa\n",
    "from IPython.display import display_html, IFrame\n",
    "# import patchworklib as pw\n",
    "sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "from hla_phase import *\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# Genuine logics changed from the main script were marked by string **X** in the original R files.\n",
    "# Simplicity changes were marked by string **Y**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For oneKG\n",
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "ref_samples = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/oneKG.samples', sep = ' ')\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "ref_samples_removed = ref_samples[~ref_samples['SAMPLE'].isin(hlatypes['Sample ID'].tolist())]\n",
    "samples_to_remove = ref_samples_removed['SAMPLE'].tolist()\n",
    "hlatypes = hlatypes[~hlatypes['Sample ID'].isin(samples_to_remove)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG_30x/oneKG.chr6.vcf.gz\"\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "\n",
    "# return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "#                             ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "#                             hla_gene_information = hla_gene_information,\n",
    "#                             hlatypes = hlatypes,\n",
    "#                             phased_vcf = phased_vcf, \n",
    "#                             reference_allele_ary = reference_allele_ary, \n",
    "#                             strict_snp_filter = True,\n",
    "#                             read_from_QUILT = False, \n",
    "#                             subset_vcf_samples = None,\n",
    "#                             sample_linker = None)\n",
    "# hlatypes = return_dict['hlatypes']\n",
    "\n",
    "# individual = 'NA12878'\n",
    "# ix = hlatypes.index[hlatypes['Sample ID'] == individual][0]\n",
    "# display_indices = np.arange(10)\n",
    "# html = '1KG-DRB1.html'\n",
    "\n",
    "# res = visualise_phase(gene, ix, hlatypes, return_dict, both_het = True)\n",
    "# compare_phase(display_indices, res, save_html = True, save_name = html)\n",
    "\n",
    "# plot_first_step_phase(c)\n",
    "# plot_last_step_phase(return_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# strand = hla_gene_information[hla_gene_information['Name'] == ('HLA-' + gene)]['Strand'].iloc[0]\n",
    "# reference_allele = reference_allele_ary[np.char.find(reference_allele_ary, gene) != -1][0]\n",
    "# nucleotides = ['A', 'T', 'C', 'G']\n",
    "\n",
    "# ipd_gen_file = ipd_gen_file_dir + gene + '_gen.txt'\n",
    "# with open(ipd_gen_file, \"r\") as file:\n",
    "#     lines = file.readlines()\n",
    "# gDNA_idx = []\n",
    "# names = []\n",
    "# i = 0\n",
    "# while len(gDNA_idx) < 2:\n",
    "#     l = lines[i]\n",
    "#     if 'gDNA' in l:\n",
    "#         gDNA_idx.append(i)\n",
    "#     elif l.lstrip(' ').split(' ')[0].startswith(gene + '*'):\n",
    "#         name = l.lstrip(' ').split(' ')[0]\n",
    "#         names.append(name)\n",
    "#     i += 1\n",
    "\n",
    "# first_base = int(lines[gDNA_idx[0]].split(' ')[-1].split('\\n')[0])\n",
    "# n_alleles = gDNA_idx[1] - gDNA_idx[0] - 3\n",
    "\n",
    "# alleles_dict = {k:'' for k in names}\n",
    "# for i, s in enumerate(lines):\n",
    "#     r = s.lstrip(' ')\n",
    "#     if r.startswith(gene):\n",
    "#         r = r.rstrip(' \\n')\n",
    "#         name = r.split(' ')[0]\n",
    "#         sequence = r.split(' ')[1:]\n",
    "#         sequence = ''.join(sequence)\n",
    "#         alleles_dict[name] = alleles_dict[name] + sequence\n",
    "\n",
    "# db = pd.DataFrame({key: list(value) for key, value in alleles_dict.items()}).T\n",
    "# db = db.drop(columns=db.columns[db.eq('|').all()])\n",
    "# db.columns = range(db.shape[1])\n",
    "\n",
    "# length = len(db.columns)\n",
    "# positions = [first_base]*length\n",
    "\n",
    "# db.loc[len(db)] = positions\n",
    "\n",
    "# r_idx = len(db) - 1\n",
    "# for i in db.columns[1:]:\n",
    "#     if db.iloc[0, i] == '.':\n",
    "#         db.iloc[r_idx, i] = db.iloc[r_idx, i-1]\n",
    "#     else:\n",
    "#         if db.iloc[r_idx, i-1] != -1:\n",
    "#             db.iloc[r_idx, i] = db.iloc[r_idx, i-1] + 1\n",
    "#         else:\n",
    "#             db.iloc[r_idx, i] = db.iloc[r_idx, i-1] + 2 # No zero in position. ATG is encoded by 1.\n",
    "# db.columns = db.iloc[r_idx]\n",
    "# db = db.iloc[:r_idx]\n",
    "\n",
    "# db = db.loc[:,db.columns >= 0] # Trims genomes before CDS, is this really necessary?\n",
    "# db = db.apply(lambda c: c.replace('-', c[0]) ,axis = 0)\n",
    "\n",
    "# if strand != 1:\n",
    "#     db.replace({'A': 't', 'C': 'g', 'G': 'c', 'T': 'a'}, inplace=True)\n",
    "#     db.replace({'a': 'A', 'c': 'C', 'g': 'G', 't': 'T'}, inplace=True)\n",
    "#     db = db.iloc[:, ::-1]\n",
    "# db.columns = range(db.shape[1])\n",
    "\n",
    "# counts = db.loc[reference_allele,:].value_counts()\n",
    "# aligned_bases = counts[counts.index.isin(nucleotides)].sum()\n",
    "\n",
    "# if strand == 1:\n",
    "#     genome_pos =  hla_gene_information[hla_gene_information['Name'] == ('HLA-' + gene)]['Start'].iloc[0]\n",
    "#     start = genome_pos\n",
    "# else:\n",
    "#     genome_pos =  hla_gene_information[hla_gene_information['Name'] == ('HLA-' + gene)]['End'].iloc[0]\n",
    "#     start = genome_pos - aligned_bases + 1\n",
    "# end = start + aligned_bases - 1\n",
    "\n",
    "# full_start = start\n",
    "# full_end = end\n",
    "\n",
    "# allele_pos = [start]*db.shape[1]\n",
    "# for i, n in enumerate(db.loc[reference_allele,:].values[1:], start = 1):\n",
    "#     if n in nucleotides:\n",
    "#         allele_pos[i] = allele_pos[i - 1] + 1\n",
    "#     else:\n",
    "#         allele_pos[i] = allele_pos[i - 1]\n",
    "\n",
    "# # ourpos -> allele_pos\n",
    "# # first_row = ourrow is the index in db of the referece allele\n",
    "\n",
    "# variants_df = []\n",
    "# for i in range(len(db)):\n",
    "#     variants_df.append(extract_snps_from_row(db.loc[reference_allele,:], db.iloc[i,], allele_pos))\n",
    "# vcf = pd.concat(variants_df).drop_duplicates()\n",
    "# vcf = vcf[(vcf['ref'].isin(nucleotides)) & (vcf['alt'].isin(nucleotides))].sort_values(by = 'pos').reset_index(drop = True)\n",
    "\n",
    "# allele_pos = np.array(allele_pos)\n",
    "# dbcopy = pd.DataFrame()\n",
    "# aligned_df = pd.DataFrame()\n",
    "# variants_to_remove = []\n",
    "\n",
    "# for i in range(vcf.shape[0]):\n",
    "#     pos, a1, a2 = vcf.iloc[i, :3]\n",
    "#     index = np.where((allele_pos == pos))[0]\n",
    "\n",
    "#     if len(index) == 1:\n",
    "#         index = index[0]\n",
    "#         dbcopy[dbcopy.shape[1]] = db[index]        \n",
    "#         aligned_df[aligned_df.shape[1]] = db[index].apply(recode_variant, args = (a2))\n",
    "#     else:\n",
    "#         variants_to_remove.append(i)\n",
    "\n",
    "# vcf = vcf[~vcf.index.isin(variants_to_remove)].reset_index(drop = True) \n",
    "# # This is a remedy when extract_snps_from_row fails to detect insertions. \n",
    "# # For DRB1 we get 2442 variants instead of 2452, fairly good I'd say\n",
    "# dbfull = db\n",
    "# db = dbcopy\n",
    "\n",
    "# sample = pd.DataFrame({\n",
    "#     \"ID_1\": aligned_df.index,\n",
    "#     \"ID_2\": aligned_df.index,\n",
    "#     \"missing\": 0\n",
    "# })\n",
    "# vcf['id'] = 'chr6:' + vcf['pos'].astype(str)\n",
    "# haps = aligned_df.T\n",
    "# cond = (haps == -1).mean(axis=1) < 0.1\n",
    "# haps = haps[cond].reset_index(drop = True)\n",
    "# vcf = vcf[cond].reset_index(drop = True)\n",
    "\n",
    "# start = vcf.loc[0, 'pos']\n",
    "# end =vcf.loc[len(vcf) - 1, 'pos']\n",
    "\n",
    "# cond = (allele_pos >= start) & (allele_pos <= end)\n",
    "# allele_pos = allele_pos[cond]\n",
    "# dbfull = dbfull.iloc[:, np.where(cond == True)[0]]\n",
    "# dbfull.columns = range(dbfull.shape[1])\n",
    "\n",
    "# # Imputing database alleles\n",
    "# for i, a in enumerate(haps.columns):\n",
    "#     haps_undetermined = haps[a] == -1\n",
    "\n",
    "#     allele_col = haps[~haps_undetermined][a]\n",
    "#     subseted_db =  haps[~haps_undetermined] \n",
    "#     pairwise_distance = subseted_db.sub(subseted_db[a], axis=0)\n",
    "#     counts = pairwise_distance.abs().sum().sort_values().iloc[1:]\n",
    "#     closest_allele_idx = 0\n",
    "#     substituted_nucleotides = np.array([np.nan]*np.sum(haps_undetermined))\n",
    "#     substituted_indices = haps_undetermined[haps_undetermined == True].index\n",
    "\n",
    "#     while np.isnan(substituted_nucleotides).any() > 0:\n",
    "#         closest_allele = counts.index[closest_allele_idx]\n",
    "\n",
    "#         target_haplotype = haps.loc[substituted_indices, closest_allele]\n",
    "#         substitutable = target_haplotype[target_haplotype.values != -1]\n",
    "#         substituted_nucleotides[np.where(np.in1d(substituted_indices, substitutable.index))[0]] = substitutable.values\n",
    "#         closest_allele_idx += 1\n",
    "\n",
    "#     haps.loc[substituted_indices, a] = substituted_nucleotides\n",
    "#     # haps has dim n_variants * n_alleles\n",
    "\n",
    "# prepared_vcf = vcf\n",
    "# alleles = db.index\n",
    "# prepared_vcf['pos']  = prepared_vcf['pos'].astype(int)\n",
    "\n",
    "# variant_alleles = haps\n",
    "\n",
    "# positions = vcf['pos'].astype(float).values\n",
    "# zz = np.zeros(haps.shape[0]) \n",
    "# for i in range(len(zz)):\n",
    "#     zz[i] = np.sum(allele_pos == positions[i])\n",
    "# zz3 = np.zeros_like(zz)\n",
    "# for i in range(len(zz3)):\n",
    "#     zz3[i] = np.sum(positions == positions[i])\n",
    "# filter_conditions = (zz3 == 1) & (zz == 1)\n",
    "\n",
    "# snpinfo = prepared_vcf.loc[filter_conditions, :].reset_index(drop = True)\n",
    "# snpinfo = snpinfo.iloc[:,[3, 0, 1, 2]]\n",
    "# variant_alleles = variant_alleles.loc[filter_conditions, :]\n",
    "# haps = variant_alleles.copy().reset_index(drop = True)\n",
    "# # haps is `n_variant \\times n_alleles` df, with each entry as 0/1 representing ref/alt from the fake vcf \"prepared_vcf\"\n",
    "\n",
    "# start = snpinfo['pos'].values[0]\n",
    "# end = snpinfo['pos'].values[-1]\n",
    "\n",
    "# snp_haps = read_vcf(\n",
    "#     full_start - 1000, \n",
    "#     full_end + 1000, \n",
    "#     phased_vcf, \n",
    "#     hlatypes)\n",
    "\n",
    "# retained_samples = snp_haps.columns[4:].tolist()\n",
    "# hlatypes = hlatypes[hlatypes['Sample ID'].isin(retained_samples)].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>370</th>\n",
       "      <th>371</th>\n",
       "      <th>372</th>\n",
       "      <th>373</th>\n",
       "      <th>374</th>\n",
       "      <th>375</th>\n",
       "      <th>376</th>\n",
       "      <th>377</th>\n",
       "      <th>378</th>\n",
       "      <th>379</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2564</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2565</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2566</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2567</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2568 rows × 380 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0    1    2    3    4    5    6    7    8    9    ...  370  371  372  \\\n",
       "0       0    0    0    0    0    0    0    0    0    0  ...    0    1    1   \n",
       "1       0    0    0    0    0    0    0    0    0    0  ...    0    0    0   \n",
       "2       0    0    0    0    0    0    0    0    0    0  ...    0    1    1   \n",
       "3       0    0    0    0    0    0    0    0    0    0  ...    0    1    1   \n",
       "4       0    0    0    0    0    0    0    0    0    0  ...    0    0    0   \n",
       "...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "2563    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   \n",
       "2564    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   \n",
       "2565    0    0    0    0    0    0    0    0    0    0  ...    0    1    1   \n",
       "2566    0    0    0    0    0    0    0    0    0    0  ...    0    1    1   \n",
       "2567    0    0    0    0    0    0    0    0    0    0  ...    0    1    1   \n",
       "\n",
       "      373  374  375  376  377  378  379  \n",
       "0       0    0    0    0    0    0    0  \n",
       "1       0    0    0    0    0    0    0  \n",
       "2       0    0    0    0    0    0    0  \n",
       "3       0    0    0    0    0    0    0  \n",
       "4       0    0    0    0    0    0    0  \n",
       "...   ...  ...  ...  ...  ...  ...  ...  \n",
       "2563    0    0    0    0    0    0    0  \n",
       "2564    0    0    0    0    0    0    0  \n",
       "2565    0    0    0    0    0    0    0  \n",
       "2566    0    0    0    0    0    0    0  \n",
       "2567    0    0    0    0    0    0    0  \n",
       "\n",
       "[2568 rows x 380 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hrcfirstalleles = snp_haps[['snp', 'pos', 'ref', 'alt']]\n",
    "hrcsecondalleles = snp_haps[['snp', 'pos', 'ref', 'alt']]\n",
    "for s in retained_samples:\n",
    "    hrcfirstalleles[s] = snp_haps[s].str.split('|').str.get(0).astype(int)\n",
    "    hrcsecondalleles[s] = snp_haps[s].str.split('|').str.get(1).astype(int)\n",
    "\n",
    "hrcfirstalleles_tmp = hrcfirstalleles[(hrcfirstalleles['pos'] >= start) & (hrcfirstalleles['pos'] <= end)]\n",
    "hrcsecondalleles_tmp = hrcsecondalleles[(hrcsecondalleles['pos'] >= start) & (hrcsecondalleles['pos'] <= end)]\n",
    "hrcfirstalleles_full = hrcfirstalleles.copy()\n",
    "hrcsecondalleles_full = hrcsecondalleles.copy()\n",
    "hrcfirstalleles = hrcfirstalleles_tmp\n",
    "hrcsecondalleles = hrcsecondalleles_tmp    \n",
    "\n",
    "original_variants = (snpinfo['pos'].astype(str) + snpinfo['ref'] + snpinfo['alt']).values\n",
    "flipped_variants = (snpinfo['pos'].astype(str) + snpinfo['alt'] + snpinfo['ref']).values\n",
    "\n",
    "def subset_snp_haps(hrcfirstalleles, original_variants = original_variants, flipped_variants = flipped_variants):    \n",
    "    hrcfirstalleles_original = hrcfirstalleles[hrcfirstalleles['snp'].isin(original_variants)]\n",
    "    hrcfirstalleles_flipped = hrcfirstalleles[hrcfirstalleles['snp'].isin(flipped_variants)]\n",
    "    if len(hrcfirstalleles_flipped) == 0:\n",
    "        hrcfirstalleles = hrcfirstalleles_original.drop(columns = ['snp', 'pos', 'ref', 'alt']).reset_index(drop = True)\n",
    "    elif len(hrcfirstalleles_flipped) == 1:\n",
    "        hrcfirstalleles_flipped.loc[:, retained_samples] = 1 - hrcfirstalleles_flipped.loc[:, retained_samples]\n",
    "        hrcfirstalleles = pd.concat([hrcfirstalleles_original, hrcfirstalleles_flipped.to_frame().T], axis = 0).sort_index()\n",
    "    else:\n",
    "        hrcfirstalleles_flipped.loc[:, retained_samples] = 1 - hrcfirstalleles_flipped.loc[:, retained_samples]\n",
    "        hrcfirstalleles = pd.concat([hrcfirstalleles_original, hrcfirstalleles_flipped]).sort_index()\n",
    "    hrcfirstalleles = hrcfirstalleles.sort_index().reset_index(drop = True)\n",
    "    hrcfirstalleles.columns = range(hrcfirstalleles.shape[1])\n",
    "    return hrcfirstalleles.T\n",
    "\n",
    "hrcfirstalleles = subset_snp_haps(hrcfirstalleles)\n",
    "hrcsecondalleles = subset_snp_haps(hrcsecondalleles)\n",
    "hrcfirstalleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 57.06666667,  77.        , 118.        ,  58.        ,\n",
       "        50.81818182,  55.        ,  61.02564103,  61.        ,\n",
       "        60.18181818,  56.8       ,  54.18181818,  52.        ])"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ixs = [29,598,600,721,1027,1168,1506,1677,1853,1880,2072,2516]\n",
    "ix = 2072\n",
    "name = 'NA19077'\n",
    "oldphase1[ixs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([50.93333333, 57.        , 62.        , 52.        , 73.18181818,\n",
       "       69.        , 59.02564103, 53.        , 58.18181818, 65.2       ,\n",
       "       54.18181818, 64.        ])"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oldphase2[ixs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# valid_corr_indices = np.where(~np.isnan(corr))[0]\n",
    "valid_corr_indices = np.where((~np.isnan(corr)) & (corr > 0.8))[0]\n",
    "\n",
    "dist11 = phase_subtraction(hrcfirstalleles, predfirstalleles, valid_corr_indices)\n",
    "dist12 = phase_subtraction(hrcfirstalleles, predsecondalleles, valid_corr_indices)\n",
    "dist21 = phase_subtraction(hrcsecondalleles, predfirstalleles, valid_corr_indices)\n",
    "dist22 = phase_subtraction(hrcsecondalleles, predsecondalleles, valid_corr_indices)\n",
    "\n",
    "phase1 = dist11 + dist22\n",
    "phase2 = dist12 + dist21\n",
    "\n",
    "d11 = dist11\n",
    "d21 = dist21\n",
    "d12 = dist12\n",
    "d22 = dist22\n",
    "\n",
    "phased1 = (\n",
    "    (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 < 4) & (phase2 > 4)) |\n",
    "    (~np.isnan(phase1) & ~np.isnan(phase2) & (phase2 - phase1 > 2) & (phase1 < 4)) |\n",
    "    ((reftypes1 != 'nan') & (reftypes2 != 'nan') & (reftypes1 == reftypes2)) |\n",
    "    (np.isnan(d21) & ~np.isnan(d12) & (d12 - d22 > 2) & (d22 < 2)) |\n",
    "    (np.isnan(d12) & ~np.isnan(d21) & (d21 - d11 > 2) & (d11 < 2))\n",
    ")\n",
    "\n",
    "phased2 = (\n",
    "    (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 > 4) & (phase2 < 4)) |\n",
    "    (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 - phase2 > 2) & (phase2 < 4)) |\n",
    "    (np.isnan(d21) & ~np.isnan(d12) & (d22 - d12 > 2) & (d12 < 2)) |\n",
    "    (np.isnan(d12) & ~np.isnan(d21) & (d11 - d21 > 2) & (d21 < 2))\n",
    ")\n",
    "\n",
    "phased1old = phased1.copy()\n",
    "phased2old = phased2.copy()\n",
    "oldphase1 = phase1.copy()\n",
    "oldphase2 = phase2.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "h1 = hrcfirstalleles.loc[ix,].values\n",
    "h2 = hrcsecondalleles.loc[ix,].values\n",
    "p1 = predfirstalleles[ix,]\n",
    "p2 = predsecondalleles[ix,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0,   1,   2,   3,   4,   7,   9,  12,  15,  17,  19,  23,  24,\n",
       "        58,  60,  62,  65,  66,  67,  70,  81,  91,  92,  94, 102, 117,\n",
       "       118, 119, 129])"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where((h1[valid_corr_indices]-p1[valid_corr_indices]) != 0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  7,   9,  12,  15,  17,  19,  37,  38,  58,  60,  62,  65,  66,\n",
       "        67,  70,  76,  77,  79,  81,  91,  92,  94, 102, 105, 127, 129])"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where((h2[valid_corr_indices]-p2[valid_corr_indices]) != 0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 25,  26,  27,  30,  31,  34,  36,  40,  42,  43,  44,  45,  47,\n",
       "        48,  49,  50,  51,  52,  53,  76,  77,  79, 127])"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where((h1[valid_corr_indices]-p2[valid_corr_indices]) != 0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0,   1,   2,   3,   4,  23,  24,  25,  26,  27,  30,  31,  34,\n",
       "        36,  37,  38,  40,  42,  43,  44,  45,  47,  48,  49,  50,  51,\n",
       "        52,  53, 105, 117, 118, 119])"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where((h2[valid_corr_indices]-p1[valid_corr_indices]) != 0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>4</th>\n",
       "      <th>122</th>\n",
       "      <th>133</th>\n",
       "      <th>235</th>\n",
       "      <th>245</th>\n",
       "      <th>2018</th>\n",
       "      <th>2030</th>\n",
       "      <th>11137</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:01</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:02</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:03</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:04</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:05</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:06</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:07</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:08</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:09</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:10</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*09:01:02:11</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*10:01:01:01</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*10:01:01:03</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*10:01:01:04</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*10:01:01:05</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRB1*10:01:02</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 4     122   133   235   245   2018  2030  11137\n",
       "DRB1*09:01:02:01     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:02     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:03     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:04     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:05     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:06     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:07     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:08     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:09     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:10     A     G     G     A     A     C     A     C\n",
       "DRB1*09:01:02:11     A     G     G     A     A     C     A     C\n",
       "DRB1*10:01:01:01     T     A     A     G     G     T     G     T\n",
       "DRB1*10:01:01:03     T     A     A     G     G     T     G     T\n",
       "DRB1*10:01:01:04     T     A     A     G     G     T     G     T\n",
       "DRB1*10:01:01:05     T     A     A     G     G     T     G     T\n",
       "DRB1*10:01:02        T     A     A     G     G     T     G     T"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "suspicious_idx = [0,   1,   2,   3,   4, 23,  24, 117]\n",
    "\n",
    "Xsidx = valid_corr_indices[suspicious_idx]\n",
    "Xs = np.array(Xsidx)\n",
    "Xspos = merged_snps.loc[Xs,:]['pos'].tolist()\n",
    "Xsposb38 = np.where(np.in1d(allele_pos, Xspos))[0]\n",
    "tmp = dbfull.loc[:, Xsposb38]\n",
    "tmp = tmp[tmp.index.str.startswith('DRB1*09:01') | tmp.index.str.startswith('DRB1*10:01')]\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32579059</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr6:32579059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32579175</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr6:32579175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32579186</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr6:32579186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>32579286</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr6:32579286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>32579296</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr6:32579296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>32581024</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr6:32581024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>32581036</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr6:32581036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>32586106</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr6:32586106</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          pos ref alt             id\n",
       "0    32579059   A   T  chr6:32579059\n",
       "3    32579175   G   A  chr6:32579175\n",
       "4    32579186   G   A  chr6:32579186\n",
       "5    32579286   A   G  chr6:32579286\n",
       "6    32579296   A   G  chr6:32579296\n",
       "43   32581024   C   T  chr6:32581024\n",
       "44   32581036   A   G  chr6:32581036\n",
       "282  32586106   T   C  chr6:32586106"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_snps.loc[Xs,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr6:32579059,chr6:32579175,chr6:32579186,chr6:32579286,chr6:32579296,chr6:32581024,chr6:32581036,chr6:32586106'"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "','.join([f'chr6:{i}' for i in allele_pos[Xsposb38]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "hrcfirstalleles = snp_haps[['snp', 'pos', 'ref', 'alt']]\n",
    "hrcsecondalleles = snp_haps[['snp', 'pos', 'ref', 'alt']]\n",
    "for s in retained_samples:\n",
    "    hrcfirstalleles[s] = snp_haps[s].str.split('|').str.get(0).astype(int)\n",
    "    hrcsecondalleles[s] = snp_haps[s].str.split('|').str.get(1).astype(int)\n",
    "\n",
    "hrcfirstalleles_tmp = hrcfirstalleles[(hrcfirstalleles['pos'] >= start) & (hrcfirstalleles['pos'] <= end)]\n",
    "hrcsecondalleles_tmp = hrcsecondalleles[(hrcsecondalleles['pos'] >= start) & (hrcsecondalleles['pos'] <= end)]\n",
    "hrcfirstalleles_full = hrcfirstalleles.copy()\n",
    "hrcsecondalleles_full = hrcsecondalleles.copy()\n",
    "hrcfirstalleles = hrcfirstalleles_tmp\n",
    "hrcsecondalleles = hrcsecondalleles_tmp    \n",
    "\n",
    "original_variants = (snpinfo['pos'].astype(str) + snpinfo['ref'] + snpinfo['alt']).values\n",
    "flipped_variants = (snpinfo['pos'].astype(str) + snpinfo['alt'] + snpinfo['ref']).values\n",
    "\n",
    "def subset_snp_haps(hrcfirstalleles, original_variants = original_variants, flipped_variants = flipped_variants):    \n",
    "    hrcfirstalleles_original = hrcfirstalleles[hrcfirstalleles['snp'].isin(original_variants)]\n",
    "    hrcfirstalleles_flipped = hrcfirstalleles[hrcfirstalleles['snp'].isin(flipped_variants)]\n",
    "    if len(hrcfirstalleles_flipped) == 0:\n",
    "        hrcfirstalleles = hrcfirstalleles_original.drop(columns = ['snp', 'pos', 'ref', 'alt']).reset_index(drop = True)\n",
    "    elif len(hrcfirstalleles_flipped) == 1:\n",
    "        hrcfirstalleles_flipped.loc[:, retained_samples] = 1 - hrcfirstalleles_flipped.loc[:, retained_samples]\n",
    "        hrcfirstalleles = pd.concat([hrcfirstalleles_original, hrcfirstalleles_flipped.to_frame().T], axis = 0).sort_index()\n",
    "    else:\n",
    "        hrcfirstalleles_flipped.loc[:, retained_samples] = 1 - hrcfirstalleles_flipped.loc[:, retained_samples]\n",
    "        hrcfirstalleles = pd.concat([hrcfirstalleles_original, hrcfirstalleles_flipped]).sort_index()\n",
    "    hrcfirstalleles = hrcfirstalleles.sort_index().reset_index(drop = True)\n",
    "    hrcfirstalleles.columns = range(hrcfirstalleles.shape[1])\n",
    "    return hrcfirstalleles.T\n",
    "\n",
    "hrcfirstalleles = subset_snp_haps(hrcfirstalleles)\n",
    "hrcsecondalleles = subset_snp_haps(hrcsecondalleles)\n",
    "\n",
    "samples = np.array(alleles).reshape(-1, 1)\n",
    "c1 = [s[0].split('*') for s in samples]\n",
    "vv = np.empty((samples.shape[0], 3), dtype=object)\n",
    "ww = [c[1] for c in c1]\n",
    "for i in range(len(c1)):\n",
    "    vv[i, 0] = c1[i][0]\n",
    "c1 = [w.split(':') for w in ww]\n",
    "for i in range(len(c1)):\n",
    "    vv[i, 1] = c1[i][0]\n",
    "    vv[i, 2] = c1[i][1]\n",
    "fourdigit = [f\"{vv[i, 1]}:{vv[i, 2]}\" for i in range(len(vv))]\n",
    "ufourdigit = np.unique(fourdigit)\n",
    "# Convert reference alleles to 4-digit resolution. vv and ww are just intermediate variables of no use\n",
    "# ufourdigit is unique 4-digit alleles\n",
    "\n",
    "haps = np.array(haps, dtype=float)\n",
    "newhaps = np.empty((haps.shape[0], len(ufourdigit)))\n",
    "for i in range(newhaps.shape[1]):\n",
    "    tt = np.where(np.array(fourdigit) == ufourdigit[i])[0]\n",
    "    if len(tt) < 2:\n",
    "        tt = np.concatenate((tt, tt))\n",
    "    newhaps[:, i] = np.mean(haps[:, tt], axis=1)\n",
    "newhaps = pd.DataFrame(newhaps, columns = ufourdigit)\n",
    "\n",
    "cols = snpinfo.columns.tolist()\n",
    "full_vcf = snp_haps.copy()[cols[1:]]\n",
    "same = pd.merge(full_vcf, snpinfo, on = cols[1:], how = 'inner')\n",
    "newhaps_original = newhaps.loc[snpinfo[snpinfo['id'].isin(same['id'].values)].index,:]\n",
    "\n",
    "snpinfo_flipped = snpinfo.iloc[:,[0, 1, 3, 2]]\n",
    "snpinfo_flipped.columns = cols\n",
    "flipped = pd.merge(full_vcf, snpinfo_flipped, on = cols[1:], how = 'inner')\n",
    "newhaps_flipped = 1 - newhaps.loc[snpinfo_flipped[snpinfo_flipped['id'].isin(flipped['id'].values)].index,:]\n",
    "\n",
    "merged_snps = pd.concat([same, flipped]).sort_values(by = 'pos', ascending = True).reset_index(drop = True)\n",
    "\n",
    "if len(newhaps_flipped) == 0:\n",
    "    newhaps = newhaps_original\n",
    "elif len(newhaps_flipped) == 1:\n",
    "    newhaps = pd.concat([newhaps_original, newhaps_flipped.to_frame().T], axis = 0).sort_index()\n",
    "else:\n",
    "    newhaps = pd.concat([newhaps_original, newhaps_flipped]).sort_index()\n",
    "newhaps2 = newhaps.reset_index(drop = True)\n",
    "\n",
    "region_pattern = f\"HLA.{gene}.\"\n",
    "cols = [i for i, col in enumerate(hlatypes.columns) if re.search(region_pattern, col)]\n",
    "\n",
    "reftypes1 = hlatypes.iloc[:, cols[0]].astype(str).tolist()\n",
    "reftypes1 = [re.sub(r\"\\*\", \"\", t) for t in reftypes1]\n",
    "reftypes1 = np.array(reftypes1)\n",
    "reftypes2 = hlatypes.iloc[:, cols[1]].astype(str).tolist()\n",
    "reftypes2 = [re.sub(r\"\\*\", \"\", t) for t in reftypes2]\n",
    "reftypes2 = np.array(reftypes2)\n",
    "\n",
    "def get_pred_alleles(reftype, newhaps2 = newhaps2):\n",
    "    predalleles = np.array(pd.DataFrame(np.nan, index=range(len(reftype)), columns=range(newhaps2.shape[0])))\n",
    "    for i, a in enumerate(reftype):\n",
    "        if (a == '-9') or (a == 'nan'):\n",
    "            pass\n",
    "        elif '/' not in a:\n",
    "            if a in newhaps2.columns:\n",
    "                predalleles[i, :] = newhaps2.loc[:, a].T\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            valid_alleles = []\n",
    "            for j in a.split('/'):\n",
    "                if ':' not in j:\n",
    "                    onefield = a.split('/')[0].split(':')[0]\n",
    "                    j = onefield + ':' + j\n",
    "                if j in newhaps2.columns:\n",
    "                    valid_alleles.append(j)\n",
    "            if len(valid_alleles) == 0:\n",
    "                pass\n",
    "            else:\n",
    "                predalleles[i, :] = newhaps2.loc[:, valid_alleles].mean(axis = 1).values\n",
    "    return predalleles\n",
    "\n",
    "predfirstalleles = get_pred_alleles(reftypes1)\n",
    "predsecondalleles = get_pred_alleles(reftypes2)\n",
    "\n",
    "obsgen = hrcfirstalleles + hrcsecondalleles\n",
    "predgen = pd.DataFrame(predfirstalleles + predsecondalleles)\n",
    "corr = np.empty(obsgen.shape[1])\n",
    "for i in range(obsgen.shape[1]):\n",
    "    with np.errstate(invalid='ignore'):\n",
    "        corr[i] = obsgen.loc[:, i].corr(predgen.loc[:, i])\n",
    "\n",
    "strict_snp_filter = True\n",
    "if strict_snp_filter:\n",
    "    valid_corr_indices = np.where((~np.isnan(corr)) & (corr > 0.8))[0]\n",
    "else:\n",
    "    valid_corr_indices = np.where(~np.isnan(corr))[0]\n",
    "\n",
    "dist11 = phase_subtraction(hrcfirstalleles, predfirstalleles, valid_corr_indices)\n",
    "dist12 = phase_subtraction(hrcfirstalleles, predsecondalleles, valid_corr_indices)\n",
    "dist21 = phase_subtraction(hrcsecondalleles, predfirstalleles, valid_corr_indices)\n",
    "dist22 = phase_subtraction(hrcsecondalleles, predsecondalleles, valid_corr_indices)\n",
    "\n",
    "phase1 = dist11 + dist22\n",
    "phase2 = dist12 + dist21\n",
    "\n",
    "d11 = dist11\n",
    "d21 = dist21\n",
    "d12 = dist12\n",
    "d22 = dist22\n",
    "\n",
    "if strict_snp_filter:\n",
    "    phased1 = (\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 < 4) & (phase2 > 4)) |\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase2 - phase1 > 2) & (phase1 < 4)) |\n",
    "        ((reftypes1 != 'nan') & (reftypes2 != 'nan') & (reftypes1 == reftypes2)) |\n",
    "        (np.isnan(d21) & ~np.isnan(d12) & (d12 - d22 > 2) & (d22 < 2)) |\n",
    "        (np.isnan(d12) & ~np.isnan(d21) & (d21 - d11 > 2) & (d11 < 2))\n",
    "    )\n",
    "\n",
    "    phased2 = (\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 > 4) & (phase2 < 4)) |\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 - phase2 > 2) & (phase2 < 4)) |\n",
    "        (np.isnan(d21) & ~np.isnan(d12) & (d22 - d12 > 2) & (d12 < 2)) |\n",
    "        (np.isnan(d12) & ~np.isnan(d21) & (d11 - d21 > 2) & (d21 < 2))\n",
    "    )\n",
    "else:\n",
    "    phased1 = (\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 < 4) & (phase2 > 4)) |\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase2 - phase1 >= 5)) |\n",
    "        ((reftypes1 != 'nan') & (reftypes2 != 'nan') & (reftypes1 == reftypes2)) |\n",
    "        (np.isnan(d21) & ~np.isnan(d12) & (d12 - d22 > 2)) |\n",
    "        (np.isnan(d12) & ~np.isnan(d21) & (d21 - d11 > 2))\n",
    "    )\n",
    "\n",
    "    phased2 = (\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 > 4) & (phase2 < 4)) |\n",
    "        (~np.isnan(phase1) & ~np.isnan(phase2) & (phase1 - phase2 >= 5)) |\n",
    "        (np.isnan(d21) & ~np.isnan(d12) & (d22 - d12 > 2)) |\n",
    "        (np.isnan(d12) & ~np.isnan(d21) & (d11 - d21 > 2))\n",
    "    )\n",
    "\n",
    "phased1old = phased1.copy()\n",
    "phased2old = phased2.copy()\n",
    "oldphase1 = phase1.copy()\n",
    "oldphase2 = phase2.copy()\n",
    "\n",
    "return_dict = {\n",
    "    'b38_start': full_start,\n",
    "    'b38_end': full_end,\n",
    "    'hlatypes': hlatypes,\n",
    "    'initial_phase_res': (oldphase1, oldphase2, phased1old, phased2old),\n",
    "    'merged_snps': merged_snps,\n",
    "    'db_hap1': predfirstalleles,\n",
    "    'db_hap2': predsecondalleles,\n",
    "    'vcf_hap1': hrcfirstalleles,\n",
    "    'vcf_hap2': hrcsecondalleles\n",
    "}\n",
    "\n",
    "phase_df = pd.DataFrame({'Sample': hlatypes['Sample ID'].values, 'first_step_phase1': phased1, 'first_step_phase2': phased2})\n",
    "\n",
    "for extension in range(50, 101, 50):\n",
    "    if np.sum(~(phased1 | phased2)) == 0:\n",
    "        break\n",
    "\n",
    "    allele1 = np.where(phased1, reftypes1, np.where(phased2, reftypes2, 'None'))\n",
    "    allele1[np.where(allele1 == '-9')[0]] = 'None'\n",
    "    allele2 = np.where(phased1, reftypes2, np.where(phased2, reftypes1, 'None'))\n",
    "    allele2[np.where(allele2 == '-9')[0]] = 'None'\n",
    "\n",
    "    all_haplotypes = np.append(allele1, allele2)\n",
    "    nameset = np.unique(all_haplotypes[all_haplotypes != 'None'])\n",
    "\n",
    "    newstart = start - extension\n",
    "    newend = end + extension\n",
    "\n",
    "    hrcfirstalleles = hrcfirstalleles_full[(hrcfirstalleles_full['pos'] >= newstart) & (hrcfirstalleles_full['pos'] <= newend)].drop(columns = ['snp', 'pos', 'ref', 'alt']).reset_index(drop = True).T\n",
    "    hrcsecondalleles = hrcsecondalleles_full[(hrcsecondalleles_full['pos'] >= newstart) & (hrcsecondalleles_full['pos'] <= newend)].drop(columns = ['snp', 'pos', 'ref', 'alt']).reset_index(drop = True).T\n",
    "    hrcalleles = pd.concat([hrcfirstalleles, hrcsecondalleles])\n",
    "\n",
    "    predmat = np.zeros((len(nameset), hrcalleles.shape[1]))\n",
    "    for i, name in enumerate(nameset):\n",
    "        predmat[i, :] = np.mean(hrcalleles.iloc[all_haplotypes == name, :], axis=0)\n",
    "    predmatallele1 = np.full((len(reftypes1), hrcalleles.shape[1]), np.nan)\n",
    "    predmatallele2 = np.full((len(reftypes1), hrcalleles.shape[1]), np.nan)\n",
    "\n",
    "    for i in range(len(reftypes1)):\n",
    "        if reftypes1[i] in nameset: \n",
    "            predmatallele1[i, :] = predmat[np.where(nameset == reftypes1[i])[0][0], :]\n",
    "        if reftypes2[i] in nameset: \n",
    "            predmatallele2[i, :] = predmat[np.where(nameset == reftypes2[i])[0][0], :]\n",
    "\n",
    "    dist11 = np.sum(np.abs(hrcfirstalleles - predmatallele1) > 0.9, axis=1)\n",
    "    dist12 = np.sum(np.abs(hrcfirstalleles - predmatallele2) > 0.9, axis=1)\n",
    "    dist21 = np.sum(np.abs(hrcsecondalleles - predmatallele1) > 0.9, axis=1)\n",
    "    dist22 = np.sum(np.abs(hrcsecondalleles - predmatallele2) > 0.9, axis=1)\n",
    "\n",
    "    phase1b = dist11 + dist22\n",
    "    phase2b = dist12 + dist21\n",
    "\n",
    "    d11 = dist11\n",
    "    d21 = dist21\n",
    "    d12 = dist12\n",
    "    d22 = dist22\n",
    "\n",
    "    phased1b = (\n",
    "        (~np.isnan(phase1b) & ~np.isnan(phase2b) & (phase1b < phase2b)) |\n",
    "        ((reftypes1 != 'nan') & (reftypes2 != 'nan') & (reftypes1 == reftypes2)) |\n",
    "        (np.isnan(d21) & ~np.isnan(d12) & (d12 - d22 > 2)) |\n",
    "        (np.isnan(d12) & ~np.isnan(d21) & (d21 - d11 > 2))\n",
    "    )\n",
    "\n",
    "    phased2b = (\n",
    "        (~np.isnan(phase1b) & ~np.isnan(phase2b) & (phase1b > phase2b)) |\n",
    "        (np.isnan(d21) & ~np.isnan(d12) & (d22 - d12 > 2)) |\n",
    "        (np.isnan(d12) & ~np.isnan(d21) & (d11 - d21 > 2))\n",
    "    )\n",
    "\n",
    "    update = ~phased1 & ~phased2\n",
    "    phased1[update] = phased1b[update]\n",
    "    phased2[update] = phased2b[update]\n",
    "\n",
    "phase_df['final_step_phase1'] = phased1\n",
    "phase_df['final_step_phase2'] = phased2\n",
    "\n",
    "phase_df['allele1'] = 'N/A'\n",
    "phase_df['allele2'] = 'N/A'\n",
    "\n",
    "for i in range(len(phase_df)):\n",
    "    p1, p2 = phase_df.loc[i, ['final_step_phase1', 'final_step_phase2']]\n",
    "    if (not p1) and p2:\n",
    "        phase_df.loc[i, 'allele1'] = reftypes2[i]\n",
    "        phase_df.loc[i, 'allele2'] = reftypes1[i]\n",
    "    elif (not p2) and p1:\n",
    "        phase_df.loc[i, 'allele1'] = reftypes1[i]\n",
    "        phase_df.loc[i, 'allele2'] = reftypes2[i]\n",
    "    else:\n",
    "        phase_df.loc[i, 'allele1'] = 'N/A'\n",
    "        phase_df.loc[i, 'allele2'] = 'N/A'\n",
    "\n",
    "return_dict['phase_df'] = phase_df\n",
    "\n",
    "unphased_indices = phase_df.index[~(phase_df['final_step_phase1'] | phase_df['final_step_phase2'])]\n",
    "unphased_df = pd.DataFrame({\n",
    "    'phase1': oldphase1[unphased_indices],\n",
    "    'phase2': oldphase2[unphased_indices],\n",
    "    'type1': reftypes1[unphased_indices],\n",
    "    'type2': reftypes2[unphased_indices]\n",
    "})\n",
    "return_dict['unphased_df'] = unphased_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Phasing commences\n",
      ">>> Parameters for this run:\n",
      ">>> - gene: B\n",
      ">>> - db directory: /well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/\n",
      ">>> - vcf file: /well/band/users/rbx225/recyclable_files/ref_panels/oneKG/oneKG.chr6.vcf.gz\n",
      ">>> - strict filtering on SNPs: True\n",
      ">>> - read from QUILT: False\n",
      ">>> - subset samples: False\n"
     ]
    }
   ],
   "source": [
    "# For oneKG\n",
    "gene = 'B'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "ref_samples = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/oneKG.samples', sep = ' ')\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "ref_samples_removed = ref_samples[~ref_samples['SAMPLE'].isin(hlatypes['Sample ID'].tolist())]\n",
    "samples_to_remove = ref_samples_removed['SAMPLE'].tolist()\n",
    "hlatypes = hlatypes[~hlatypes['Sample ID'].isin(samples_to_remove)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/oneKG.chr6.vcf.gz\"\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information,\n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            strict_snp_filter = True,\n",
    "                            read_from_QUILT = False, \n",
    "                            subset_vcf_samples = None,\n",
    "                            sample_linker = None)\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "# individual = 'NA12878'\n",
    "# ix = hlatypes.index[hlatypes['Sample ID'] == individual][0]\n",
    "# display_indices = np.arange(10)\n",
    "# html = '1KG-DRB1.html'\n",
    "\n",
    "# res = visualise_phase(gene, ix, hlatypes, return_dict, both_het = True)\n",
    "# compare_phase(display_indices, res, save_html = True, save_name = html)\n",
    "\n",
    "plot_first_step_phase(c)\n",
    "plot_last_step_phase(return_dict)\n",
    "\n",
    "# reference_exclude_samplelist_file = ''\n",
    "# if reference_exclude_samplelist_file != '':\n",
    "#     samples_to_remove = samples_to_remove + lcwgsus.read_tsv_as_lst(reference_exclude_samplelist_file)\n",
    "# if samples_to_remove is not none, save to \"hlauntyped.exclude.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiEAAAG3CAYAAABiwIVEAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAA9hAAAPYQGoP6dpAABpL0lEQVR4nO3deVxU5f4H8M8ZBoZNQFZFCVJUVEDMVLTcxXC7hUYu18qtxcy0zTRLsUxtz9vPFtOUm4pdF65lailqm7llFq4pbqQpCsgmDMzM8/uD4DoxA7Mwc2bg8369zqt4zvOc+Z5xdL4825GEEAJEREREdqaQOwAiIiJqnJiEEBERkSyYhBAREZEsmIQQERGRLJiEEBERkSyYhBAREZEsmIQQERGRLJiEEBERkSyYhBAREZEsmIQQmeD8+fOQJAnjx4+3+2unpKRAkiTs2bPHZq8RERGBiIgIm13fHKtWrYIkSVi1apXcoRCRjTEJoUZPkqRaD34ZEhHZhlLuAIgcxbx58wyWx8XFoUWLFjhx4gR8fX3tHFXjk5SUhPj4eDRv3lzuUIjIxpiEEP0lJSWl1vNRUVH2CaSR8/X1ZbJH1EhwOIbIBMbmhIwfPx6SJOH8+fP4+OOPERMTA3d3d4SEhOCRRx7BjRs3alxr9+7dePTRR9GhQwf4+PjAw8MDHTt2xLx581BaWmp1rFUxnT17Fu+88w6ioqLg7u6Oli1b4umnn0ZhYaHRtjdv3sTzzz+P2267DSqVCpGRkVi8eDEMPWx71apVGDlyJFq1agUPDw/4+Pjgrrvuwr///W+D1z5z5gwmT56M1q1bw93dHU2bNkX79u3x2GOPITc3V++6hobBquatmBOjEAJLlixBhw4d4O7ujhYtWuDJJ59EQUGBQ82DIWqs2BNCVA9mzpyJr7/+GsOHD8egQYOwe/duLF++HL///ju+/fZbvbqvv/46Tp48iZ49e2Lo0KEoLS3Fjz/+iFdeeQW7d+/Grl27oFRa/1fz6aefxnfffYcHHngA9957L77++mu89957+P777/HDDz/A3d1dr35FRQUGDRqEy5cvY/DgwVAqlfjvf/+L2bNno7S0FPPnz9erP2XKFHTo0AG9e/dG8+bNcf36dXz11Vd4+OGHcfLkSSxcuLC67uXLl9GtWzcUFRVhyJAhuP/++1FWVoZz585h9erVmDZtGgICAuq8J3NjnDp1Kj788EOEhobi0UcfhZubG7744gscOHAAFRUVcHV1teIdJiKrCaJGDoAAIObNm1fjWLlypRBCiHPnzgkA4uGHH9Zr+/DDDwsA4rbbbhMXLlyoLq+oqBC9evUSAMS+ffv02mRlZQmdTlcjjtmzZwsAIi0tTa983rx5AoDYvXu3SfdTFVNAQIA4f/58dblWqxUjRowQAMQrr7yi1yY8PFwAEIMHDxY3b96sLr969arw9fUVPj4+ory8XK/NmTNnarx2WVmZ6Nu3r1AqlSI7O7u6fMmSJQKAePfdd2u0KS4u1nvNlStXCgDV772lMX733XcCgGjbtq3Iz8+vLler1dV/NuHh4TXiISL74XAM0V/mz59f4zB1ZczcuXNx2223Vf+sVCoxYcIEAMDBgwf16rZq1QqSJNW4xrPPPgsA+Oabbyy8A33Tp09HeHh49c8KhQJvvvkmFAoFPv30U4Nt3n//fXh4eFT/HBwcjHvvvReFhYU4deqUXt3WrVvXaK9SqfDkk09Co9Fg165d1eVV9+vp6VmjjZeXl95r1sXUGFNTUwEAc+bMgZ+fX3W5m5sbFi1aZPLrEZHtMAkh+osQosZh6t4cd955Z42ysLAwAEB+fr5eeUlJCRYuXIiuXbvC19cXCoUCkiQhMDAQAHDp0iXrbuQvffr0qVHWqlUrhIWF4fz58zXmq/j5+RlMLIzdx8WLFzF16lRERUXB09Ozeknz/fffX+M+/vGPf8Db2xtTp05FcnIyli1bhmPHjhmcx1Ebc2L85ZdfAAB33313jfrx8fH1MuRFRNbh30KiemBoNUfVl5xWq60uq6ioQP/+/XHgwAFER0dj1KhRCAoKqp6bMH/+fKjV6nqJKSQkxGB5s2bNcOHCBRQUFOj1EBhbkWLoPs6ePYtu3bohPz8fvXr1wqBBg+Dr6wsXFxecP38eqampevcRHh6OAwcOICUlBdu3b8eGDRsAVCYPM2fOxJNPPmnSPZkTY0FBAQDD74OLi4tJc1CIyLaYhBDZ0ebNm3HgwAE8/PDDNYZ6/vzzzxoTK61x9epVtGvXrkb5lStXABj/QjfFO++8g9zcXKxcubLGiqG0tLTqoZBbtW/fHp9//jk0Gg1+/fVX7Ny5E++//z6mTZsGLy+v6uGr+uLj4wOg8n1o1aqV3jmtVovc3Fy0aNGiXl+TiMzD4RgiOzpz5gwAYOTIkTXO/X0VjbUMXe/s2bPIzs5GRESEXi+Iuay5D6VSiS5duuCFF15AWloaACA9Pd3iWIzp3LkzAOCHH36ocW7fvn3QaDT1/ppEZB4mIUR2VLUvxe7du/XKz549ixdeeKFeX2vJkiW4cOFC9c86nQ7PP/88dDqd1b0Oxu7j66+/xvLly2vUP3DgAK5evVqjvKrs78uF68NDDz0EAHjttdeqh2YAoLy8HC+++GK9vx4RmY/DMUR2NHz4cERGRuLdd9/F0aNH0blzZ1y8eBFbtmzB0KFDcfHixXp7rbvvvhtxcXEYNWoUfH198fXXX+PXX39Fly5dMHPmTKuu/cQTT2DlypV44IEHMHLkSLRo0QJHjx7F9u3b8cADD+Dzzz/Xq7927VosXboUffr0QWRkJJo2bYqsrCx8+eWXUKlUmD59ulXxGNKnTx88+uijWLZsGTp27IiRI0fC1dUVX375JXx9fREaGgqFgr+HEcmJSQiRHXl5eWHXrl2YNWsW9uzZg++//x6tWrXCyy+/jGeeeabGl7c13nnnHaSnp+OTTz7B+fPnERAQgOnTp+OVV16xuuchNjYWu3fvxksvvYStW7dCo9GgU6dO2LRpE/z8/Grcx5gxY6BWq7F3714cPnwYpaWlaNGiBUaPHo1nn30W0dHRVsVjzIcffoioqCh8/PHH+OijjxAQEICkpCQsXLgQLVu2RGRkpE1el4hMIwlz18gRkUMbP348UlNTce7cOW5LbsTp06fRtm1bjB49unpeChHZH/siiajBunLlCnQ6nV7ZzZs3MWPGDACGJ9YSkf0wCSGiBuu9997D7bffjocffhizZs3C+PHj0a5dO2zduhVDhw5lEkKNVlFREWbOnIlBgwYhKCgIkiTV+STxW+Xk5GD8+PEIDAyEp6cnevTogYyMDLPjYBJCRA1WQkICYmJikJGRgffeew/r169HUFAQ3nzzTaSnpxvcPp+oMcjNzcWyZcugVqtx3333mdVWrVZjwIAByMjIwJIlS7B582aEhIQgMTHR7K0GOCeEiIiokan66pckCdevX0dQUBDmzZtnUm/IBx98gKlTp2Lv3r3o0aMHAFRPTvf29sb+/ftNjoM9IURERI1M1bOeLJGeno527dpVJyBA5SaE48aNw4EDB8x6/hWTECIiIjLZ0aNHERsbW6O8quzYsWMmX6tR7BOi0+lw+fJlNGnShGPARERUKyEEioqKbL6hXVlZGcrLy62+jhDC4HebSqWCSqWy+vp/l5ubC39//xrlVWW5ubkmX6tRJCGXL1+uftQ3ERGRKbKzs9GyZUubXLusrAy3h3vjSo627sp18Pb2RnFxcY1yU+d4WKK2X+jN+WW/USQhTZo0AVD5gap6siYREZEhhYWFCAsLq/7usIXy8nJcydHiws8R8GlieW9LYZEO4V3OG/x+s0UvCAAEBAQY7O3Iy8sDAIO9JMY0iiSkKivz8fFhEkJERCaxx/C9dxMJ3k0sfx0d7P/9FhMTg8zMzBrlVWXmPIaBE1OJiIhkohU6qw97S0pKwsmTJ/WW4mo0GqxevRrdu3dHaGioyddqFD0hREREpG/btm0oKSlBUVERAOD48ePYsGEDAGDIkCHw9PTEpEmTkJqaiqysLISHhwMAJk6ciKVLlyI5ORmLFy9GcHAwPvjgA5w6dQo7d+40KwYmIURERDLRQUAHy/cMtabtlClTcOHCheqf169fj/Xr1wNA9QMwtVottFotbt3XVKVSISMjAzNnzsS0adNw8+ZNxMXFYdu2bejTp49ZMTSKHVMLCwvh6+uLgoICzgkhIqJa2eM7o+o1Lp9qafXE1NB2fzjt9xvnhBAREZEsOBxDREQkE60Q0FoxIGFNW0fAJISIiEgmcs4JcQRMQoiIiGSig4C2ESchZs0J2bVrFyZOnIioqCh4eXmhRYsWuPfee/Hzzz/XqHv48GEMHDgQ3t7e8PPzw4gRI3D27FmTX2vnzp3o0aMHPD09ERgYiPHjxyMnJ8eccImIiMiBmZWEfPjhhzh//jymT5+OrVu3YsmSJcjJyUF8fDx27dpVXe/kyZPo27cvysvL8Z///Aeffvopfv/9d/Tq1QvXrl2r83W+/fZbDB48GCEhIdi8eTOWLFmCnTt3YsCAAVCr1ebfJTVYao0G/z15ApO+SMf9/0nDzB3bcfjPy3KHRURkkqrhGGsOZ2bWEt2cnBwEBwfrlRUXFyMyMhLR0dHVm5Q88MAD2L17N7KysqqXDF24cAFt2rTB008/jddff73W1+nWrRtKSkrw66+/QqmsHDHau3cv7rrrLnzwwQeYMmWKWTfJJboNU05JMcZtWo8z+XlQANABcJEkaIXAP2M64ZW+A/jUZCIymz2X6P5+IgRNrFiiW1SkQ9v2V532+82sO/97AgJUPr2vQ4cOyM7OBlC5deuWLVswcuRIvTckPDwc/fr1Q3p6eq2vcenSJRw8eBAPPvhgdQICAD179kTbtm3rbE+NgxACU776Audu5AOoTECA/80UX5P5K1b9+otM0RERkSms3iekoKAAhw8fRseOHQEAWVlZKC0tRWxsbI26sbGxOHPmDMrKyoxe7+jRo9V1DbWvOk+N229Xr+CXK3/Wujxt2c8HodXZ/7kKRESm0tXD4cysXh0zdepUlJSUYM6cOQBQ/XhfQ4/y9ff3hxAC+fn5aN68ucHr1dXe0OODb6VWq2vMGyksLKz7RsipfH/xQvXQizFXS4px/kY+WvsH2DEyIiLTaa1cHWNNW0dgVU/Iyy+/jDVr1uDdd99Fly5d9M7VNhZvyji9sTp1tV20aBF8fX31jrCwsDpfj5yLRqcz6XOkcfKNfIiIGjKLk5D58+djwYIFeO211/Dkk09WlwcEVP7WaajHIi8vD5Ikwc/Pz+h162pvqIfkVrNnz0ZBQYHeUTVfhRqOzs2aQ1PHUEsTNzdE+PrZJyAiIgtohfWHM7MoCZk/fz5SUlKQkpKCF198Ue9c69at4eHhgczMzBrtMjMzERkZCXd3d6PXjo6Orq5rqH3VeWNUKhV8fHxqHNSw9AqPQJiPL1yM9IYoJAljYzpBpeR+fETkuBr7nBCzk5BXX30VKSkpeOmllzBv3rwa55VKJYYPH45NmzahqKiouvzixYvYvXs3RowYUev1W7RogW7dumH16tXQarXV5fv27cOpU6fqbE+Ng0KS8NHQf8DLzU0vEan6v66hLTCje095giMiIpOYtU/I22+/jeeeew6JiYkGE5D4+HgAlZuVde3aFXfccQdmzZqFsrIyzJ07F3l5eThy5AiCgoKq2yiVSvTp0wcZGRnVZXv27EFCQgKGDx+OJ554Ajk5OZg1axZ8fX1x6NAhqFQqs26S+4Q0XH8WFeHfv/2C/548geJyNcJ9/fDP2DiMbN8Rbi4ucodHRE7InvuEHD4eAm8r9gkpLtLhjg7Ou0+IWUlI37598e233xo9f+ulfv75Z7zwwgv46aefoFQq0b9/f7z11lto3bq1fgCShD59+mDPnj165Tt27MDcuXNx5MgReHp6YtiwYXjzzTcN7lVSFyYhRERkKnsmIYeOWZ+E3NmxkSQhzopJCBERmcqeScj+Y82sTkK6d7zitN9vVm9WRkRERGQJLh0gIiKSiRYStLD8GVfWtHUETEKIiIhkohMSdMLyRMKato6AwzFEREQkC/aEEBERyYTDMURERCQLLRTQWjEooa27ikPjcAwRERHJgj0hREREMhFWTkwVTj4xlUkIERGRTDgnhIiIiGShFQpohRVzQpx8z3POCSEiIiJZsCeEiIhIJjpI0FnRH6CDc3eFMAkhIiKSSWOfE8LhGCIiIpIFe0KIiIhkYv3EVA7HEBERkQUq54RY8QA7DscQERERmY89IURERDLRWfnsGK6OISIiIos09jkhHI4hIiIiWbAnhIiISCY6KLhZGREREdmfVkjQWvEkXGvaOgImIURERDLRWjkxVevkPSGcE0JERESyYE8IERGRTHRCAZ0Vq2N0Tr46hkkIERGRTDgcQ0RERCQD9oQQERHJRAfrVrjo6i8UWTAJISIikon1+4Q494CG2dEXFRVh5syZGDRoEIKCgiBJElJSUmrUkyTJ6BEVFVXn6/Tt29dg28TERHNDJiIiIgdkdk9Ibm4uli1bhk6dOuG+++7D8uXLDdb76aefapTt378fM2bMQFJSkkmv1apVK6xZs0avzM/Pz9yQiYiIHJL1z45x7p4Qs5OQ8PBw5OfnQ5IkXL9+3WgSEh8fX6Ps448/hiRJmDRpkkmv5eHhYfA6REREDYEOEnSwZk5II9sxVZIsu+GioiKsX78effr0QWRkpEXXICIioobDbv0469atQ0lJCSZPnmxym6ysLPj7+0OpVKJ169aYM2cOSktLbRglERGR/VQNx1hzODO7rY5ZsWIF/Pz8MHLkSJPq33333Rg1ahSioqJQWlqKbdu24Y033sAPP/yA3bt3Q6Ew/Mar1Wqo1Wq9ssLCQqvjJyIiqm/Wb1bGJKROx44dw/79+zF16lS4u7ub1GbBggV6Pw8ZMgQRERF47rnnsHnzZqOTWxctWoT58+dbHTMREZGt6YQEnTX7hDj5U3TtkkKtWLECAMwaijFk3LhxAIB9+/YZrTN79mwUFBToHdnZ2Va9LhEREdU/m/eElJeX47PPPkOXLl0QFxdXL9c0NhQDACqVCiqVql5eh4iIyJZ0Vg7HNLrNysz1xRdf4Pr16yYvy61NamoqAMPLf4mIiJxN1VN0rTmcmUU9Idu2bUNJSQmKiooAAMePH8eGDRsAVM7d8PT0rK67YsUKeHh4YOzYscaDUCrRp08fZGRkAAC+//57vPbaa0hKSkKrVq1QVlaGbdu2YdmyZejfvz+GDx9uSdhERETkQCxKQqZMmYILFy5U/7x+/XqsX78eAHDu3DlEREQAALKzs/HNN99g3Lhx8PX1NXo9rVYLrVZb/XPz5s3h4uKCV199FdevX4ckSWjTpg1eeeUVPPvss7UOxxARETkLLSRordhwzJq2jsCiJOT8+fMm1QsLC9NLLowRQuj9HBkZia+++sqS0IiIiJyGtUMqzj4c49zRExERkdOy22ZlREREpE8L64ZU6h5rcGzsCSEiIpKJnKtjiouLMWPGDISGhsLd3R1xcXFYt26dSW13796NhIQEBAcHw9vbG7GxsfjXv/5l0hSMW7EnhIiIqBEaMWIEDh48iMWLF6Nt27ZYu3YtxowZA51OV+uK1p07d+Kee+5B79698cknn8DLywtffPEFpk+fjqysLCxZssTkGJiEEBERycTah9BZ2nbr1q3YsWNHdeIBAP369cOFCxfw/PPPY9SoUXBxcTHYdtWqVXB1dcWWLVvg5eUFABg4cCBOnTqFVatWmZWEcDiGiIhIJgISdFYcwsL5JOnp6fD29kZycrJe+YQJE3D58mXs37/faFtXV1e4ubnBw8NDr9zPz8/k58NVYRJCREQkk6qeEGsOSxw9ehTt27eHUqk/IBIbG1t93pjHH38c5eXleOqpp3D58mXcuHEDn332GdLT0zFz5kyz4uBwDBERkZMrLCysUVbbs9Ryc3PRqlWrGuX+/v7V543p3r07du3aheTkZCxduhQA4OLigkWLFuHZZ581K24mIURERDLRCQk6YfkS3aq2YWFhNc7NmzcPKSkpRttKkvHXre3czz//jKSkJHTv3h0ff/wxvLy8sGvXLrz00ksoKyvDyy+/bHL8TEKIiIhkorXyKbpVbbOzs+Hj46N3rrYnygcEBBjs7cjLywPwvx4RQ6ZOnYqQkBCkp6dXT17t168fFAoFUlJS8M9//tNgL4shnBNCRETk5Hx8fGoctSUhMTExOHHiBDQajV55ZmYmACA6Otpo2yNHjqBLly41Vs907doVOp0OJ06cMDluJiFEREQyqRqOseawRFJSEoqLi7Fx40a98tTUVISGhqJ79+5G24aGhuLQoUM1Nib76aefAAAtW7Y0OQ4OxxAREclEBwV0VvQHWNp28ODBSEhIwJQpU1BYWIjIyEikpaVh+/btWL16dXUvx6RJk5CamoqsrCyEh4cDAJ5++mk89dRTGD58OB577DF4enoiIyMDb7/9NgYOHIhOnTqZHAeTECIiokZo06ZNmDNnDubOnYu8vDxERUUhLS0No0ePrq6j1Wqh1Wr1nnY/bdo0tGjRAu+++y4mT56M0tJSREREYN68eXj66afNikESt165gSosLISvry8KCgpqTNwhIiK6lT2+M6peY8r3I6DydrX4OuriCnzYa5PTfr+xJ4SIiEgm9bVE11lxYioRERHJgj0hREREMhFCAZ0VD7ATVrR1BExCiIiIZKKFBK2FD6Grau/MmIQQERHJRCesm9ehc/KlJc7dj0NEREROiz0hREREMtFZOSfEmraOgEkIERGRTHSQoLNiXoc1bR2Bc6dQRERE5LTYE0JERCQTrZCgtWJiqjVtHQGTECIiIpk09jkhzh09EREROS32hBAREclEByufHePkE1OZhBAREclEWLk6Rjh5EmL2cExRURFmzpyJQYMGISgoCJIkISUlpUa98ePHQ5KkGkdUVJTJr7Vz50706NEDnp6eCAwMxPjx45GTk2NuyEREROSAzO4Jyc3NxbJly9CpUyfcd999WL58udG6Hh4e2LVrV40yU3z77bcYPHgwhg4dis2bNyMnJwcvvPACBgwYgEOHDkGlUpkbOhE1QIVqNdJPHsPOs1lQa7SIDgnBP6Nj0do/QO7QiOqkE1YOxzS21THh4eHIz8+HJEm4fv16rUmIQqFAfHy8RYE9//zzaNu2LTZs2AClsjLM22+/HXfddRc+/fRTTJkyxaLrElHDceJaDh5M34D8slIAgADwy5XLSD1yGC/37ofxcXfIGyBRHbg6xkxVwyq2dOnSJRw8eBAPPvhgdQICAD179kTbtm2Rnp5u09cnIsdXWlGBh/67EQXqMghUJiAAoBUCAsAr3+3G9xfPyxcgkQmqekKsOZyZTVOo0tJSNGvWDC4uLmjZsiWefPJJ5OXl1dnu6NGjAIDY2Nga52JjY6vPE1HjteX0KeSW3oRWGH6MqIskYfnhQ3aOiojMYbPVMZ06dUKnTp0QHR0NoHKOx7vvvouMjAwcPHgQ3t7eRtvm5uYCAPz9/Wuc8/f3rz5viFqthlqt1isrLCy05BaIyIF9f+E8FJIEnZEkRCsEfsy+CJ0QUNi495bIUo392TE2S0KefvppvZ8TEhLQuXNn3H///fjkk09qnDfE2LBPbcNBixYtwvz5880LloicjlYICCMJSBVdVR0mIeSgGvvEVLvOaElKSoKXlxf27dtXa72AgMpZ7YZ6PPLy8gz2kFSZPXs2CgoK9I7s7GzrAicihxPXrFmt5xWShI5BwXBROPfEPaKGzO5/O4UQUNTxj0LVEE5mZmaNc5mZmdXnDVGpVPDx8alxEFHDcn/7aKhclEY7o3VCYGJcF7vGRGQuTky1ow0bNuDmzZt1Lttt0aIFunXrhtWrV0Or1VaX79u3D6dOncKIESNsHSoRObimHh54f8gwuCgUcLlluKVq/sfojjG4L6q9XOERmaSxJyEWzQnZtm0bSkpKUFRUBAA4fvw4NmzYAAAYMmQIrl27hrFjx2L06NGIjIyEJEn49ttv8d5776Fjx46YPHmyfhBKJfr06YOMjIzqstdffx0JCQlITk7GE088gZycHMyaNQvR0dGYMGGCpfdLRA3IgNtbY8uYB7HqyGF8nXUG5VotooOD8VCnzkhs3cbm2wkQkXUkUdfMLgMiIiJw4cIFg+fOnTsHX19fTJo0Cb/88guuXr0KrVaL8PBwJCUl4cUXX4Svr69+EJKEPn36YM+ePXrlO3bswNy5c3HkyBF4enpi2LBhePPNNxEcHGxWvIWFhfD19UVBQQGHZoiIqFb2+M6oeo2ErY/B1cvN4utUlJRjx5CPnfb7zaKekPPnz9dZZ9OmTSZfz1gelJCQgISEBJOvQ0RE5EwErFtma3YvgoPhtHEiIiKShc32CSEiIqLaNfZ9QpiEEBERyYRJCBEREcmisSchnBNCREREsmBPCBERkUwae08IkxAiIiKZCCFBWJFIWNPWEXA4hoiIiGTBnhAiIiKZ6CBZtVmZNW0dAZMQIiIimTT2OSEcjiEiIiJZsCeEiIhIJo19YiqTECIiIplwOIaIiIhIBuwJISIikgmHY4iIiEgWwsrhGCYhREREZBEBQAjr2jszzgkhIiIiWbAnhIiISCY6SJC4YyoRERHZW2OfmMrhGCIiIpIFe0KIiIhkohMSpEa8WRmTECIiIpkIYeXqGCdfHsPhGCIiIpIFe0KIiIhk0tgnpjIJISIikkljT0I4HENERESyYE8IERGRTLg6hoiIiGTB1TFmKioqwsyZMzFo0CAEBQVBkiSkpKTo1dFqtXjnnXeQmJiIli1bwtPTE+3bt8esWbNw48YNk16nb9++kCSpxpGYmGhuyERERA6pMgmRrDjkvgPrmN0Tkpubi2XLlqFTp0647777sHz58hp1SktLkZKSgjFjxmDy5MkIDAzE4cOHsWDBAnz55Zc4dOgQPDw86nytVq1aYc2aNXplfn5+5oZMREREDsjsJCQ8PBz5+fmQJAnXr183mIR4eHjg3LlzCAgIqC7r27cvbrvtNiQnJ2Pjxo0YN25cna/l4eGB+Ph4c0MkIiJyCo19dYzZSYgk1X3DLi4ueglIlW7dugEAsrOzzX1ZIiKiBkf8dVjT3pnZdYnurl27AAAdO3Y0qX5WVhb8/f2hVCrRunVrzJkzB6WlpbYMkYiIiOzEbknIpUuXMGvWLNx5550YNmxYnfXvvvtuvPPOO9i4cSO++OILDBkyBG+88QYSExOh0+mMtlOr1SgsLKxxEBERORrrJqVaN5RTXFyMGTNmIDQ0FO7u7oiLi8O6detMbr9582b06dMHPj4+8PLyQseOHbFs2TKzYrDLEt28vDwMGTIEQgh8/vnnUCjqzn0WLFig9/OQIUMQERGB5557Dps3b0ZSUpLBdosWLcL8+fPrJW4iIiKbknE8ZsSIETh48CAWL16Mtm3bYu3atRgzZgx0Oh3Gjh1ba9vFixdjzpw5ePzxxzF79my4urri5MmTKC8vNysGmych+fn5SEhIwKVLl7Br1y60atXK4muNGzcOzz33HPbt22c0CZk9ezaeeeYZvbLCwkKEhYVZ/LpEREQNydatW7Fjx47qxAMA+vXrhwsXLuD555/HqFGj4OLiYrDtzz//jDlz5mDRokWYOXNmdfmAAQPMjsOmwzH5+fkYOHAgzp07hx07diA2NrZerltbT4pKpYKPj0+Ng4iIyOFYOxRj4XBMeno6vL29kZycrFc+YcIEXL58Gfv37zfa9v/+7/+gUqkwbdo0i177VjZLQqoSkLNnz+Kbb75B586drb5mamoqAHDZLhERNQhVO6Zac1ji6NGjaN++PZRK/QGRqs6Co0ePGm373XffoX379ti4cSPatWsHFxcXtGzZErNmzbLPcMy2bdtQUlKCoqIiAMDx48exYcMGAJVzNyRJwj333INffvkF7733HjQaDfbt21fdPigoCK1bt/5fEEol+vTpg4yMDADA999/j9deew1JSUlo1aoVysrKsG3bNixbtgz9+/fH8OHDLQmbiIioQTK0AEOlUkGlUhmsn5uba3B6hL+/f/V5Yy5duoRr167hqaeewquvvooOHTogIyMDixcvRnZ2do1NRmtjURIyZcoUXLhwofrn9evXY/369QCAc+fOAQAOHjwIAJg+fXqN9g8//DBWrVpV/bNWq4VWq63+uXnz5nBxccGrr76K69evQ5IktGnTBq+88gqeffZZkya2EhERObr62qzM0LzHefPm1Xisyq1q2/ertnM6nQ5FRUVIS0vD6NGjAVTOJykpKcF7772H+fPnIzIy0qT4LUpCzp8/X2cdYUYf0d/rRkZG4quvvjI3LCIiIudixbyO6vao3AT07/MfjfWCAEBAQIDB3o68vDwA/+sRMdb2ypUruOeee/TKBw8ejPfeew+HDx82OQlhlwIREZFM6mtOiKEFGbUlITExMThx4gQ0Go1eeWZmJgAgOjraaFtji0yqOhTMGa1gEkJERNTIJCUlobi4GBs3btQrT01NRWhoKLp372607ciRIwFUzg+91datW6FQKNC1a1eT47DLZmVERERkgEyblQ0ePBgJCQmYMmUKCgsLERkZibS0NGzfvh2rV6+u3iNk0qRJSE1NRVZWFsLDwwFULuP9+OOP8cQTT+D69evo0KEDdu7ciaVLl+KJJ56ormcKJiFEREQykfMpups2bcKcOXMwd+5c5OXlISoqSm+yKfC/hSO3zt10dXXFjh078OKLL2LhwoXIy8vD7bffjsWLF9fYLLQukjBnBqmTKiwshK+vLwoKCrhxGRER1coe3xlVr3HbsrlQeLpbfB3dzTJcfPQVp/1+Y08IERGRnBp8V4BxTEKIiIhkIudwjCPg6hgiIiKSBXtCiIiI5CLT6hhHwSSEiIhINtJfhzXtnReHY4iIiEgW7AkhIiKSC4djiIiISBZMQoiIiEgW9fQUXWfFOSFEREQkC/aEEBERyUSIysOa9s6MSQgREZFcGvmcEA7HEBERkSzYE0JERCSXRj4xlUkIERGRTCRReVjT3plxOIaIiIhkwZ4QIiIiuTTyialMQoiIiOTSyOeEcDiGiIiIZMGeECIiIrlwOIaIiIhkwSSEiIiIZNHIkxDOCSEiIiJZsCeEiIhILlwdY56ioiLMnDkTgwYNQlBQECRJQkpKisG6hw8fxsCBA+Ht7Q0/Pz+MGDECZ8+eNfm1du7ciR49esDT0xOBgYEYP348cnJyzA2ZiIjIIVXtmGrN4czMTkJyc3OxbNkyqNVq3HfffUbrnTx5En379kV5eTn+85//4NNPP8Xvv/+OXr164dq1a3W+zrfffovBgwcjJCQEmzdvxpIlS7Bz504MGDAAarXa3LCJiKiBEtqrEMXvQ5c3Drq8cRDF70Nor8odFpnA7OGY8PBw5OfnQ5IkXL9+HcuXLzdYb+7cuVCpVNiyZQt8fHwAAF26dEGbNm3w1ltv4fXXX6/1dZ5//nm0bdsWGzZsgFJZGebtt9+Ou+66C59++immTJlibuhERNTAiLLdEDemAdAA0FWWlR8Cij8G/N6H5N5P1vjqxImp5pEkCZJU+xiURqPBli1bMHLkyOoEBKhMYPr164f09PRa21+6dAkHDx7Egw8+WJ2AAEDPnj3Rtm3bOtsTEVHDJzQXIW48CaACVQlIJR2ACogbT0JoLsoTHJnEJqtjsrKyUFpaitjY2BrnYmNjcebMGZSVlRltf/To0eq6htpXnSciosZL3FyDyoTDUHeAAKCDuLnWvkGRWWyShOTm5gIA/P39a5zz9/eHEAL5+fkWt686b4harUZhYWGNg4iIGhj1HgDaWipo/6rjuCRYOTFV7huwkk33Calt2KauIZ3a6tTWdtGiRfD19dU7wsLC6g6WiIicTG0JSBWNzaOwStUSXWsOJ2aTJCQgIAAADPZY5OXlQZIk+Pn5WdzeUA9JldmzZ6OgoEDvyM7ONvMOiIjI4bneCcCllgouf9UhR2WTJKR169bw8PBAZmZmjXOZmZmIjIyEu7u70fbR0dHVdQ21rzpviEqlgo+PT42DiIgaFsnrn6hrOEbyGmevcCwj6uFwYjZJQpRKJYYPH45NmzahqKiouvzixYvYvXs3RowYUWv7Fi1aoFu3bli9ejW02v99wPbt24dTp07V2Z6IiBo+yTUGUpMX//rp1h6Ryv+XmrwIydX4L60OoZEnIRZt275t2zaUlJRUJxjHjx/Hhg0bAABDhgyBp6cn5s+fj65du2LYsGGYNWsWysrKMHfuXAQGBuLZZ5/VD0KpRJ8+fZCRkVFd9vrrryMhIQHJycl44oknkJOTg1mzZiE6OhoTJkyw9H6JiKgBkbzGA64dIUpSgfKfKgvdekDyehiSW1dZYzOFtbueOvuOqRYlIVOmTMGFCxeqf16/fj3Wr18PADh37hwiIiIQFRWFPXv24IUXXsD9998PpVKJ/v3746233kJQUJDe9bRarV6PBwD07dsXW7duxdy5czF8+HB4enpi2LBhePPNN6FSqSwJm4iIGiDJratTJBxUk0VJyPnz502q16VLF+zcubPOekIYTuUSEhKQkJBgTmhERETOo5HvmMqn6BIREcmlkSchNt0nhIiIiMgY9oQQERHJhBNTiYiISB7W7nrKHVOJiIiIzMeeECIiIrk08ompTEKIiIhk0tjnhHA4hoiIiGTBnhAiIiK5cDiGiIiIZGHlcAyTECIiIrJMI+8J4ZwQIiIikgV7QoiIiOTSyHtCmIQQERHJhEt0iYiIiGTAJISIiIhkweEYIiIiuTTyOSHsCSEiIiJZsCeEiIhIJo19YiqTECIiIjk5eSJhDQ7HEBERkSzYE2Km4vJyLD24DzuyzqBcp0Vk0wCMjYlF/9tbQyFJcodHRETOpJFPTGUSYoY3f/wOH/18UO/P/I/CQuy5cA6DWkXi/cHD4OriIlt8RETkXBr7nBAOx5hoypbN+PBvCcitdpw9g6UH99s1JiIicnKiHg4nxiTEBN+cOY2vz56ptY4AkPrrYag1GvsERUREZIXi4mLMmDEDoaGhcHd3R1xcHNatW2f2dV566SVIkoTo6Giz23I4pg5CCLyQ8Y1JdQvUapzJy0XH4BAbR0VERA2BnMMxI0aMwMGDB7F48WK0bdsWa9euxZgxY6DT6TB27FiTrnHkyBG89dZbCAmx7HuPSUgdTuZeR4G6zOT6Tt4zRkRE9iTTxNStW7dix44d1YkHAPTr1w8XLlzA888/j1GjRsGljjmOGo0GEyZMwGOPPYZff/0V169fNzsODsfU4Wpxscl1PZWuaOMfYMNoiIiIrJeeng5vb28kJyfrlU+YMAGXL1/G/v11z3FcvHgx8vLy8Nprr1kcB5OQOgR6eppcd3zcHVAp2blEREQmkmli6tGjR9G+fXso//adFRsbW32+NsePH8eCBQvw4Ycfwtvb27IgYMMkZPz48ZAkyeixb98+o21XrVpltN2VK1dsFbJBHYOC0cqvaZ31bvPxxVPde9ghIiIiaiiq5oRYcwBAYWFhjUOtVht93dzcXPj7+9coryrLzc012lan02HixIkYMWIEhgwZYtX92+zX9pdffhmPP/54jfLhw4dDpVKha9eudV5j5cqViIqK0isLCLDvcIckSZjTuy8mf5FuNOFs5dcUX48bDxcFO5aIiMj+wsLCapTNmzcPKSkpRttItWywWdu5d955B6dPn8YXX3xhVoyG2CwJad26NVq3bq1X9u233+L69et46aWX6pzwAgDR0dG48847bRWiyfpFtMLHw+7F3D0ZuHLLHBFXhQLP9bgbj3SpO6EiIiKqoZ4mpmZnZ8PHx0fvlEqlMtosICDAYG9HXl4eABjsJQGAixcvYu7cuVi8eDHc3Nxw48YNAJWTVHU6HW7cuAGVSgUPDw+TwrfrBIYVK1ZAkiRMnDjRni9bLwa2ikS/iFbYdykbfxYVIcDTE3eFhcONO6QSEZGl6ikJ8fHxqZGE1CYmJgZpaWnQaDR680IyMzMBwOieH2fPnkVpaSmmT5+O6dOn1zjftGlTTJ8+He+9955JcdgtCSkoKMCGDRswYMAA3H777Sa1GTZsGK5duwZfX1/07dsXr7zyikWbodQXF4UCd4WFy/b6RERE9SEpKQmffPIJNm7ciFGjRlWXp6amIjQ0FN27dzfYLi4uDrt3765RPmPGDBQUFGDlypVo2bKlyXHYLQlJS0tDaWkpJk2aVGfdZs2aYc6cOYiPj4ePjw8yMzOxePFixMfH48cff0SnTp2MtlWr1TUm4xQWFlodPxERUX2Ta7OywYMHIyEhAVOmTEFhYSEiIyORlpaG7du3Y/Xq1dVTJiZNmoTU1FRkZWUhPDwcfn5+6Nu3b43r+fn5QaPRGDxXG7slIStWrEBAQACSkpLqrJuYmIjExMTqn3v37o2hQ4ciJiYGc+fOxebNm422XbRoEebPn18vMRMREdmUjE/R3bRpE+bMmYO5c+ciLy8PUVFRSEtLw+jRo6vraLVaaLVaCGGbrTglYasr3+K3335Dp06dzBonMmTw4ME4fPgwrl69arSOsZ6QsLAwFBQUmDVmRkREjU9hYSF8fX1t+p1R9Rrtn1wIF5W7xdfRqstw4v9edNrvN7v0hKxYsQIAMHnyZKuuI4SAoo5lsCqVqtYZwUREROQYbJ6EqNVqrF69Gt26dbNqUum5c+fw448/YuDAgfUYneMqUqux+dQJnMq9Dg+lEgmtI3Fn8xa1rt0mIiInI+NwjCOweRLy3//+F3l5eUZ7Qf4+6QUABg4ciN69eyM2NrZ6Yuobb7wBSZLw6quv2jpkWeSV3sT640ex82wWckqKcbmoCFohoPyr52f5Lz+jS/NQLBt2H5qauP6aiIgcHJMQ21qxYgW8vLz0JrrcytCkl5iYGHz++ed46623UFpaiuDgYPTv3x8vv/wy2rZta+uQ7e63q1fw0H83oEitrvF50uh01f9/5MqfmPxlOjYkj2GPCBEROT2bJyHffPNNredXrVqFVatW6ZW9++67NozIsdysqMD4zRtRXF5eZ0KrFQK/XPkT+/7IRo+w2+wSHxER2Y7012FNe2fGh53I7ItTJ3CjrAw6ExcpKRUKfJ112sZRERGRXcj0FF1HwSREZj9mX4DCjKEVIQRKKipsGBEREZF92PXZMVSTVidgzk4tOiHQxt++TxImIiLbkGvHVEfBnhCZdQltAXP601wUCoxo39F2ARERkf1wOIbkdH/7jnBXKuucXFQ1ZPNqv4EI9PS0fWBEREQ2xiREZr7u7vhw6L1wdXGByy1zQ/6elNzRLBQr/zECozrG2DdAIiKyrUbaCwJwTohD6B0ega1jH8Jnvx3B12dOo1yrRXRwCMbFdkJMcAjcla7wdbf82QJEROSYGvucECYhDqJVU3/M69Mf8/r0lzsUIiKyl0a+YyqHY4iIiEgW7AkhIiKSCYdjiIiISB4cjiEiIiKyP/aEmKGkrAz/OrQPFwpuIDqoGf7RLgq3+frJHRYRETkpDseQSXqv/AR/FBVW//xNVhbe2fcjAGBp4jAMbttOrtAsUlBWhn//+gvWHz+K66U3oVQo0DEoGI916YY+4RGQzHieDRERWaiRD8cwCTFB9If/ws1aHho3dfsWjDh7Fm8lDrZjVJY7fi0HYzZ+jqLycr3y/Zf+wP5LfyApqgPeTEg068F6RERE5uKckDos+uHbWhOQKpt+P45rxUV2iMg6ZZoKPJS+oUYCcqv0k8ex6shhO0ZFRNRI8dkxVJvlhw+ZXPcf69bYMJL6seX3U8grK62z3opffobOnMf7EhGR2armhFhzODMmIXUw58/36s0Sm8VRX/ZmX6zzYXkA8GdxES7fMgeGiIiovnFOSCOjFcLkxIodIURENtbIJ6ayJ6QeeSkdP6fr0jzUpHohXt4IbdLExtEQETVukhBWH86MSUgd/FUqk+t+M26CDSOpH0lRHeBpQrI0qXMXuCj48SAisilOTKXafDD0XpPqRfj6ormPj42jsV4TlQofD78PrrUkGP9oG4UJcXfYMSoiImqMmITUoVvLMEzt0r3WOv3Cw7Hr4cl2ish6d4WF4+tx4zE6Ohbebm5QSBJULi6IDgrGsmH34t17hrAXhIjIDhr76hjHn8TgAJ69626M6NABE79Ix8WCGxCofOOSO8ZgXp/+cHOCuSB/F+HXFAv7J2Bh/wS5QyEiarwa+cRU5/v2lMntTf2x++FJcodBRETUYDAJISIikgkfYEdERETyaOTDMZx9SERERLJgT4iD0up0+PnPy8grLUXzJk0QGxwCiU+1JSJqUBr7cIzNekL27NkDSZIMHvv27auzfU5ODsaPH4/AwEB4enqiR48eyMjIsFW4DkMIgbm7d6Ld/72L0Rs/xxNbv0DS52vQe+Un2Jt9Ue7wiIioPjXyzcps3hOycOFC9OvXT68sOjq61jZqtRoDBgzAjRs3sGTJEgQHB2Pp0qVITEzEzp070adPH1uGLJvvzp/F+C/SDZ67VFyEcenrsSYpGT3CbrNzZEREZCvO3pthDZsnIW3atEF8fLxZbVasWIGjR49i79696NGjBwCgX79+6NSpE2bOnIn9+/fbIlRZ/SPtMxy9llNnvce3bMaRx5/k0AwRETk9h5yYmp6ejnbt2lUnIACgVCoxbtw4HDhwAJcuXZIxuvo38YuNJiUgAFBUUY7j16/ZOCIiIrILIaw/nJjNk5CpU6dCqVTCx8cH99xzD3744Yc62xw9ehSxsbE1yqvKjh07Vu9xyqVIrcae8+fNarMz64xtgiEiIrvitu024uvri+nTp6Nv374ICAjAmTNn8Oabb6Jv37746quvcM899xhtm5ubC39//xrlVWW5ublG26rVaqjVar2ywsJCC+/C9jYcP2p2m5ziIhtEQkREZF82S0I6d+6Mzp07V//cq1cvJCUlISYmBjNnzqw1CQFQ65yH2s4tWrQI8+fPNz9gmWQXFpjdxkulskEkRERkd9yszH78/PwwbNgw/PbbbygtLTVaLyAgwGBvR15eHgAY7CWpMnv2bBQUFOgd2dnZ1gdvIxF+Tc1u09Td3QaREBGRvUk66w9nZveJqeKvSTS19WbExMQgMzOzRnlVWW1LfFUqFXx8fGocjuof7aLMbhPs3cQGkRAREdmXXZOQ/Px8bNmyBXFxcXCv5bf5pKQknDx5Um8prkajwerVq9G9e3eEhobaI1y78HP3wON3dDW5vptCgUGtIm0YERER2Q03K7ONsWPH4rbbbsOdd96JwMBAnD59Gm+//TauXr2KVatWVdebNGkSUlNTkZWVhfDwcADAxIkTsXTpUiQnJ2Px4sUIDg7GBx98gFOnTmHnzp22Clk2z9/VC65KF7x/oO6dZJ/teTeacE4IEVGD0Ni3bbdZEhIbG4vPP/8cH330EYqLi+Hv74+7774bn332Gbp2/d9v/lqtFlqttnqYBqgcUsnIyMDMmTMxbdo03Lx5E3Fxcdi2bVuD3C1VkiQ8HX8XptzZDV+fPo2PDx/Eqdzregmut6sbnulxFx7u1NnodYiIiJyJJIST73RigsLCQvj6+qKgoMCh54fcSicEDl76A9mFBWjq7oG7bwuHSsnnDRIR2Zo9vjOqXqPbP16F0tXyxQaaijIc+OJlp/p+uxW/1RyUQpLQvWUYuiNM7lCIiMhGOBxDRERE8uA+IWQJtUaDnJJilFZUyB0KERGRU2JPiJlO5+ViTsYOHL5yGTohoAAwpE07PNW9ByL9A+QOz2yZV69gx7ksuLsokdyhI4K8vOUOiYio0eBwDJmkRK1GzMf/V6NcB2DL6VPIOJuFdcmjERMcYv/gzFRSXo6UPTux8eQJvfK3fvoBnZs1x5oRD8DdiSbBCiFw8PIlbDpxDNduliDEyxv3d4hG52bNa90Uj4hIdtY+CdfJ15Y4zzeNjFr96+0665RqNXj26634etx4h/7iu1xQgLtTlxs9/8uVP5Hw2af49uHJUCgcf7ROrdFg2rYt2HkuCy6SBK0QcJEkrDuWiWFt2uHtQYPh6uIid5hERGSA43/LyMyUBKTKmfw8HL5y2YbRWOdKUWGtCUiVS0VF+NyCp/vK4dXv92DX+bMAAO1fvxFU/fer06fw9k8/yBQZEVHdqoZjrDmcGZOQWhzIvmh2m9MGHrznKBJWrzK57ic/H7RdIPUkr/Qm/nMsEzoj3ZECwL9/O4Iitdq+gRERmaqRb9vOJKQWo9PXm93Gw9XVBpFY76c/slFixkqenJslNoymfuzNvgiNrvZHSJZpNDh4+ZKdIiIiInNwTkg9uyssXO4QDFqbecSs+q4Kx59HUaE17RnWFTqtjSMhIrJMY18dw56Qenb4T8f8rbtca94XcbcWLW0USf2JNmElkgSgY1Cw7YMhIrKETlh/ODEmIbXwcXUzu82Os1k2iMR6PVveZlb9+X362SiS+tMmIABdQ1vCxchqJBdJQp/w29HSx9fOkRERkSmYhNRiz4RHzG6j1mpsEIn1RkXHQGHi0uGxHWPRrIlzPAjprYRE+Ht41khEFJKEEG9vLByQIFNkREQm4MRUMsbP3R39IiJMrq+QJEQFBNkuICu4K13x3j1D6qw3qXMXzO83wA4R1Y8wX19sGfsgHrmjK/w9PCABCPL0xBN3dscXo8ehmXcTuUMkIjJKgpVLdOW+ASsxCanDin+MxFt9B5pUVyFJSO4YbeOILDesbRTWjRyFVn5N9co9lUpM6xqPA5OnYE6vvnBxgk3KbhXk6YWZd/XCoUeewJlpz2D/5Cl4psdd8PfwlDs0IqLaVe2Yas1hoeLiYsyYMQOhoaFwd3dHXFwc1q1bV2e7TZs2YcyYMYiMjISHhwciIiLwz3/+E6dPnzY7Bq6OMcGI2E4YEdsJE9M3YE/2hRrnXSQJOiGweMAgBHl6yRCh6bq1aImdD02EWqOBWquBt5vK5GEaZ+DIu9USETmSESNG4ODBg1i8eDHatm2LtWvXYsyYMdDpdBg7dqzRdq+//jqaNWuGOXPmoFWrVsjOzsbChQtxxx13YN++fejYsaPJMUhCOPnG8yYoLCyEr68vCgoK4ONj3VyHCq0Wn/12BKt+PYw/CgshAegTfjse69IV3VuG1U/AREQkm/r8zqjrNe7unwKl0t3i62g0ZfhhV4rZsW7duhVDhw6tTjyqDBo0CMeOHcPFixfhYuSRFzk5OQgO1l91ePnyZUREROChhx7C8uV178xdhT0hZnJ1ccHEzl0wIe4OlGk0UCoUfDYJERFZxtrJpRa2TU9Ph7e3N5KTk/XKJ0yYgLFjx2L//v3o2bOnwbZ/T0AAIDQ0FC1btkR2drZZcTjX4L8DkSQJHq6uTECIiMjpHD16FO3bt4fyb09Mj42NrT5vjrNnz+LChQtmDcUA7AlxeBVaLRSS5HSTRYmIqG6SEJCsmBVR1bawsLDGOZVKBZVKZbBdbm4uWrVqVaPc39+/+rypNBoNJk2aBG9vbzz99NMmtwOYhDgkrU6HN378HutPHMWNsjIAQHyLlphyZ3f0Co+QNzgzXLtZgs0nT+Bc/p/Qlv+KviG/IcRDDbXiTrQPewy+ns65k6moOAWo9wAoB5QdAFUfSBL/KhGRBXR/Hda0BxAWVnNO4rx585CSkmK0aW0T+U2d5C+EwKRJk/D9999j48aNBuOoDf/ldDB9P3kbF0sF/rf6u/L/9136A/su/YG5vfthfNwdMkZomo8OHcDbP/0ArRCQoIOACutPd8OY1sfxcufVUOd+juM33kSH0MFyh2oyoSuAuPEMUP49KkcyFQA0gCIE8PsXJLfOMkdIRI1VdnZ2jYmpxnpBACAgIMBgb0deXh6A//WI1EYIgcmTJ2P16tVITU3Fvffea2bUnBPiMC4VFKDVv978WwIC6CcjAq98txtn8/PsH6AZ0o7+hjf2fg/tX92E4q8vbAEJa7M64O3fusFDWY4w7XO4XlRzybMjEkIHkf8oUL73rxIdgL92x9Vdg8gbD6E5K1d4ROSkqoZjrDkAwMfHp8ZRWxISExODEydOQKPR3+U7MzMTABAdXfueV1UJyMqVK7F8+XKMGzfOovtnEuIAfr54Hr1SP0HlH4exLrD/la/N/M0eYVlEo9Nh4fe7YHzKtoTlv8eiqFwFd6UGpy99ZM/wLFf+I1DxCwBDDwLUASiHKFlh56CIyOnJtG17UlISiouLsXHjRr3y1NRUhIaGonv37sZDFgKPPPIIVq5ciY8//hgTJkywLAhwOEZ25Votkv+7EaZtvltZJ/PqFZvGZI3frl5BSYUWtd+PAp9ntcOj7X+Dv+Ine4VmFVH2FQAXGE5CUFleugXwfc2OURERWWbw4MFISEjAlClTUFhYiMjISKSlpWH79u1YvXp19R4hkyZNQmpqKrKyshAeHg4AeOqpp7BixQpMnDgRMTEx2LdvX/V1VSoVOnc2fWiaSYjMVhw+aHYbldJxlwXvPPOzSfXWn4vCYx1+g4vkmA/8q0FXiLpnj5VCCC0kyXH/fIjIwVi59bo1bTdt2oQ5c+Zg7ty5yMvLQ1RUFNLS0jB69OjqOlqtFlqtFrfua/rll18CAD799FN8+umnetcMDw/H+fPnTY6BSYjM0k+eMLOFwKDWbWwSS324eOOySfWKNCpodcD1iraItHFM9cIlHJXDZcZ6QgAomjEBISKzVD2Izpr2lvL29saSJUuwZMkSo3VWrVqFVatW6ZWZk2TUhXNCZKbWmNsTIOG+qA42iaU+tPbzNaleoOomXBRAcNAjNo6ofkieyag1AYECkueYWs4TERkg4wPsHAGTEJm1Cww0o7bAfe3aw9vNzWbxWOvxbvfClNlSz8Xux+HC+9EqqIdd4rKWpGwFeE0xctYFULYBPB+ya0xERM7OZknIrl27MHHiRERFRcHLywstWrTAvffei59/rnvOwKpVqyBJksHjyhXHnZRpiad73G1GbQlT7jQ+Y9kReKg80C1EgvGJqQJN3UoR2+x23Nl2oT1Ds5rkPQOSz6uAS4tbSt0Bz9GQ/NdCUjj2E5SJyPFIOusPZ2azOSEffvghcnNzMX36dHTo0AHXrl3D22+/jfj4eHz99dfo379/nddYuXIloqKi9MoCAgJsFbIs2gcGYVSHGHx+PLPOukNat0EbJ7j/daOexdDPXseJfBfoJyMCTd3KsHuEBk2CPpYrPItJkgR4jgI8kgHtOUCUAy63MfkgIsvJODHVEdgsCVm6dGmNJ+0lJiYiMjISCxcuNCkJiY6Oxp133mmrEB3GwgEJaO3fFAt/+M5onZFRHbBwwCA7RmWdrx58AUcuncTc3ZtxvbQC3q4aPHuHCoM6ToCk15PgfCRJAShbyx0GEZHTs1kSYuhRv97e3ujQoYPZj/pt6CRJwuQ7umJS5ztx4noONp88gcN//okKnRZ3hYVjYucuCPD0lDtMs8W1iMIX46LqrkhE1FhZseFYdXsnZtclugUFBTh8+LBJvSAAMGzYMFy7dg2+vr7o27cvXnnllTq3knVmkiShQ1AIOgSFyB0KERHZQX09RddZ2TUJmTp1KkpKSjBnzpxa6zVr1gxz5sxBfHw8fHx8kJmZicWLFyM+Ph4//vgjOnXqZLStWq2GWq3WKzP0iGMiIiKSl92SkJdffhlr1qzB+++/jy5dutRaNzExEYmJidU/9+7dG0OHDkVMTAzmzp2LzZs3G227aNEizJ8/v97iJiIishlOTLW9+fPnY8GCBXjttdfw5JNPWnSNiIgI3H333Xp71Bsye/ZsPPPMM3plhYWFCAsLs+h1DfmzqAivfLsLeWU30TEoBDN79oK7q2u9Xd8ehBD46vQpHLx8CT4qFYa1jUK7AHP2LCEiIqsJ1P1EiLraOzGbJyHz589HSkoKUlJS8OKLL1p1LSEEFIratzZRqVS1Pr7YGufz89H/M/198g9evoxVv/6C0R1jnGb1ysxvtmHDyeMABPzc1NAICUsP7kdsoIRP730A/l4t5Q6xURNCALo8QJIAqWnl0mAiogbIpknIq6++ipSUFLz00kuYN2+eVdc6d+4cfvzxRwwcOLCeojNPq3+9Xev5dccy4aF0xct9+tkpIsv0XPERcm4WYmLbTDwYeRxh3kUAgMy8QHxyshOS1n2Cb8b0hcrTse+jIRJCB5SmQZSsBLQXKwtdIgCvyYBHMpMRogaIE1Nt5O2338bcuXORmJiIoUOH1hhGiY+PB2D4McEDBw5E7969ERsbWz0x9Y033oAkSXj11VdtFbJBh7Kz8UD6f1DZ51X7l8DKXw/jxV594FJHb41cntq2BUrdRbzb/TAGh53VO9fB7zr+1TMDS452weZfD+P+LkFQuDn2SiSd7iZQ+ApQtg1AKQAXwLUT4JMChatzLQ0WQkAUvAiUbYLe50x7AaLwJaDiBOAzl4kIUUMjYOWckHqLRBY2S0KqHvW7fft2bN++vcb5qscCG3pMcExMDD7//HO89dZbKC0tRXBwMPr374+XX34Zbdu2tVXIBlUmIEBdCUiVradPYXi79rYLyEIanQ4nLn+PzkEFGHrb2RrnXf7Km6ZH/4wxGUNxf6vHIYK/d9gvPZ2uELjWHxC3rnzSAhWHgdx7ofNdCoWHPL1mFlHv+SsBAfT/Vfnr/0vXAO73AKp4OwdGRDbFiam2sWfPHpPqGXpM8Lvvvlv/AVmgriEYQ07n5togEutodTr8I+0zlOg88M/We6HVSXBRGP7ganQS/hGRBYgcoOJXwC3OvsGaKn/y3xKQWwmg4CnoVL9BobDrKnSLiZtrAbjA+JN6XSBuroXEJISIGhDHHDdwYjEhzeQOoYbtZ07jZG4OrpR6o51fntEEBACUCoF2vnlQaxVA+U92jNJ0Ol0hUHGkjloa4OanddRxIJpTMJ6AoPKc5pS9oiEie9HVw+HEmITUs4GtHO+ZIqt+PQxfl1IAEso0tfcM6ASg1irhpnDgT7baxOSo3PizeByOZMK2/BIflEfU0FRNTLXmcGZMQuqZI86hOHcjHwVaDwACW/9oDY3OeIwSgAPXQiBJANxq31RONpKpe7K42DSMeuU+BLX/dZQgeQyxVzRERHbBJKQWbgBMn3oskNTW8SakAoCvyh2AAi08i/Dv0x1RoVNAayAR0egkXC31RL9mFwHJF3Dtav9gTeHWEyZNFHYfbvNQ6ovkOQaQvGH4r6QLoGgKeIy0d1hEZGtVE1OtOZwYk5BanHzqWZi6KgaQ8OY9g20ZjsXubdceEgBXhRYXin0x6fvBKNa4QgigQieh4q+E5M+b3vhPVhvEBuYCTV5wyF4dAFAo3AFV3zpqeQLuI+wRTr2QXIIg+f8bUAT8VaJE9bxxRQikpp9BUjSVKzwispVGnoQ4x9IBGf00ahx6fL4axvcJqSx/8o6uUDjol/bYmE5I/fUwzhc3RZsm15GZF4S7vhyHoWFZ6BxwFVqhwM/XQnB/+BE8FfMboGwDycPBv8B9lwK5QwHtOQMnlYD/mjp313U0kmsHIGg3ULYDouIQAAmSW3dANQCSxL+qRNTwONe/0jIICQnB2eoeEYGaezhImN4tHs/c3VuW+EwR6OmJdSNHo6WPD04XBaJU64YI7zycLmiK/55vgwjPPLzT43v0DM0HlFGQmq6CJDn2R0OhUAIB24AmLwGKFgA8AMkP8EgGgvdC4dZR7hAtIklukDyGQuEzDwqfuZDc72ECQtSQsSeETFGZiAAZv/+OZ7dvgYskYcE9QzC4bTuZIzNNm4AA7Hl4Mr67cB4HL/8BnRC4q4U/egbuB7SuANwhuQ8E3Ho6fAJSRaFQAF4PVR5ERM5IB9NH/Y21d2JMQsw0oG1bHGn7TN0VHZBCktA34nb0jbj9llLH3pqdiIgaLiYhREREMuED7IiIiEgefHYMOZrzN/KRW3oTH+39DFkFpXBz0eHedlGYEv9PuUMjAEKXB2ivA4qmkFyC5A6HbiF0BYD2CqDwg+QSInc4RHXTCUCyIpHQMQmhejJz28fYcf46egScxeWypsgt98blm00gIOGtA3/io8OvYfnQYega3knuUOuk01wCipcC5cchdAUQUEDh1gnwmQOFMqDuCzgYnXovkP88gGt65cLlNkh+70ByjZUnMAvoNNlA0SKg/CAgSgGFL+AxCpLXOEgKf7nDs4jQXIQoehtQf42qmXrCtQukJk9Dcusmb3BEZBSTEAeRtGYhPPN/Qc+mXlCW3kS3ZhXoEvgb2vjkYtXpWKw+0xE3rpfin/9Jxy/T2sDL3YRnjchApysHricDuhPVZdJfB8qzIa5tgUbRAi7BO5xi6alOdwPIGQzAyNORtRchcu8H/NdAcnPQHWZvoSv5N1C04G+F14CS/4O4uQYIWAdJebvhxg5KaC5A5CYDogh6SwUqfoHIewjw+xCSez/Z4iOqFYdjSG6fHtqMivI8/K6NROZLadAVFaHFU49huV8nNPcowrJe29G09DJmvnIILk2aYFJQCdaNf1PusA3LGQjgitHTkgRI4hK0V++Cstl++8VlAZ2uDMjpDaCszroi/1Eg+LDD7jILAEL9bc0ERK9CPsSNJ4GALQ59H38nihb+lYD8/SnElWsfRcEsQPUDJJOfOURkT9bu9eHcSYhzbAjRwG3M/AFnbwbieoEC2qJiVOTm4Y9/LYMm/wZyyrzwwMZeWPLMHmhy86AtKsKvlx3zj013cxdqS0BupUA+dGW7bRuQtYregikJCABAlAAVB20ajrVE0Qd1V9KcBioO2T6YeiK0VwH1HtRMQKprACIfUDv4Z42okXLMb7NG5LMfXseZ4kCotUpIfv5oPm0KlAEB0OTm4s/3P0TJ2Yv4/d1PceFiOUJaeKD5tCkQPg46bl/0vMlVhQBEQYrtYqkPpRvMqi4qztgoEOsJXTGg+cWEmhJQ/rPN46k32mzU/ZugC6C9aI9oiMzXyHdMZRIis3//XoAKoaz+Z1TZ1E8/EXnv/6DJzYN7kB/+/e/2cGvqC5VCI2vMRokik6tKEiBEng2DqQ83zazvyFsXlptR18VmUdQ7yceESrq/nlBM5IB0wvrDiTEJkZkOCvz94XjKpn4IenCMXr2ICffDK6AJdFCgR4iJQwQOTsBd7hDqYOa8CJVjPkUZQOVzdaRAEyoKQNXD1tHUH2UbwCUCtf9ZKQBVgp0CIiJzMAmRWYWu8o9AuqVLWZN/A9c+S9Or98eq9Vh7OARRPtfx/rCX7Bqj6cyb5+zSxFHv4y8Kc/aZcHfopceSpIDk9WDdFZVxkFydZyt/SZIgNXkaxodkJMDzQUgujvtnQ42c0Fl/ODEmITLzU1YOYYi/fpPT5N/An+9/CE1uLpQBAWg+40koAwJQnFOAz2Ztx6stv4SrSiVnyMa5DzG5qk4ACq97bRhMPfB+3PS6fkttF0d98ZoEKGvZM0MRAqnp/9kvnnoiuQ+G5PMqAHdU9ogoUflPmwR4/BNSk5myxkdUK84JITl9Nuo1NHcvAiCguXEDf77/wf8SkGmPw+P2cPxj/r1o2swLRTmFSH6kEH/88YfcYRvmMx9VvSHG/l5U/Z1RBP1mv7gsJHncD8C37oo+b0Dh3svm8VhLktwgBXwKeM8CFLfs9CoFAF5TIQXtgOQSLF+AVpA8R0EK3luZjHhNgtTkeUhBe6DwnesU+9EQNVb82ykzXw9PjAjXIf1iEf4oVcGliTcACc2nTYFvkCfCvK6jv88RLPi8BYY+fBMhzVqgSZMmcodtkELhBV3gHiB3GCTcqNyNGFWTUCvrFJdL8A0/JWeYJpMkNyAwDeL6PwHkG6jhDgT8AIWrKZMjHYMkuUHyngh4T4QQ5QCUkKSG8buIpPAGPB+w6qnoRHanE7Bqrw8nn5jKJMQBPHvPq5h8swjj/jMfEc+OhCgvh5d/OZops/Fmz65o0mwL3Dyb47vv/0CTJk3g62vCb+cyUSiDgZAD0FWchKLoLaDiPAA3SF5DIXlPga+TfeFJykggeDdQ9iXEzY2A7s/KXoQmc6BQdZE7PKtIkpvcIRARd0wlR+Dr2QRfjn+r1jotW7a0UzTWU7hGAf7L5Q6jXkgKT8BzFCTPUXKHQkQNjYCVSUi9RSIL5/q1lIiIiBoM9oQQERHJhcMxZIk1vx3Bgu/3oEKrhdLFBU9374nH7nSuR4YLoYEo+w7QHAMkD0geQyG5NJc7LCKixkOng1W7Leuce58QJiFm+va3Ldh46r84VRyErgElCG9ShK3Zt+P1vd/hzb3f48ijU+Ht7tg7geo0V4Cb24Cbb+DWB3+J4jcg4An4rYbC3Tk2rNLpyoDyw0DFbwCkyk2pVH0huZiyO6jjEEIDqL+HqPgVKPkJwFFU7vQ5EPB5BQoXx1wRZYzQFQA310KUrgd0uYAiCJLHA4DnGEgK57qXWwlRCqh/BHSFgDIMcL3TqZ44TORobDonpLi4GDNmzEBoaCjc3d0RFxeHdevWmdQ2JycH48ePR2BgIDw9PdGjRw9kZGTYMlyjfvrpJyxYsADDn+iC1IwPEdKkDCMiTiJIKsBX/72Mc9t+QttzX6E8Pw+dPzHhSaUy0RWvhu5KW4hrvYGbiyCE1kBP3k3gxgjoSv4jR4gm02muQ3clFsiJBW6MB0reAUrehih8EeJaT+iuDIBOa2hZreMR6m8hrvWCuPEYUPIBgF8AVABQA+qvgGtdoCt6X+YoTSe0VyCu3wtR/B6g/QMQpYD2IkTx2xC5SRDa63KHaDYhBETJCoicnhA3noAonAWR90+I6wkQ5QfkDo+cWSPfrMymPSEjRozAwYMHsXjxYrRt2xZr167FmDFjoNPpMHbsWKPt1Go1BgwYgBs3bmDJkiUIDg7G0qVLkZiYiJ07d6JPnz62DFvPrFmz8Prrr+uVNX96Grxu74Si/3sLV3/PAQB8DQDK7xD20mzsu3gB8beF2y1GU+gKFgGlKyFE5b4dwP/+a1DRSxCqLpCUre0Snzl0mjzges86amUD17pD578JCjfH7dUR5Qcg8h9Dnd2xJe9D59IKCs+hdonLGuLG84DusqEzlclIwQuQ/FfYPS6rlHwEUfxuzXLtHxB54wH/tZDc4uwdFTUEjXxOiCSEbe5g69atGDp0aHXiUWXQoEE4duwYLl68CBcXw0/r/OCDDzB16lTs3bsXPXpUPkxLo9GgU6dO8Pb2xv79+82KpbCwEL6+vigoKICPj+kbS33yySd49NFHDZ5zCQmB9urVGuXNn3kKER07YO/Ex8yK0ZZ0Oh2QE2V+Q/d/QOFX+7JhOeiudAJQanJ9KfhI5TJbB6TLHQ1UHDa5vqLZ7zaMxnpCcx7i+qC6KwZmQKEMs31A9UDo8iFy7gJg7OnVCsC1KxQBn9kzLLIhS78zLHmNgYEToVRYvmePRleOndc/tWmstmSz4Zj09HR4e3sjOTlZr3zChAm4fPlyrYlEeno62rVrV52AAIBSqcS4ceNw4MABXLp0yVZh65n3yTNGzxlKQMKenQr38DDcrKiwZVjmyzN+H7VS76nXMOpD5S6fpicgACCKVtsmGCsJ7Z9mJSAAIHTFNoqmfgj1j6ZVLHzVtoHUp7LtuHXuVE06oGI/hPaKvSKihkQnrD+cmM2SkKNHj6J9+/ZQKvVHfGJjY6vP19a2qp6htseOHavHSI2buzgBAd1rxmHI+DcSobwtAoCEyKb+No3LbJptlrVzwKczCs0F8xuVOug8HZ35c1ZERbYNAqlHWhN/QSj/EUKobRtLfdFdB2C417ZmPSLzCKGz+nBmNktCcnNz4e9f88u4qiw3N9cmbdVqNQoLC2sclvj2agTCxyfDs2vt23OHjr4X7WP90ESphoukxeoRD1j0erZjYabs2r5+w6gPkrcFjcrqPYx64dIMMPdJJ7qaPXAORdnOxIoVQMVxm4ZSbxQhqL0npKpeUN11iEiPTVfH1LZ0ra5lbZa2XbRoEXx9ffWOsDDLxp4lCOiEhOYPjgI8PIzWu7xuM44fzsN9YccRG9wS7kpHW/k83LK5Sz4v1Xsk1lIoLdnHxDG3u5cU/oBrvHmNlI79RSeZ9TRhY3MsHIx7IgDXWiooALe7ILmE2CsiakiElUMxTj4x1WZJSEBAgMEei7y8PAAw2NNRH21nz56NgoICvSM727Iu7D4hF1FQocLFt/8PKK19HkLqi9/g4C9+2DjK+KofuSiavW1+I2VPKByxJwQAYOYk08A024RRDyTfeebVV3awUST1ozKx6m1CTTczek3kJSl8IDV51shZBQBXSE2et2dI1JA08iW6NktCYmJicOLECWg0+r/tZGZmAgCio40vm4yJiamuZ25blUoFHx+fGoclLl4qwqV3/oWKCxdNqr895UMcPHjQoteyNUky57MaBUXgKhtGY6VgcyZzSlA4cO+BpGwF+H5oWmWPUc6xMVbTJQBq27BPAjySICmcZya/5DUBks8rgCJA/4QyClLAGkiujp0ckgPT6aw/nJjNkpCkpCQUFxdj48aNeuWpqakIDQ1F9+7da2178uRJvRU0Go0Gq1evRvfu3REaGmqrsPVc2ROG8gs1e1FUUYZ/g3N3d0dIiGN2ySqa/V5rIvK/8jegaPaFvcKyiEKhAPyNT2zWq9vslI2jsZ7CYwDQdHPtlZTRUPg6x4oShcILCNgESEZ2RlV2gNTkBfsGVQ8kz9GQgr6D1PTfkPz+D1LAF1AE/heSq2mT14moJpslIYMHD0ZCQgKmTJmCTz75BLt378ajjz6K7du344033qjeI2TSpElQKpW4cOF/qx4mTpyIjh07Ijk5GWvXrsXOnTvxwAMP4NSpUzU2DrOl+Hj98XrX8DBEPDcFnWaMQWhP/aGKSZMm4eTJk7jtttvsFp+5FM1+h6T8HkDNnjxJqjyvaHaffAGaQeHm9teeGcbG6pMdfk+NWylU7SGFnAI8n4P+cFMA4PM6FIGb5ArNIgrXSEiB3wBeUwFFKCB5Ai6RkJq8BCkgDZLCkgnG8pMkV0iqeEjugyC5WrD3DtHfNfLhGJvOoNy0aRPmzJmDuXPnIi8vD1FRUUhLS8Po0aOr62i1Wmi1Wty6Z5pKpUJGRgZmzpyJadOm4ebNm4iLi8O2bdvsulvqI488AgBYtWoVoqKisGLFLbs8PjYPkyZNwvHjxzFx4sTquo5OERQCwHm+nOuiaGaf5dr2IEkSJJ9HAR/DG+Q5G8klAFKT6UCT6XKHQuSwhE4HIVk+pOLsS3RttmOqI7HH7ndERNQw2HPH1P6eo6GUrNgxVZRj1811Tvv95mhrSYmIiBoPIWDxXk7V7Z0XkxAiIiK56AQgNd4kxKablREREREZw54QIiIiuQgBwIrJpewJISIiIksInbD6sFRxcTFmzJiB0NBQuLu7Iy4uDuvWrTOpbU5ODsaPH4/AwEB4enqiR48eyMjIMDsG9oQQERE1QiNGjMDBgwexePFitG3bFmvXrsWYMWOg0+kwdqzxR5Co1WoMGDAAN27cwJIlSxAcHIylS5ciMTERO3fuNGsrDSYhREREchE6WDccY1nbrVu3YseOHdWJBwD069cPFy5cwPPPP49Ro0ZVbyr6dytWrMDRo0exd+9e9OjRo7ptp06dMHPmTL3dzuvC4RgiIiKZyDUck56eDm9vbyQnJ+uVT5gwAZcvX641kUhPT0e7du2qExAAUCqVGDduHA4cOIBLly6ZHAeTECIiIrkInfWHBY4ePYr27dtDqdQfEImNja0+X1vbqnqG2h47ZvpO1o1iOKZqU9jCwkKZIyEiIkdX9V1hjw3FNaiwaq8yDSoAGP5+U6lUUKlUBtvl5uaiVatWNcr9/f2rzxuTm5tbXc/ctn/XKJKQoqIiAEBYWJjMkRARkbMoKiqCr6+vTa7t5uaGZs2a4YcrW62+lre3t8Hvt3nz5iElJcVoO0mSLDpnbdtbNYokJDQ0FNnZ2WjSpIlZb87fFRYWIiwsDNnZ2U65R399aOzvQWO/f4DvAe+/4d+/EAJFRUUIDQ212Wu4u7vj3LlzKC8vt/paQgiD323GekEAICAgwGCPRV5eHgAY7Omoj7Z/1yiSEIVCgZYtW9bb9Xx8fBrsXz5TNfb3oLHfP8D3gPffsO/fVj0gt3J3d4e7u7vNX8eQmJgYpKWlQaPR6M0LyczMBABER0fX2raq3q1Maft3nJhKRETUyCQlJaG4uBgbN27UK09NTUVoaCi6d+9ea9uTJ0/qraDRaDRYvXo1unfvblYPUqPoCSEiIqL/GTx4MBISEjBlyhQUFhYiMjISaWlp2L59O1avXl29R8ikSZOQmpqKrKwshIeHAwAmTpyIpUuXIjk5GYsXL0ZwcDA++OADnDp1Cjt37jQrDiYhREREjdCmTZswZ84czJ07F3l5eYiKikJaWhpGjx5dXUer1UKr1eqtFFKpVMjIyMDMmTMxbdo03Lx5E3Fxcdi2bZtZu6UCTELMolKpMG/evFon+zR0jf09aOz3D/A94P037vtvSLy9vbFkyRIsWbLEaJ1Vq1Zh1apVNcpDQkKQmppqdQySsMdCaCIiIqK/4cRUIiIikgWTECIiIpIFkxAiIiKSBZMQAMXFxZgxYwZCQ0Ph7u6OuLg4rFu3zqS2OTk5GD9+PAIDA+Hp6YkePXogIyPDxhHXr127dmHixImIioqCl5cXWrRogXvvvRc///xznW1XrVoFSZIMHleuXLFD9Nbbs2eP0XvYt29fne0bwmdg/PjxRt+Dut4HZ/sMFBUVYebMmRg0aBCCgoIgSZLRra0PHz6MgQMHwtvbG35+fhgxYgTOnj1r8mvt3LkTPXr0gKenJwIDAzF+/Hjk5OTU051YzpT3QKvV4p133kFiYiJatmwJT09PtG/fHrNmzcKNGzdMep2+ffsa/FwkJibW/02RU+LqGAAjRozAwYMHsXjxYrRt2xZr167FmDFjoNPpMHbsWKPt1Go1BgwYgBs3bmDJkiUIDg7G0qVLkZiYiJ07d5q9VEkuH374IXJzczF9+nR06NAB165dw9tvv434+Hh8/fXX6N+/f53XWLlyJaKiovTKAgICbBWyTSxcuBD9+vXTK6tr57+G8hl4+eWX8fjjj9coHz58OFQqFbp27VrnNZzlM5Cbm4tly5ahU6dOuO+++7B8+XKD9U6ePIm+ffsiLi4O//nPf1BWVoa5c+eiV69eOHLkCIKCgmp9nW+//RaDBw/G0KFDsXnzZuTk5OCFF17AgAEDcOjQIVlXl5jyHpSWliIlJQVjxozB5MmTERgYiMOHD2PBggX48ssvcejQIXh4eNT5Wq1atcKaNWv0yvz8/OrrVsjZiUbuq6++EgDE2rVr9coTEhJEaGio0Gg0RtsuXbpUABB79+6tLquoqBAdOnQQ3bp1s1nM9e3q1as1yoqKikRISIgYMGBArW1XrlwpAIiDBw/aKjyb2717twAg1q9fb3bbhvIZMGTPnj0CgHjppZdqredsnwGdTid0Op0QQohr164JAGLevHk16iUnJ4vAwEBRUFBQXXb+/Hnh6uoqZs6cWefrdO3aVXTo0EFUVFRUl/34448CgPjggw+svxErmPIeaDQacf369Rpt169fLwCIzz77rM7X6dOnj+jYsWO9xEwNU6MfjklPT4e3tzeSk5P1yidMmIDLly/rbUtrqG27du3Qo0eP6jKlUolx48bhwIEDuHTpks3irk/BwcE1yry9vdGhQwdkZ2fLEJHzaCifAUNWrFgBSZIwceJEuUOpV1VDArXRaDTYsmULRo4cqfd8lPDwcPTr1w/p6em1tr906RIOHjyIBx98UO+5HD179kTbtm3rbG9rprwHLi4uBnuyunXrBgD8t4HqRaNPQo4ePYr27dvr/UMBALGxsdXna2tbVc9Q22PHjtVjpPZVUFCAw4cPo2PHjibVHzZsGFxcXODv748RI0bU+r45qqlTp0KpVMLHxwf33HMPfvjhhzrbNNTPQEFBATZs2IABAwbg9ttvN6lNQ/gMVMnKykJpaanRP9szZ86grKzMaPuqezfW3pnfm127dgGAyf82ZGVlwd/fH0qlEq1bt8acOXNQWlpqyxDJiTT6OSG5ublo1apVjfKqRxEbelzxrW0NPbLYlLaOburUqSgpKcGcOXNqrdesWTPMmTMH8fHx8PHxQWZmJhYvXoz4+Hj8+OOP6NSpk50itpyvry+mT5+Ovn37IiAgAGfOnMGbb76Jvn374quvvsI999xjtG1D/QykpaWhtLQUkyZNqrNuQ/gM/F3Vn5uxP1shBPLz89G8eXOL2jvr5+LSpUuYNWsW7rzzTgwbNqzO+nfffTdGjRqFqKgolJaWYtu2bXjjjTfwww8/YPfu3VAoGv3vwY1eo09CANTaLVlXl6U1bR3Vyy+/jDVr1uD9999Hly5daq2bmJioN9O9d+/eGDp0KGJiYjB37lxs3rzZ1uFarXPnzujcuXP1z7169UJSUhJiYmIwc+bMWpMQoGF+BlasWIGAgAAkJSXVWbchfAaMsfbP1lgdZ/xc5OXlYciQIRBC4PPPPzcpgViwYIHez0OGDEFERASee+45bN682aTPFzVsjT4NDQgIMPhbSV5eHgDDv8nUR1tHNX/+fCxYsACvvfYannzySYuuERERgbvvvtuk5a2Oys/PD8OGDcNvv/1Wa9dxQ/wM/Pbbbzh06BDGjRtn8QoOZ/8MVM2FMPZnK0lSrSs86mrvbJ+L/Px8JCQk4NKlS9ixY4fB3mNTjRs3DgCc9rNB9avRJyExMTE4ceIENBqNXnlmZiaA2pdoxsTEVNczt60jmj9/PlJSUpCSkoIXX3zRqmsJIZy+q1X89Vil2n5rbWifAaCyFwQAJk+ebNV1nPkz0Lp1a3h4eBj9s42MjIS7u7vR9lV/7sbaO9PnIj8/HwMHDsS5c+ewY8cOg/NcLOGsnw2qX43+U5CUlITi4mJs3LhRrzw1NRWhoaHo3r17rW1Pnjypt4JGo9Fg9erV6N69O0JDQ20Wd3179dVXkZKSgpdeegnz5s2z6lrnzp3Djz/+iPj4+HqKzv7y8/OxZcsWxMXF1fpl05A+A0DlvierV69Gt27drPqidPbPgFKpxPDhw7Fp0yYUFRVVl1+8eBG7d+/GiBEjam3fokULdOvWDatXr4ZWq60u37dvH06dOlVne0dRlYCcPXsW33zzjd6wpaWqnrzqrJ8NqmeyLhB2EAkJCaJp06Zi2bJlYteuXeKRRx4RAMTq1aur60ycOFG4uLiI8+fPV5eVlZWJjh07irCwMLFmzRqxY8cOkZSUJJRKpdizZ48ct2KRt956SwAQiYmJ4qeffqpxVDH0HgwYMEDMnz9fpKeni4yMDPHee++J0NBQ0aRJE5GZmSnH7ZhtzJgx4oUXXhDr168Xu3fvFsuWLRPt2rUTSqVS7Nixo7peQ/4MVFm3bp0AIJYtW2bwfEP5DGzdulWsX79efPrppwKASE5OFuvXrxfr168XJSUlQgghTpw4Iby9vUXv3r3F1q1bxaZNm0R0dLQIDQ0VOTk5etdzcXER/fv31yvbvXu3UCqVIikpSezYsUOsWbNGhIWFiejoaFFWVma3ezWmrvfg5s2bomvXrkKSJLFkyZIa/y6cOXNG73p/fw++++47cc8994iPPvpIfPPNN+KLL74QU6ZMqa6n1WrtfcvkgJiEiMqNuZ566inRrFkz4ebmJmJjY0VaWppenYcfflgAEOfOndMrv3LlinjooYeEv7+/cHd3F/Hx8XpfXM6gT58+AoDRo4qh92DGjBmiQ4cOokmTJkKpVIrQ0FAxbtw4cerUKRnuxDKLFi0ScXFxwtfXV7i4uIigoCCRlJQkDhw4oFevIX8GqiQkJAgvLy9RWFho8HxD+QyEh4cb/bzfem+HDh0SAwYMEJ6ensLHx0fcd999Nb58hRACgOjTp0+N8m+++UbEx8cLd3d34e/vLx566CGDmwPKoa734Ny5c7X+u/Dwww/rXe/v78Hp06fFkCFDRIsWLYRKpRLu7u4iJiZGvPbaaw6RhJFjkIT4a+CbiIiIyI4a/ZwQIiIikgeTECIiIpIFkxAiIiKSBZMQIiIikgWTECIiIpIFkxAiIiKSBZMQIiIikgWTECIiIpIFkxAiIiKSBZMQIiIikgWTECIiIpIFkxAiIiKSxf8DFoD9wUnTSvsAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "oldphase1, oldphase2, phased1, phased2 = return_dict['initial_phase_res']\n",
    "phase_df = return_dict['phase_df']\n",
    "phased1 = phase_df['final_step_phase1'].values\n",
    "phased2 = phase_df['final_step_phase2'].values\n",
    "\n",
    "plt.scatter(oldphase1, oldphase2, c=1 + phased1.astype(int) + 2 * phased2.astype(int))\n",
    "indices = phase_df.index[phase_df['allele1'] == 'N/A']\n",
    "plt.scatter(oldphase1[indices], oldphase2[indices], c = 'black', marker = 'x')\n",
    "plt.title('Final phasing')\n",
    "plt.colorbar()\n",
    "#     plt.xlim((-25, 200))\n",
    "#     plt.ylim((-25, 200))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unphased samples: 13\n",
      "Percentage agreed with QUILT phasing for DRB1:  97.4559686888454 %\n"
     ]
    }
   ],
   "source": [
    "quilt_phase = pd.read_csv('/well/band/users/rbx225/recyclable_files/hla/1KG_phased_' + gene + '.tsv', sep = '\\t')\n",
    "\n",
    "phase_df = return_dict['phase_df']\n",
    "phased_indices = phase_df.index[(phase_df['final_step_phase1'] | phase_df['final_step_phase2'])]\n",
    "print('Number of unphased samples:', len(phase_df) - len(phased_indices))\n",
    "\n",
    "df1 = phase_df.loc[phased_indices,:]\n",
    "df2 = quilt_phase.loc[phased_indices,:]\n",
    "\n",
    "matched_haps = df1['final_step_phase1'] == df2['phase1_final']\n",
    "percet_matched_haps = (matched_haps.sum() / len(df1)) * 100\n",
    "print(f'Percentage agreed with QUILT phasing for {gene}: ', percet_matched_haps, '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Phasing commences\n",
      ">>> Parameters for this run:\n",
      ">>> - gene: A\n",
      ">>> - db directory: /well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/\n",
      ">>> - vcf file: /well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\n",
      ">>> - strict filtering on SNPs: False\n",
      ">>> - read from QUILT: True\n",
      ">>> - subset samples: True\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "index 2549 is out of bounds for axis 0 with size 5",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 27\u001b[0m\n\u001b[1;32m     24\u001b[0m sample_linker \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/well/band/users/rbx225/GAMCC/\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m SAMPLE_LINKER_FILE)\n\u001b[1;32m     25\u001b[0m sample_linker \u001b[38;5;241m=\u001b[39m {k:v \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(sample_linker[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSample_Name\u001b[39m\u001b[38;5;124m'\u001b[39m], sample_linker[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChip_Name\u001b[39m\u001b[38;5;124m'\u001b[39m])}\n\u001b[0;32m---> 27\u001b[0m return_dict \u001b[38;5;241m=\u001b[39m \u001b[43mphase_hla_on_haplotypes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mipd_gen_file_dir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mipd_gen_file_dir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mhla_gene_information\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mhla_gene_information\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     30\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mhlatypes\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mhlatypes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mphased_vcf\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mphased_vcf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mreference_allele_ary\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mreference_allele_ary\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mstrict_snp_filter\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mread_from_QUILT\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     35\u001b[0m \u001b[43m                            \u001b[49m\u001b[43msubset_vcf_samples\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msubset_vcf_samples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[43m                            \u001b[49m\u001b[43msample_linker\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msample_linker\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     38\u001b[0m hlatypes \u001b[38;5;241m=\u001b[39m return_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhlatypes\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     40\u001b[0m individual \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGAM951638\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "File \u001b[0;32m/gpfs3/well/band/users/rbx225/software/QUILT_sus/QUILT/Python/hla_phase.py:436\u001b[0m, in \u001b[0;36mphase_hla_on_haplotypes\u001b[0;34m(gene, ipd_gen_file_dir, hla_gene_information, hlatypes, phased_vcf, reference_allele_ary, strict_snp_filter, read_from_QUILT, subset_vcf_samples, sample_linker)\u001b[0m\n\u001b[1;32m    434\u001b[0m     target_haplotype \u001b[38;5;241m=\u001b[39m haps\u001b[38;5;241m.\u001b[39mloc[substituted_indices, closest_allele]\n\u001b[1;32m    435\u001b[0m     substitutable \u001b[38;5;241m=\u001b[39m target_haplotype[target_haplotype\u001b[38;5;241m.\u001b[39mvalues \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m--> 436\u001b[0m     \u001b[43msubstituted_nucleotides\u001b[49m\u001b[43m[\u001b[49m\u001b[43msubstitutable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;241m=\u001b[39m substitutable\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m    437\u001b[0m     closest_allele_idx \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    439\u001b[0m haps\u001b[38;5;241m.\u001b[39mloc[substituted_indices, a] \u001b[38;5;241m=\u001b[39m substituted_nucleotides\n",
      "\u001b[0;31mIndexError\u001b[0m: index 2549 is out of bounds for axis 0 with size 5"
     ]
    }
   ],
   "source": [
    "# For GAMCC\n",
    "gene = 'A'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "\n",
    "subset_vcf_samples = lcwgsus.read_tsv_as_lst('/well/band/users/rbx225/GAMCC/data/sample_tsvs/fv_gm_names.tsv')\n",
    "subset_vcf_samples = ','.join(subset_vcf_samples)\n",
    "\n",
    "sample_linker = pd.read_csv('/well/band/users/rbx225/GAMCC/' + SAMPLE_LINKER_FILE)\n",
    "sample_linker = {k:v for k, v in zip(sample_linker['Sample_Name'], sample_linker['Chip_Name'])}\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information, \n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            strict_snp_filter = False,\n",
    "                            read_from_QUILT = True, \n",
    "                            subset_vcf_samples = subset_vcf_samples,\n",
    "                            sample_linker = sample_linker)\n",
    "\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "individual = 'GAM951638'\n",
    "ix = hlatypes.index[hlatypes['Sample ID'] == individual][0]\n",
    "display_indices = np.arange(10)\n",
    "html = 'GAMCC-DRB1.html'\n",
    "\n",
    "res = visualise_phase(gene, ix, hlatypes, return_dict, both_het = True)\n",
    "compare_phase(display_indices, res, save_html = True, save_name = html)\n",
    "\n",
    "plot_first_step_phase(return_dict)\n",
    "plot_last_step_phase(return_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'phase_trios' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m trios[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallele2\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mN/A\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     10\u001b[0m undetermined_trios \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m---> 12\u001b[0m trios \u001b[38;5;241m=\u001b[39m trios\u001b[38;5;241m.\u001b[39mapply(func \u001b[38;5;241m=\u001b[39m \u001b[43mphase_trios\u001b[49m, args \u001b[38;5;241m=\u001b[39m (gene, hlatypes), axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m     14\u001b[0m kgp_samples_idx \u001b[38;5;241m=\u001b[39m hlatypes\u001b[38;5;241m.\u001b[39mindex[hlatypes[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSample ID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(trios[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIndividual ID\u001b[39m\u001b[38;5;124m'\u001b[39m])]\u001b[38;5;241m.\u001b[39mto_numpy()\n\u001b[1;32m     15\u001b[0m kgp_samples_phase \u001b[38;5;241m=\u001b[39m phased1[kgp_samples_idx]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'phase_trios' is not defined"
     ]
    }
   ],
   "source": [
    "# Checking with QUILT phasing\n",
    "retained_samples = hlatypes['Sample ID'].tolist()\n",
    "trios = pd.read_csv('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/integrated_call_samples_v3.20200731.ALL.ped', sep = '\\t')\n",
    "trios = trios[trios['Individual ID'].isin(retained_samples)]\n",
    "trios = trios[(trios['Paternal ID'].isin(retained_samples)) | (trios['Maternal ID'].isin(retained_samples))]\n",
    "trios = trios[['Individual ID', 'Paternal ID', 'Maternal ID']]\n",
    "\n",
    "trios['allele1'] = 'N/A'\n",
    "trios['allele2'] = 'N/A'\n",
    "undetermined_trios = []\n",
    "\n",
    "trios = trios.apply(func = phase_trios, args = (gene, hlatypes), axis = 1)\n",
    "\n",
    "kgp_samples_idx = hlatypes.index[hlatypes['Sample ID'].isin(trios['Individual ID'])].to_numpy()\n",
    "kgp_samples_phase = phased1[kgp_samples_idx]\n",
    "\n",
    "phased_res = pd.DataFrame({'Sample ID': trios['Individual ID'].tolist(), 'allele1': reftypes1[kgp_samples_idx], 'allele2': reftypes2[kgp_samples_idx]})\n",
    "\n",
    "for i in range(len(phased_res)):\n",
    "    if kgp_samples_phase[i] == False:\n",
    "        tmp = phased_res.iloc[i, 1]\n",
    "        phased_res.iloc[i, 1] = phased_res.iloc[i, 2]\n",
    "        phased_res.iloc[i, 2] = tmp\n",
    "\n",
    "res = pd.merge(phased_res, trios[['Individual ID', 'allele1', 'allele2']], left_on = 'Sample ID', right_on = 'Individual ID', suffixes = ('_pred', '_true'))\n",
    "res = res.drop(columns = 'Individual ID')\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A-14:54-T', 'G-09:01-G', 'G-08:03-T', 'A-14:54-G', 'A-14:04-T', 'G-09:01-G']\n",
      "['A-11:04-T', 'G-03:01-G', 'A-11:04-T', 'G-12:01-G', 'A-03:01-G', 'G-14:54-T']\n",
      "['G-12:02-C', 'A-12:02-A', 'G-12:02-C', 'G-10:01-C', 'G-12:02-C', 'A-08:03-A']\n",
      "['A-03:01-T', 'G-01:01-A', 'A-03:01-T', 'G-15:01-A', 'G-01:01-A', 'G-01:01-A']\n",
      "['A-12:01-T', 'G-16:02-A', 'A-13:01-T', 'G-12:01-T', 'G-16:02-A', 'A-11:01-T']\n",
      "['G-07:01-G', 'A-03:01-A', 'G-07:01-G', 'G-15:01-A', 'G-16:02-A', 'A-03:01-A']\n",
      "['G-01:02-T', 'A-14:06-G', 'G-01:02-T', 'G-04:11-T', 'G-04:07-T', 'A-14:06-G']\n"
     ]
    }
   ],
   "source": [
    "# Checking 1KG trios\n",
    "\n",
    "full_start = return_dict['b38_start']\n",
    "full_end = return_dict['b38_end']\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "retained_samples = hlatypes['Sample ID'].tolist()\n",
    "trios = pd.read_csv('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/integrated_call_samples_v3.20200731.ALL.ped', sep = '\\t')\n",
    "trios = trios[trios['Individual ID'].isin(retained_samples)]\n",
    "trios = trios[(trios['Paternal ID'].isin(retained_samples)) & (trios['Maternal ID'].isin(retained_samples))]\n",
    "trios = trios[['Individual ID', 'Paternal ID', 'Maternal ID']]\n",
    "\n",
    "trio_samples_all = trios.values.ravel().tolist()\n",
    "\n",
    "df = read_vcf(\n",
    "        full_start - 5000, \n",
    "        full_end + 5000, \n",
    "        phased_vcf, \n",
    "        hlatypes, \n",
    "        False, \n",
    "        ','.join(trio_samples_all), \n",
    "        None)\n",
    "\n",
    "df = df[(df['ref'].str.len() == 1) & (df['alt'].str.len() == 1)]\n",
    "\n",
    "df_up = df[df['pos'] < full_start]\n",
    "df_down = df[df['pos'] > full_end]\n",
    "\n",
    "phase_df = return_dict['phase_df']\n",
    "phase_df = phase_df[phase_df['Sample'].isin(trio_samples_all)]\n",
    "\n",
    "trio_haps = {}\n",
    "\n",
    "for child in trios['Individual ID']:\n",
    "    mat_id, pat_id = trios[trios['Individual ID'] == child][['Maternal ID', 'Paternal ID']].values[0]\n",
    "    \n",
    "    child_alleles = phase_df[phase_df['Sample'] == child][['allele1', 'allele2']].values[0]\n",
    "    mat_alleles = phase_df[phase_df['Sample'] == mat_id][['allele1', 'allele2']].values[0]\n",
    "    pat_alleles = phase_df[phase_df['Sample'] == pat_id][['allele1', 'allele2']].values[0]\n",
    "    \n",
    "    df_up_child = df_up[['pos', 'ref', 'alt'] + [child, mat_id, pat_id]]\n",
    "    df_up_child = df_up_child[~df_up_child[child].str.contains(r'\\.')]\n",
    "    df_up_child['pat'] = df_up_child[child].str.split('|').str.get(0)\n",
    "    df_up_child['mat'] = df_up_child[child].str.split('|').str.get(1)\n",
    "    het_sites = (df_up_child['mat'] != df_up_child['pat'])\n",
    "    \n",
    "    upstream_idx = het_sites[het_sites].index[-1]\n",
    "    ref, alt, mat, pat = df_up_child.loc[upstream_idx, ['ref', 'alt', 'mat', 'pat']]\n",
    "    \n",
    "    mat_genotypes = df_up_child.loc[upstream_idx, mat_id].split('|')\n",
    "    pat_genotypes = df_up_child.loc[upstream_idx, pat_id].split('|')\n",
    "    \n",
    "    if mat_genotypes[0] == '0':\n",
    "        mat_hap1 = ref + '-' + mat_alleles[0] + '-'\n",
    "    elif mat_genotypes[0] == '1':\n",
    "        mat_hap1 = alt + '-' + mat_alleles[0] + '-'\n",
    "    else:\n",
    "        mat_hap1 = 'N' + '-' + mat_alleles[0] + '-'\n",
    "        \n",
    "    if mat_genotypes[1] == '0':\n",
    "        mat_hap2 = ref + '-' + mat_alleles[1] + '-'\n",
    "    elif mat_genotypes[1] == '1':\n",
    "        mat_hap2 = alt + '-' + mat_alleles[1] + '-'\n",
    "    else:\n",
    "        mat_hap2 = 'N' + '-' + mat_alleles[1] + '-'\n",
    "    \n",
    "    if pat_genotypes[0] == '0':\n",
    "        pat_hap1 = ref + '-' + pat_alleles[0] + '-'\n",
    "    elif pat_genotypes[0] == '1':\n",
    "        pat_hap1 = alt + '-' + pat_alleles[0] + '-'\n",
    "    else:\n",
    "        pat_hap1 = 'N' + '-' + pat_alleles[0] + '-'\n",
    "        \n",
    "    if pat_genotypes[1] == '0':\n",
    "        pat_hap2 = ref + '-' + pat_alleles[1] + '-'\n",
    "    elif pat_genotypes[1] == '1':\n",
    "        pat_hap2 = alt + '-' + pat_alleles[1] + '-'\n",
    "    else:\n",
    "        pat_hap2 = 'N' + '-' + pat_alleles[1] + '-'\n",
    "    \n",
    "    if pat == '0':\n",
    "        child_hap1 = ref + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = alt + '-' + child_alleles[1] + '-'\n",
    "    elif pat == '1':\n",
    "        child_hap1 = alt + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = ref + '-' + child_alleles[1] + '-'\n",
    "    else:\n",
    "        child_hap1 = 'N' + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = 'N' + '-' + child_alleles[1] + '-'\n",
    "    \n",
    "    df_down_child = df_down[['pos', 'ref', 'alt'] + [child, mat_id, pat_id]]\n",
    "    df_down_child = df_down_child[~df_down_child[child].str.contains(r'\\.')]\n",
    "    df_down_child['pat'] = df_down_child[child].str.split('|').str.get(0)\n",
    "    df_down_child['mat'] = df_down_child[child].str.split('|').str.get(1)\n",
    "    het_sites = (df_down_child['mat'] != df_down_child['pat'])\n",
    "    downstream_idx = het_sites[het_sites].index[0]\n",
    "    ref, alt, mat, pat = df_down_child.loc[downstream_idx, ['ref', 'alt', 'mat', 'pat']]\n",
    "    \n",
    "    mat_genotypes = df_down_child.loc[downstream_idx, mat_id].split('|')\n",
    "    pat_genotypes = df_down_child.loc[downstream_idx, pat_id].split('|')\n",
    "    \n",
    "    if mat_genotypes[0] == '0':\n",
    "        mat_hap1 = mat_hap1 + ref\n",
    "    elif mat_genotypes[0] == '1':\n",
    "        mat_hap1 = mat_hap1 + alt\n",
    "    else:\n",
    "        mat_hap1 = mat_hap1 + 'N'\n",
    "        \n",
    "    if mat_genotypes[1] == '0':\n",
    "        mat_hap2 = mat_hap2 + ref\n",
    "    elif mat_genotypes[1] == '1':\n",
    "        mat_hap2 = mat_hap2 + alt\n",
    "    else:\n",
    "        mat_hap2 = mat_hap2 + 'N'\n",
    "    \n",
    "    if pat_genotypes[0] == '0':\n",
    "        pat_hap1 = pat_hap1 + ref\n",
    "    elif pat_genotypes[0] == '1':\n",
    "        pat_hap1 = pat_hap1 + alt\n",
    "    else:\n",
    "        pat_hap1 = pat_hap1 + 'N'\n",
    "        \n",
    "    if pat_genotypes[1] == '0':\n",
    "        pat_hap2 = pat_hap2 + ref\n",
    "    elif pat_genotypes[1] == '1':\n",
    "        pat_hap2 = pat_hap2 + alt\n",
    "    else:\n",
    "        pat_hap2 = pat_hap2 + 'N'\n",
    "    \n",
    "    if pat == '0':\n",
    "        child_hap1 = child_hap1 + ref\n",
    "        child_hap2 = child_hap2 + alt\n",
    "    elif pat == '1':\n",
    "        child_hap1 = child_hap1 + alt\n",
    "        child_hap2 = child_hap2 + ref\n",
    "    else:\n",
    "        break\n",
    "    \n",
    "    trio_haps[child] = [child_hap1, child_hap2, pat_hap1, pat_hap2, mat_hap1, mat_hap2]\n",
    "    \n",
    "for v in trio_haps.values():\n",
    "    print(v)\n",
    "    \n",
    "# So after checking the phased vcfs it is always correct that (for these trios) the phasing designation is pat|mat\n",
    "# Given this information we don't need to check for het sites\n",
    "# However if we do check there are some alleles that don't perfectly match the snp haplotypes.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RETAIN THIS!\n",
    "# gamcc_hla = pd.read_csv('/well/band/users/rbx225/GAMCC/data/hla_direct_sequencing/HLA_direct_sequencing_all.csv')\n",
    "# gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Included Alleles']]\n",
    "# for s in gamcc_hla['SampleID'].unique():\n",
    "#     tmps = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "#     for l in HLA_GENES:\n",
    "#         tmpl = tmps[tmps['Locus'] == l]\n",
    "#         repeat = 2 - tmpl.shape[0]\n",
    "#         if repeat == 2:\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, np.nan]\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, np.nan]\n",
    "#         if repeat == 1:\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, tmpl.iloc[0,2]]\n",
    "                \n",
    "# gamcc_hla = gamcc_hla[gamcc_hla['Locus'] == gene]\n",
    "# gamcc_hla = gamcc_hla[['SampleID', 'Included Alleles']]\n",
    "\n",
    "# gamcc_hla = pd.merge(gamcc_hla[::2], gamcc_hla[1::2], on = 'SampleID')\n",
    "# gamcc_hla.columns = ['Sample ID', 'HLA-DRB1 1', 'HLA-DRB1 2']\n",
    "# gamcc_hla\n",
    "\n",
    "# x = gamcc_hla.iloc[0, 1]\n",
    "# indices = []\n",
    "# db_alleles = db.index.to_numpy()\n",
    "\n",
    "# for a in x.split('/'):\n",
    "#     old_a = a\n",
    "#     while (gene + '*' + a) not in db.index and ':' in a:\n",
    "#         a = a.rsplit(':', 1)[0]\n",
    "#     if ':' in a:\n",
    "#         indices.append(np.where(db_alleles == gene + '*' + a)[0][0])\n",
    "\n",
    "# x = db.iloc[np.array(indices), :]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
