{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import pickle\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "# import sqlite3\n",
    "# from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "# from plotnine import *\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "# from collections import Counter\n",
    "# import seaborn as sns\n",
    "# import matplotlib.colors as mcolors\n",
    "# from matplotlib.ticker import FuncFormatter\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "import psa\n",
    "import pywfa\n",
    "from IPython.display import display_html, IFrame\n",
    "# import patchworklib as pw\n",
    "sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "from hla_phase import *\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# Genuine logics changed from the main script were marked by string **X** in the original R files.\n",
    "# Simplicity changes were marked by string **Y**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For oneKG\n",
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "ref_samples = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/oneKG.samples', sep = ' ')\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "ref_samples_removed = ref_samples[~ref_samples['SAMPLE'].isin(hlatypes['Sample ID'].tolist())]\n",
    "samples_to_remove = ref_samples_removed['SAMPLE'].tolist()\n",
    "hlatypes = hlatypes[~hlatypes['Sample ID'].isin(samples_to_remove)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/oneKG.chr6.vcf.gz\"\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information,\n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            strict_snp_filter = True,\n",
    "                            read_from_QUILT = False, \n",
    "                            subset_vcf_samples = None,\n",
    "                            sample_linker = None)\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "individual = 'NA12878'\n",
    "ix = hlatypes.index[hlatypes['Sample ID'] == individual][0]\n",
    "display_indices = np.arange(10)\n",
    "html = '1KG-DRB1.html'\n",
    "\n",
    "res = visualise_phase(gene, ix, hlatypes, return_dict, both_het = True)\n",
    "compare_phase(display_indices, res, save_html = True, save_name = html)\n",
    "\n",
    "plot_first_step_phase(return_dict)\n",
    "plot_last_step_phase(return_dict)\n",
    "# reference_exclude_samplelist_file = ''\n",
    "# if reference_exclude_samplelist_file != '':\n",
    "#     samples_to_remove = samples_to_remove + lcwgsus.read_tsv_as_lst(reference_exclude_samplelist_file)\n",
    "# if samples_to_remove is not none, save to \"hlauntyped.exclude.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unphased samples: 13\n",
      "Percentage agreed with QUILT phasing for DRB1:  97.4559686888454 %\n"
     ]
    }
   ],
   "source": [
    "quilt_phase = pd.read_csv('/well/band/users/rbx225/recyclable_files/hla/1KG_phased_' + gene + '.tsv', sep = '\\t')\n",
    "\n",
    "phase_df = return_dict['phase_df']\n",
    "phased_indices = phase_df.index[(phase_df['final_step_phase1'] | phase_df['final_step_phase2'])]\n",
    "print('Number of unphased samples:', len(phase_df) - len(phased_indices))\n",
    "\n",
    "df1 = phase_df.loc[phased_indices,:]\n",
    "df2 = quilt_phase.loc[phased_indices,:]\n",
    "\n",
    "matched_haps = df1['final_step_phase1'] == df2['phase1_final']\n",
    "percet_matched_haps = (matched_haps.sum() / len(df1)) * 100\n",
    "print(f'Percentage agreed with QUILT phasing for {gene}: ', percet_matched_haps, '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> Phasing commences\n",
      ">>> Parameters for this run:\n",
      ">>> - gene: A\n",
      ">>> - db directory: /well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/\n",
      ">>> - vcf file: /well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\n",
      ">>> - strict filtering on SNPs: False\n",
      ">>> - read from QUILT: True\n",
      ">>> - subset samples: True\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "index 2549 is out of bounds for axis 0 with size 5",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 27\u001b[0m\n\u001b[1;32m     24\u001b[0m sample_linker \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/well/band/users/rbx225/GAMCC/\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m SAMPLE_LINKER_FILE)\n\u001b[1;32m     25\u001b[0m sample_linker \u001b[38;5;241m=\u001b[39m {k:v \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(sample_linker[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSample_Name\u001b[39m\u001b[38;5;124m'\u001b[39m], sample_linker[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChip_Name\u001b[39m\u001b[38;5;124m'\u001b[39m])}\n\u001b[0;32m---> 27\u001b[0m return_dict \u001b[38;5;241m=\u001b[39m \u001b[43mphase_hla_on_haplotypes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mipd_gen_file_dir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mipd_gen_file_dir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mhla_gene_information\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mhla_gene_information\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     30\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mhlatypes\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mhlatypes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mphased_vcf\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mphased_vcf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mreference_allele_ary\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mreference_allele_ary\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mstrict_snp_filter\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m                            \u001b[49m\u001b[43mread_from_QUILT\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     35\u001b[0m \u001b[43m                            \u001b[49m\u001b[43msubset_vcf_samples\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msubset_vcf_samples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[43m                            \u001b[49m\u001b[43msample_linker\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msample_linker\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     38\u001b[0m hlatypes \u001b[38;5;241m=\u001b[39m return_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhlatypes\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     40\u001b[0m individual \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGAM951638\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "File \u001b[0;32m/gpfs3/well/band/users/rbx225/software/QUILT_sus/QUILT/Python/hla_phase.py:436\u001b[0m, in \u001b[0;36mphase_hla_on_haplotypes\u001b[0;34m(gene, ipd_gen_file_dir, hla_gene_information, hlatypes, phased_vcf, reference_allele_ary, strict_snp_filter, read_from_QUILT, subset_vcf_samples, sample_linker)\u001b[0m\n\u001b[1;32m    434\u001b[0m     target_haplotype \u001b[38;5;241m=\u001b[39m haps\u001b[38;5;241m.\u001b[39mloc[substituted_indices, closest_allele]\n\u001b[1;32m    435\u001b[0m     substitutable \u001b[38;5;241m=\u001b[39m target_haplotype[target_haplotype\u001b[38;5;241m.\u001b[39mvalues \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m--> 436\u001b[0m     \u001b[43msubstituted_nucleotides\u001b[49m\u001b[43m[\u001b[49m\u001b[43msubstitutable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;241m=\u001b[39m substitutable\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m    437\u001b[0m     closest_allele_idx \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    439\u001b[0m haps\u001b[38;5;241m.\u001b[39mloc[substituted_indices, a] \u001b[38;5;241m=\u001b[39m substituted_nucleotides\n",
      "\u001b[0;31mIndexError\u001b[0m: index 2549 is out of bounds for axis 0 with size 5"
     ]
    }
   ],
   "source": [
    "# For GAMCC\n",
    "gene = 'A'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "\n",
    "subset_vcf_samples = lcwgsus.read_tsv_as_lst('/well/band/users/rbx225/GAMCC/data/sample_tsvs/fv_gm_names.tsv')\n",
    "subset_vcf_samples = ','.join(subset_vcf_samples)\n",
    "\n",
    "sample_linker = pd.read_csv('/well/band/users/rbx225/GAMCC/' + SAMPLE_LINKER_FILE)\n",
    "sample_linker = {k:v for k, v in zip(sample_linker['Sample_Name'], sample_linker['Chip_Name'])}\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information, \n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            strict_snp_filter = False,\n",
    "                            read_from_QUILT = True, \n",
    "                            subset_vcf_samples = subset_vcf_samples,\n",
    "                            sample_linker = sample_linker)\n",
    "\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "individual = 'GAM951638'\n",
    "ix = hlatypes.index[hlatypes['Sample ID'] == individual][0]\n",
    "display_indices = np.arange(10)\n",
    "html = 'GAMCC-DRB1.html'\n",
    "\n",
    "res = visualise_phase(gene, ix, hlatypes, return_dict, both_het = True)\n",
    "compare_phase(display_indices, res, save_html = True, save_name = html)\n",
    "\n",
    "plot_first_step_phase(return_dict)\n",
    "plot_last_step_phase(return_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'phase_trios' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m trios[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallele2\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mN/A\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     10\u001b[0m undetermined_trios \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m---> 12\u001b[0m trios \u001b[38;5;241m=\u001b[39m trios\u001b[38;5;241m.\u001b[39mapply(func \u001b[38;5;241m=\u001b[39m \u001b[43mphase_trios\u001b[49m, args \u001b[38;5;241m=\u001b[39m (gene, hlatypes), axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m     14\u001b[0m kgp_samples_idx \u001b[38;5;241m=\u001b[39m hlatypes\u001b[38;5;241m.\u001b[39mindex[hlatypes[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSample ID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(trios[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIndividual ID\u001b[39m\u001b[38;5;124m'\u001b[39m])]\u001b[38;5;241m.\u001b[39mto_numpy()\n\u001b[1;32m     15\u001b[0m kgp_samples_phase \u001b[38;5;241m=\u001b[39m phased1[kgp_samples_idx]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'phase_trios' is not defined"
     ]
    }
   ],
   "source": [
    "# Checking with QUILT phasing\n",
    "retained_samples = hlatypes['Sample ID'].tolist()\n",
    "trios = pd.read_csv('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/integrated_call_samples_v3.20200731.ALL.ped', sep = '\\t')\n",
    "trios = trios[trios['Individual ID'].isin(retained_samples)]\n",
    "trios = trios[(trios['Paternal ID'].isin(retained_samples)) | (trios['Maternal ID'].isin(retained_samples))]\n",
    "trios = trios[['Individual ID', 'Paternal ID', 'Maternal ID']]\n",
    "\n",
    "trios['allele1'] = 'N/A'\n",
    "trios['allele2'] = 'N/A'\n",
    "undetermined_trios = []\n",
    "\n",
    "trios = trios.apply(func = phase_trios, args = (gene, hlatypes), axis = 1)\n",
    "\n",
    "kgp_samples_idx = hlatypes.index[hlatypes['Sample ID'].isin(trios['Individual ID'])].to_numpy()\n",
    "kgp_samples_phase = phased1[kgp_samples_idx]\n",
    "\n",
    "phased_res = pd.DataFrame({'Sample ID': trios['Individual ID'].tolist(), 'allele1': reftypes1[kgp_samples_idx], 'allele2': reftypes2[kgp_samples_idx]})\n",
    "\n",
    "for i in range(len(phased_res)):\n",
    "    if kgp_samples_phase[i] == False:\n",
    "        tmp = phased_res.iloc[i, 1]\n",
    "        phased_res.iloc[i, 1] = phased_res.iloc[i, 2]\n",
    "        phased_res.iloc[i, 2] = tmp\n",
    "\n",
    "res = pd.merge(phased_res, trios[['Individual ID', 'allele1', 'allele2']], left_on = 'Sample ID', right_on = 'Individual ID', suffixes = ('_pred', '_true'))\n",
    "res = res.drop(columns = 'Individual ID')\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A-14:54-T', 'G-09:01-G', 'G-08:03-T', 'A-14:54-G', 'A-14:04-T', 'G-09:01-G']\n",
      "['A-11:04-T', 'G-03:01-G', 'A-11:04-T', 'G-12:01-G', 'A-03:01-G', 'G-14:54-T']\n",
      "['G-12:02-C', 'A-12:02-A', 'G-12:02-C', 'G-10:01-C', 'G-12:02-C', 'A-08:03-A']\n",
      "['A-03:01-T', 'G-01:01-A', 'A-03:01-T', 'G-15:01-A', 'G-01:01-A', 'G-01:01-A']\n",
      "['A-12:01-T', 'G-16:02-A', 'A-13:01-T', 'G-12:01-T', 'G-16:02-A', 'A-11:01-T']\n",
      "['G-07:01-G', 'A-03:01-A', 'G-07:01-G', 'G-15:01-A', 'G-16:02-A', 'A-03:01-A']\n",
      "['G-01:02-T', 'A-14:06-G', 'G-01:02-T', 'G-04:11-T', 'G-04:07-T', 'A-14:06-G']\n"
     ]
    }
   ],
   "source": [
    "# Checking 1KG trios\n",
    "\n",
    "full_start = return_dict['b38_start']\n",
    "full_end = return_dict['b38_end']\n",
    "hlatypes = return_dict['hlatypes']\n",
    "\n",
    "retained_samples = hlatypes['Sample ID'].tolist()\n",
    "trios = pd.read_csv('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/integrated_call_samples_v3.20200731.ALL.ped', sep = '\\t')\n",
    "trios = trios[trios['Individual ID'].isin(retained_samples)]\n",
    "trios = trios[(trios['Paternal ID'].isin(retained_samples)) & (trios['Maternal ID'].isin(retained_samples))]\n",
    "trios = trios[['Individual ID', 'Paternal ID', 'Maternal ID']]\n",
    "\n",
    "trio_samples_all = trios.values.ravel().tolist()\n",
    "\n",
    "df = read_vcf(\n",
    "        full_start - 5000, \n",
    "        full_end + 5000, \n",
    "        phased_vcf, \n",
    "        hlatypes, \n",
    "        False, \n",
    "        ','.join(trio_samples_all), \n",
    "        None)\n",
    "\n",
    "df = df[(df['ref'].str.len() == 1) & (df['alt'].str.len() == 1)]\n",
    "\n",
    "df_up = df[df['pos'] < full_start]\n",
    "df_down = df[df['pos'] > full_end]\n",
    "\n",
    "phase_df = return_dict['phase_df']\n",
    "phase_df = phase_df[phase_df['Sample'].isin(trio_samples_all)]\n",
    "\n",
    "trio_haps = {}\n",
    "\n",
    "for child in trios['Individual ID']:\n",
    "    mat_id, pat_id = trios[trios['Individual ID'] == child][['Maternal ID', 'Paternal ID']].values[0]\n",
    "    \n",
    "    child_alleles = phase_df[phase_df['Sample'] == child][['allele1', 'allele2']].values[0]\n",
    "    mat_alleles = phase_df[phase_df['Sample'] == mat_id][['allele1', 'allele2']].values[0]\n",
    "    pat_alleles = phase_df[phase_df['Sample'] == pat_id][['allele1', 'allele2']].values[0]\n",
    "    \n",
    "    df_up_child = df_up[['pos', 'ref', 'alt'] + [child, mat_id, pat_id]]\n",
    "    df_up_child = df_up_child[~df_up_child[child].str.contains(r'\\.')]\n",
    "    df_up_child['pat'] = df_up_child[child].str.split('|').str.get(0)\n",
    "    df_up_child['mat'] = df_up_child[child].str.split('|').str.get(1)\n",
    "    het_sites = (df_up_child['mat'] != df_up_child['pat'])\n",
    "    \n",
    "    upstream_idx = het_sites[het_sites].index[-1]\n",
    "    ref, alt, mat, pat = df_up_child.loc[upstream_idx, ['ref', 'alt', 'mat', 'pat']]\n",
    "    \n",
    "    mat_genotypes = df_up_child.loc[upstream_idx, mat_id].split('|')\n",
    "    pat_genotypes = df_up_child.loc[upstream_idx, pat_id].split('|')\n",
    "    \n",
    "    if mat_genotypes[0] == '0':\n",
    "        mat_hap1 = ref + '-' + mat_alleles[0] + '-'\n",
    "    elif mat_genotypes[0] == '1':\n",
    "        mat_hap1 = alt + '-' + mat_alleles[0] + '-'\n",
    "    else:\n",
    "        mat_hap1 = 'N' + '-' + mat_alleles[0] + '-'\n",
    "        \n",
    "    if mat_genotypes[1] == '0':\n",
    "        mat_hap2 = ref + '-' + mat_alleles[1] + '-'\n",
    "    elif mat_genotypes[1] == '1':\n",
    "        mat_hap2 = alt + '-' + mat_alleles[1] + '-'\n",
    "    else:\n",
    "        mat_hap2 = 'N' + '-' + mat_alleles[1] + '-'\n",
    "    \n",
    "    if pat_genotypes[0] == '0':\n",
    "        pat_hap1 = ref + '-' + pat_alleles[0] + '-'\n",
    "    elif pat_genotypes[0] == '1':\n",
    "        pat_hap1 = alt + '-' + pat_alleles[0] + '-'\n",
    "    else:\n",
    "        pat_hap1 = 'N' + '-' + pat_alleles[0] + '-'\n",
    "        \n",
    "    if pat_genotypes[1] == '0':\n",
    "        pat_hap2 = ref + '-' + pat_alleles[1] + '-'\n",
    "    elif pat_genotypes[1] == '1':\n",
    "        pat_hap2 = alt + '-' + pat_alleles[1] + '-'\n",
    "    else:\n",
    "        pat_hap2 = 'N' + '-' + pat_alleles[1] + '-'\n",
    "    \n",
    "    if pat == '0':\n",
    "        child_hap1 = ref + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = alt + '-' + child_alleles[1] + '-'\n",
    "    elif pat == '1':\n",
    "        child_hap1 = alt + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = ref + '-' + child_alleles[1] + '-'\n",
    "    else:\n",
    "        child_hap1 = 'N' + '-' + child_alleles[0] + '-'\n",
    "        child_hap2 = 'N' + '-' + child_alleles[1] + '-'\n",
    "    \n",
    "    df_down_child = df_down[['pos', 'ref', 'alt'] + [child, mat_id, pat_id]]\n",
    "    df_down_child = df_down_child[~df_down_child[child].str.contains(r'\\.')]\n",
    "    df_down_child['pat'] = df_down_child[child].str.split('|').str.get(0)\n",
    "    df_down_child['mat'] = df_down_child[child].str.split('|').str.get(1)\n",
    "    het_sites = (df_down_child['mat'] != df_down_child['pat'])\n",
    "    downstream_idx = het_sites[het_sites].index[0]\n",
    "    ref, alt, mat, pat = df_down_child.loc[downstream_idx, ['ref', 'alt', 'mat', 'pat']]\n",
    "    \n",
    "    mat_genotypes = df_down_child.loc[downstream_idx, mat_id].split('|')\n",
    "    pat_genotypes = df_down_child.loc[downstream_idx, pat_id].split('|')\n",
    "    \n",
    "    if mat_genotypes[0] == '0':\n",
    "        mat_hap1 = mat_hap1 + ref\n",
    "    elif mat_genotypes[0] == '1':\n",
    "        mat_hap1 = mat_hap1 + alt\n",
    "    else:\n",
    "        mat_hap1 = mat_hap1 + 'N'\n",
    "        \n",
    "    if mat_genotypes[1] == '0':\n",
    "        mat_hap2 = mat_hap2 + ref\n",
    "    elif mat_genotypes[1] == '1':\n",
    "        mat_hap2 = mat_hap2 + alt\n",
    "    else:\n",
    "        mat_hap2 = mat_hap2 + 'N'\n",
    "    \n",
    "    if pat_genotypes[0] == '0':\n",
    "        pat_hap1 = pat_hap1 + ref\n",
    "    elif pat_genotypes[0] == '1':\n",
    "        pat_hap1 = pat_hap1 + alt\n",
    "    else:\n",
    "        pat_hap1 = pat_hap1 + 'N'\n",
    "        \n",
    "    if pat_genotypes[1] == '0':\n",
    "        pat_hap2 = pat_hap2 + ref\n",
    "    elif pat_genotypes[1] == '1':\n",
    "        pat_hap2 = pat_hap2 + alt\n",
    "    else:\n",
    "        pat_hap2 = pat_hap2 + 'N'\n",
    "    \n",
    "    if pat == '0':\n",
    "        child_hap1 = child_hap1 + ref\n",
    "        child_hap2 = child_hap2 + alt\n",
    "    elif pat == '1':\n",
    "        child_hap1 = child_hap1 + alt\n",
    "        child_hap2 = child_hap2 + ref\n",
    "    else:\n",
    "        break\n",
    "    \n",
    "    trio_haps[child] = [child_hap1, child_hap2, pat_hap1, pat_hap2, mat_hap1, mat_hap2]\n",
    "    \n",
    "for v in trio_haps.values():\n",
    "    print(v)\n",
    "    \n",
    "# So after checking the phased vcfs it is always correct that (for these trios) the phasing designation is pat|mat\n",
    "# Given this information we don't need to check for het sites\n",
    "# However if we do check there are some alleles that don't perfectly match the snp haplotypes.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RETAIN THIS!\n",
    "# gamcc_hla = pd.read_csv('/well/band/users/rbx225/GAMCC/data/hla_direct_sequencing/HLA_direct_sequencing_all.csv')\n",
    "# gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Included Alleles']]\n",
    "# for s in gamcc_hla['SampleID'].unique():\n",
    "#     tmps = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "#     for l in HLA_GENES:\n",
    "#         tmpl = tmps[tmps['Locus'] == l]\n",
    "#         repeat = 2 - tmpl.shape[0]\n",
    "#         if repeat == 2:\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, np.nan]\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, np.nan]\n",
    "#         if repeat == 1:\n",
    "#             gamcc_hla.loc[len(gamcc_hla)] = [s, l, tmpl.iloc[0,2]]\n",
    "                \n",
    "# gamcc_hla = gamcc_hla[gamcc_hla['Locus'] == gene]\n",
    "# gamcc_hla = gamcc_hla[['SampleID', 'Included Alleles']]\n",
    "\n",
    "# gamcc_hla = pd.merge(gamcc_hla[::2], gamcc_hla[1::2], on = 'SampleID')\n",
    "# gamcc_hla.columns = ['Sample ID', 'HLA-DRB1 1', 'HLA-DRB1 2']\n",
    "# gamcc_hla\n",
    "\n",
    "# x = gamcc_hla.iloc[0, 1]\n",
    "# indices = []\n",
    "# db_alleles = db.index.to_numpy()\n",
    "\n",
    "# for a in x.split('/'):\n",
    "#     old_a = a\n",
    "#     while (gene + '*' + a) not in db.index and ':' in a:\n",
    "#         a = a.rsplit(':', 1)[0]\n",
    "#     if ':' in a:\n",
    "#         indices.append(np.where(db_alleles == gene + '*' + a)[0][0])\n",
    "\n",
    "# x = db.iloc[np.array(indices), :]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
